---
title: "Bulk gene and transcript identification analysis (20M reads)"
author: "Yupei & Ash"
version: "v2.0"
output:
  html_document:
    code_folding: hide
    toc: yes
    toc_float: yes
    number_sections: yes
    toc_depth: 4
  pdf_document:
    toc: yes
    toc_depth: '4'
params:
  random_seed: 2024
  sirv_ercc_gtf: '/vast/projects/LongBench/reference_files/SIRV_Set4_Norm_Sequences_20210507/SIRV_ERCC_longSIRV_multi-fasta_20210507.gtf'
  sequins_gtf: '/vast/projects/LongBench/reference_files/Sequin_resources/v2.4/rnasequin_annotation_2.4.gtf'
  sequins_tsv: '/vast/projects/LongBench/reference_files/Sequin_resources/v2.4/rnasequin_isoforms_2.4.tsv'
  human_gtf: '/vast/projects/LongBench/reference_files/GRCh38/gencode.v44.annotation.gtf'
  sirv_csv: '/vast/projects/LongBench/reference_files/SIRV_Set4_Norm_Sequences_20210507/SIRV_set4_concentration.csv'
  ont_bulk_oarfish_dir: '/vast/projects/LongBench/analysis/main_workflow/result/rarefraction_analysis/oarfish/ont_bulk/20M'
  pb_bulk_oarfish_dir: '/vast/projects/LongBench/analysis/main_workflow/result/rarefraction_analysis/oarfish/pb_bulk/20M'
  ont_drnd_oarfish_dir: '/vast/projects/LongBench/analysis/main_workflow/result/rarefraction_analysis/oarfish/dRNA_bulk/20M'
  ill_bulk_salmon_dir: '/vast/projects/LongBench/analysis/main_workflow/result/rarefraction_analysis/salmon/20M'
  Tx2Gene.map.rds: "/vast/projects/LongBench/analysis/workflow/rmarkdown/RDS/Tx2Gene.map.rds"
  bulk_meta: '/vast/projects/LongBench/sequencing_data/illumina_bulk/metadata.txt'
  cache_dir: NULL
  fig.path: 'figures/'
  out.rds: '/vast/projects/LongBench/analysis/reports/RDS/bulk_de_human_summary.rds'
---


# Setup:
```{r Input filename setup, include=FALSE, eval=F}
# This chunk is used only when running the file chunk by chunk manually
params <- list(
  random_seed = 2024,
  cache_dir = NULL,
  sirv_ercc_gtf = '/vast/projects/LongBench/reference_files/SIRV_Set4_Norm_Sequences_20210507/SIRV_ERCC_longSIRV_multi-fasta_20210507.gtf',
  sequins_gtf = '/vast/projects/LongBench/reference_files/Sequin_resources/v2.4/rnasequin_annotation_2.4.gtf',
  sequins_tsv = '/vast/projects/LongBench/reference_files/Sequin_resources/v2.4/rnasequin_isoforms_2.4.tsv',
  human_gtf = '/vast/projects/LongBench/reference_files/GRCh38/gencode.v44.annotation.gtf',
  sirv_csv = '/vast/projects/LongBench/reference_files/SIRV_Set4_Norm_Sequences_20210507/SIRV_set4_concentration.csv',
  ont_bulk_oarfish_dir = '/vast/projects/LongBench/analysis/main_workflow/result/rarefraction_analysis/oarfish/ont_bulk/20M',
  pb_bulk_oarfish_dir = '/vast/projects/LongBench/analysis/main_workflow/result/rarefraction_analysis/oarfish/pb_bulk/20M',
  ont_drnd_oarfish_dir = '/vast/projects/LongBench/analysis/main_workflow/result/rarefraction_analysis/oarfish/dRNA_bulk/20M',
  ill_bulk_salmon_dir = '/vast/projects/LongBench/analysis/main_workflow/result/rarefraction_analysis/salmon/20M',
  bulk_meta = "/vast/projects/LongBench/sequencing_data/illumina_bulk/metadata.txt",
  out.rds= '/vast/projects/LongBench/analysis/reports/RDS/bulk_de_human_summary.rds'
)
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  cache = TRUE,
  fig.width = 15, 
  fig.height = 10, 
  warning = F, 
  fig.path = params$fig.path, 
  dev = c("png", "svg"),
  cache.path = ifelse(is.null(params$cache_dir), 
                      paste0(tools::file_path_sans_ext(knitr::current_input()), "_cache/"), 
                      params$cache_dir))
```


```{r Libraries,include=FALSE}
library(limma)
library(edgeR)
library(tidyverse)
library(RColorBrewer)
library(dplyr)
library(tidyr)
library(ggplot2)
library(ggrepel)
library(glue)
library(ggplot2)
library(gridExtra)
library(ggExtra)
library(knitr)
library(kableExtra)
library(rtracklayer)
library(scales)
library(UpSetR)
library(ggtree)
library(aplot)
library(patchwork)
source("/vast/projects/LongBench/analysis/workflow/scripts/Rfunctions.R")

# setup some conflict functions
rename <- dplyr::rename
select <- dplyr::select

# Set color palette
color_palette <- c(
  PacBio = "#df1995",
  ONT = "#00789b",
  ONT_1 = "#04476c",
  ONT_2 = "#00789b",
  Illumina = "#e88b20"
)
```

```{r function for splitting the transcript origin}
is.sequins <- function(dge) {grepl("^R", rownames(dge))}
is.human <- function(dge) {grepl("^ENS", rownames(dge))}
is.sirv_or_ercc <- function(dge) {!is.human(dge) & !is.sequins(dge)} # thie is the whole SIRV set4 (ERCC + SIRV E0 + Long SIRV)
is.sirv_E0 <- function(dge) {grepl("^SIRV\\d{3}$", rownames(dge))} 
is.sirv_all <- function(dge) {grepl("^SIRV", rownames(dge))} # including all SIRV E0 and Long SIRV
is.long_sirv <- function(dge) {grepl("^SIRV\\d{4}", rownames(dge))}
is.ercc <- function(dge) {!is.human(dge) & !is.sequins(dge) & !is.sirv_all(dge)}
```

```{r, ge transcripts to gene map}
list2env(readRDS(params$Tx2Gene.map.rds), .GlobalEnv)
# This will load 
# "SIRV.G.Tx.map"  "sequins.G.Tx.map" "human.G.Tx.map"  
```

# Load quantification
```{r Create DGE from input, echo=F}
bulk.meta <- read.csv(params$bulk_meta)
rownames(bulk.meta) <- bulk.meta$sample

# SR salmon
ill_bulk.dge <- get_dge_from_salmon(params$ill_bulk_salmon_dir, ".*/")
# LR oarfish
ont_bulk.dge.oarfish <- get_dge_from_oarfish(params$ont_bulk_oarfish_dir, ".*/")
pb_bulk.dge.oarfish <- get_dge_from_oarfish(params$pb_bulk_oarfish_dir, ".*/")
ont_drna.dge.oarfish <- get_dge_from_oarfish(params$ont_drnd_oarfish_dir, ".*/")

# Extract human transcript counts
ill_bulk.dge <- ill_bulk.dge[is.human(ill_bulk.dge), ]
ont_bulk.dge.oarfish <- ont_bulk.dge.oarfish[is.human(ont_bulk.dge.oarfish), ]
pb_bulk.dge.oarfish <- pb_bulk.dge.oarfish[is.human(pb_bulk.dge.oarfish), ]
ont_drna.dge.oarfish <- ont_drna.dge.oarfish[is.human(ont_drna.dge.oarfish), ]
```


### 1. Aggregate transcript counts to gene level using `tximport` with unfiltered transcript counts
```{r, echo=F}
library(tximport) # install latest version from github

# Extract tx_name and gene_id columns from human and sequins
human_tx2gene <- human.G.Tx.map[, c("tx_name", "gene_id")]
sequins_tx2gene <- sequins.G.Tx.map[, c("tx_name", "gene_id")]
SIRV_tx2gene <- SIRV.G.Tx.map[, c("tx_name", "gene_id")]
combined_tx2gene <- rbind(human_tx2gene, sequins_tx2gene,SIRV_tx2gene)

# Get quant.sf file paths
get_quant_paths <- function(dirs, file_name) {
  # Generate file paths for quantification files
  quant_paths <- file.path(dirs, file_name)
  return(quant_paths)
}

# Define directories with their corresponding expected file names
oarfish_dirs <- list(
  ill_bulk = list(dir = params$ill_bulk_salmon_dir, file_name = "quant.sf"),
  ont_bulk = list(dir = params$ont_bulk_oarfish_dir, file_name = ".quant"),
  pb_bulk = list(dir = params$pb_bulk_oarfish_dir, file_name = ".quant"),
  drna_bulk = list(dir = params$ont_drnd_oarfish_dir, file_name = ".quant")
)

# Apply function to each directory to get all quant file paths
quant_paths <- lapply(oarfish_dirs, function(dir_info) {
  # Get the directory and file name from dir_info
  dir <- dir_info$dir
  file_name <- dir_info$file_name
  
  # List all subdirectories and get the quant file paths
  sub_dirs <- list.dirs(dir, full.names = TRUE, recursive = FALSE)
  get_quant_paths(sub_dirs, file_name)
})

# Flatten the list of quant_paths into a named list
quant_paths <- lapply(quant_paths, unlist)

# Function to import DGE from tximport
get_dge_from_txi <- function(quant_dir, sample_prefix_regex, type) {
  # Import data using tximport
  txi <- tximport(quant_dir,
                  type = type,
                  tx2gene = combined_tx2gene,
                  ignoreAfterBar = TRUE,
                  countsFromAbundance = "no") # raw counts
  counts <- as.data.frame(txi$counts, stringAsFactors = FALSE)
  sample_names <- gsub(sample_prefix_regex, '', basename(dirname(quant_dir)))
  colnames(counts) <- sample_names
  rst.dge <- DGEList(counts = counts)
  rst.dge$genes <- data.frame(Length=txi$length %>% rowMeans())
  rownames(rst.dge$samples) <- sub(sample_prefix_regex, '', rownames(rst.dge$samples))
  colnames(rst.dge$counts) <- sub(sample_prefix_regex, "", colnames(rst.dge$counts))
  # colnames(rst.dge$gene$Length) <- sub(sample_prefix_regex, "", colnames(rst.dge$gene$Length))
  
  rst.dge$samples <- merge(rst.dge$samples %>% select(-group), bulk.meta, by = "row.names", all.x = TRUE) %>% select(-Row.names)
  rst.dge$samples <- rst.dge$samples[match(colnames(rst.dge$counts), rst.dge$samples$sample),]
  return(rst.dge)
}

# Process each quant path for different bulk types
ill_bulk.gene.dge <- get_dge_from_txi(quant_paths$ill_bulk, ".*/", "salmon")
ont_bulk.gene.dge <- get_dge_from_txi(quant_paths$ont_bulk, ".*/", "oarfish")
pb_bulk.gene.dge <- get_dge_from_txi(quant_paths$pb_bulk, ".*/", "oarfish")
drna_bulk.gene.dge <- get_dge_from_txi(quant_paths$drna_bulk, ".*/", "oarfish")

# Extract human gene counts
ill_bulk.gene.dge <- ill_bulk.gene.dge[is.human(ill_bulk.gene.dge), ]
ont_bulk.gene.dge <- ont_bulk.gene.dge[is.human(ont_bulk.gene.dge), ]
pb_bulk.gene.dge <- pb_bulk.gene.dge[is.human(pb_bulk.gene.dge), ]
drna_bulk.gene.dge <- drna_bulk.gene.dge[is.human(drna_bulk.gene.dge), ]

# remove .* 
rownames(ill_bulk.gene.dge) <- ill_bulk.gene.dge %>% rownames %>% sub("\\..*", "", .)
rownames(ont_bulk.gene.dge) <- ont_bulk.gene.dge %>% rownames %>% sub("\\..*", "", .)
rownames(pb_bulk.gene.dge) <- pb_bulk.gene.dge %>% rownames %>% sub("\\..*", "", .)
rownames(drna_bulk.gene.dge) <- drna_bulk.gene.dge %>% rownames %>% sub("\\..*", "", .)
```
# Gene discovery analysis
## Get `gene_discovery` 

Based on `filterByExpr` for each dataset, using default parameters
```{r gene identification}
# 1. Get union (superset) of genes
all_genes <- Reduce(union, lapply(
  list(ill_bulk.gene.dge, ont_bulk.gene.dge, pb_bulk.gene.dge, drna_bulk.gene.dge),
  rownames
))

# Pad each DGE with missing genes (set to 0)
pad_dge <- function(dge, all_genes) {
  missing <- setdiff(all_genes, rownames(dge))
  if (length(missing) > 0) {
    pad <- matrix(0, nrow = length(missing), ncol = ncol(dge$counts),
                  dimnames = list(missing, colnames(dge$counts)))
    dge$counts <- rbind(dge$counts, pad)
  }
  dge[all_genes, ]
}

ill_bulk.gene.dge <- pad_dge(ill_bulk.gene.dge, all_genes)
ont_bulk.gene.dge <- pad_dge(ont_bulk.gene.dge, all_genes)
pb_bulk.gene.dge  <- pad_dge(pb_bulk.gene.dge, all_genes)
drna_bulk.gene.dge <- pad_dge(drna_bulk.gene.dge, all_genes)

# 2. Gene discovery plots
gene_discovery <- tibble(
  gene_id   = all_genes,
  Illumina  = filterByExpr(ill_bulk.gene.dge, group = ill_bulk.gene.dge$samples$group),
  `ONT cDNA`= filterByExpr(ont_bulk.gene.dge, group = ont_bulk.gene.dge$samples$group),
  PacBio    = filterByExpr(pb_bulk.gene.dge, group = pb_bulk.gene.dge$samples$group),
  `ONT dRNA`= filterByExpr(drna_bulk.gene.dge, group = drna_bulk.gene.dge$samples$group)
) %>%
  filter(Illumina | `ONT cDNA` | PacBio | `ONT dRNA`)
```


###  Add biotype and gene length
Let's adding some more feature for plotting

```{r}
# Getting the gene biotype and gene id, using illumina data here but will be identical for all datasets
tx_ref_df <- ill_bulk.dge$genes %>%
  rownames_to_column("tx_id") %>%
  select(tx_id, Length) %>%
  tibble() %>%
  mutate(
    biotype = strsplit(tx_id, "\\|") %>% sapply(function(x) x[8]),
    gene_id = strsplit(tx_id, "\\|") %>% sapply(function(x) x[2]),
    tx_id = strsplit(tx_id, "\\|") %>% sapply(function(x) x[1])
  )

gene_length_lookup <- tx_ref_df %>%
    group_by(gene_id) %>%
    summarise(mean_length = mean(Length, na.rm = TRUE), .groups = "drop") %>%
    mutate(gene_id= sub("\\..*", "", gene_id))

# Connect to Ensembl
library(biomaRt)
ensembl <- useEnsembl(biomart = "genes", dataset = "hsapiens_gene_ensembl")

gene_discovery <- gene_discovery %>%
    left_join(gene_length_lookup, by = c("gene_id")) %>%
    rename(`Gene length` = mean_length)
gene_discovery$gene_id <- gene_discovery$gene_id %>% sub("\\..*", "", .)
biotypes <- getBM(
  attributes = c("ensembl_gene_id", "gene_biotype"),
  filters = "ensembl_gene_id",
  values = gene_discovery$gene_id,
  mart = ensembl
)


gene_discovery <- gene_discovery %>%
  left_join(biotypes, by = c("gene_id" = "ensembl_gene_id")) %>%
  rename(biotype = gene_biotype) %>%  
  mutate(biotype = ifelse(biotype %in% 
            names(sort(table(biotype), decreasing = TRUE)[1:6]), 
            biotype, "Others"))


gene_discovery$biotype <- factor(gene_discovery$biotype, levels = c(sort(unique(gene_discovery$biotype)) %>% setdiff("Others"), "Others"))
```

### Add intronic Gene and exon count
Boolean for intronic genes. Each gene is checked if it is located inside the intron of the other gene. The specific definition:
`findOverlaps(exons, merged_introns, ignore.strand=TRUE, type='within',  select="all", minoverlap =50)`
```{r Add intronic gene and exon count}

rds_list <- readRDS("/home/users/allstaff/you.yu/LongBench/analysis/workflow/rmarkdown/RDS/intronic_gene_and_exon_count.rds")
overlaps_tibble <- rds_list$exon_intron_overlaps
overlaps_tibble$exon_gene <- sub("\\..*","", overlaps_tibble$exon_gene)
overlaps_tibble$intron_gene <- sub("\\..*", "", overlaps_tibble$intron_gene)

exons_count <- rds_list$exons_count  %>% mutate(Gene = sub("\\..*", "", Gene))

rm(rds_list)

# Merge the intronic gene information into the gene_discovery tibble
gene_discovery <- gene_discovery %>% mutate(
  "Intronic Gene" = gene_discovery$gene_id %in% overlaps_tibble[overlaps_tibble$intron_gene %in% gene_discovery$gene_id,]$exon_gene
)  %>% left_join(exons_count, by = c("gene_id" = "Gene"))
```

## Gene discovery plots

###  Main Upset plot
```{r GENE: Main Upset plot, fig.height=4, fig.width=8}
library(ComplexUpset)
ComplexUpset::upset(
  gene_discovery %>% mutate(
    biotype = case_when(
      biotype =="protein_coding" ~ "Protein-coding",
      TRUE ~ "Non-protein-coding"
    )
  ),
  intersect = c('Illumina', "ONT cDNA", "PacBio", "ONT dRNA"),
  annotations = list(
        # 1st method - passing list:
        "Gene length" = list(
          aes = aes(x = intersection, y = `Gene length`),
          # provide a list if you wish to add several geoms
          geom = geom_boxplot(na.rm = TRUE, outliers = FALSE)
        )
        # ,
        # "BioType" = list(
        #   aes = aes(x = intersection, fill = biotype),
        #   geom = geom_bar(position = "fill")
        # )
  ),
  base_annotations=list(
        'Intersection size'=ComplexUpset::intersection_size(
            counts=TRUE,
            mapping=aes(fill=biotype),
            text=element_text(size = 3, vjust=-0.2)
        ) + scale_fill_manual(values = c("Non-protein-coding"="#BAC5CB",
                              "Protein-coding"='#B5C952'))
    ),
  set_sizes=(
        ComplexUpset::upset_set_size(
            geom=geom_bar(
                aes(fill=biotype),
                width=0.8
            ) ,
            position='right'
        )+ scale_fill_manual(values = c("Non-protein-coding"="#BAC5CB",
                              "Protein-coding"='#B5C952')) + 
          theme(legend.position = "none")
  ),
  guides='over',
  wrap = TRUE,
  sort_intersections_by=c('degree', 'cardinality')
)

```

### Upset plot with addition feature

```{r Prepare plot_df and intersetion order}
plot_df <- gene_discovery %>% mutate(
  "Intronic Gene" = gene_discovery$gene_id %in% overlaps_tibble[overlaps_tibble$intron_gene %in% gene_discovery$gene_id,]$exon_gene,
  "Single exon" = gene_discovery$Exon_Count == 1
)


intersection_summary <- plot_df %>%
  group_by(Illumina, `ONT cDNA`, PacBio, `ONT dRNA`) %>%
  summarise(prop_intronic = mean(`Intronic Gene`, na.rm = TRUE),
            mean_gene_length = mean(`Gene length`, na.rm = TRUE), .groups = "drop")

platform_cols <- c("Illumina", "ONT cDNA", "PacBio", "ONT dRNA")

intersection_summary <- intersection_summary %>%
  rowwise() %>%
  mutate(
    intersection_list = list(platform_cols[unlist(across(all_of(platform_cols)))])
  ) %>%
  ungroup()

ordered_intersections_by_intronic_gene <- intersection_summary %>%
  arrange(desc(prop_intronic)) %>% pull(intersection_list)

ordered_intersections_by_gene_length <- intersection_summary %>%
  arrange(desc(mean_gene_length)) %>% pull(intersection_list)

```

```{r,  Expression of the intronic and single-exon gene, fig.height=6, fig.width=10}
ComplexUpset::upset(
  plot_df,
  intersect = c('Illumina', "ONT cDNA", "PacBio", "ONT dRNA"),
  annotations = list(
        "Exon counts" = list(
            aes = aes(x = intersection, y = Exon_Count),
            # provide a list if you wish to add several geoms
            geom = geom_boxplot(na.rm = TRUE, outliers = FALSE)
          ),
        # "Single-exon Genes" = (
        #     ggplot(mapping=aes(fill = `Single exon`)) +
        #       geom_bar(position = "fill")+
        #       scale_fill_manual(values = c("#BAC5CB",'#B5C952')) + 
        #       ylab('Single-exon gene proportion')
        #   ),
        "Intronic Genes" = (
            ggplot(mapping=aes(fill = `Intronic Gene`)) +
              geom_bar(position = "fill")+
              scale_fill_manual(values = c("#BAC5CB",'#B5C952')) + 
              ylab('Intronic gene proportion')
          )
  ),
  set_sizes=(
        ComplexUpset::upset_set_size(
            geom=geom_bar(
                width=0.8
            )
        )+ theme(legend.position = "none")
    ),
  wrap = TRUE,
  sort_intersections=FALSE,
  intersections = ordered_intersections_by_intronic_gene
)

```




### Jaccard similarity
```{r gene identification Jaccard similarity, fig.height=3, fig.width=4.5}

# Jaccard similarity
library(ComplexHeatmap)
binary_matrix <- gene_discovery %>% 
  select( Illumina, `ONT cDNA`, PacBio ,`ONT dRNA`) %>% 
  mutate(across(everything(), as.integer))

# Compute Jaccard similarity
jaccard_sim <- function(x, y) {
  sum(x & y) / sum(x | y)
}

platforms <- colnames(binary_matrix)
jaccard_matrix <- outer(platforms, platforms, Vectorize(function(i, j) {
  jaccard_sim(binary_matrix[[i]], binary_matrix[[j]])
}))

# Convert to matrix format
colnames(jaccard_matrix) <- rownames(jaccard_matrix) <- platforms

label_func <- function(j, i, x, y, width, height, fill) {
    grid.text(sprintf("%.2f", jaccard_matrix[i, j]), x, y, gp = gpar(fontsize = 10))
}
Heatmap(jaccard_matrix, 
        name = "Jaccard Similarity", 
        col = colorRampPalette(c("grey", "#06317F"))(50),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(round(jaccard_matrix[i, j], 2), x, y, 
                    gp = gpar(col = "white"))  # Set font color to white
        },
        na_col = "white")  # Keeps upper triangle blank
```

### Detected and undetected gene length distribution
```{r GENE: discovered gene distribution, fig.height=3, fig.width=5}
gene_discovery$length_bin <- cut(gene_discovery$`Gene length`, 
                                 breaks = c(0, 500, 1000, 1500, 2000, 2500, Inf), 
                                 labels = c("0-0.5k", "0.5-1k", "1-1.5k", "1.5-2k", "2-2.5k", "2.5k+"))


p1 <- gene_discovery %>%
  pivot_longer(cols = c("Illumina", "ONT cDNA", "PacBio", "ONT dRNA"), 
               names_to = "platform", values_to = "keep") %>%
  filter(keep) %>%
  filter(!is.na(length_bin)) %>% 
  count(length_bin, platform) %>%  # Summarize counts per category
  ggplot(aes(x = length_bin, y = n, color = platform, group = platform)) +
  # platform-specific lines
  geom_line(linewidth = 1) +
  geom_point(size = 2) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) + 
  labs(title = "Detected Gene", 
       x = "Gene length (average transcript length)", 
       y = "# of Gene detected") + 
  scale_color_manual(values = color_palette[c("Illumina", "ONT_1", "ONT_2", "PacBio")] %>% unname())
```


```{r Plarform specific absense gene, fig.height=3, fig.width=8}

p2 <- gene_discovery %>%
  pivot_longer(cols = c("Illumina", "ONT cDNA", "PacBio", "ONT dRNA"),
               names_to = "platform", values_to = "keep") %>%
  group_by(gene_id) %>%
  mutate(n_detected = sum(keep)) %>%
  ungroup() %>%
  filter(keep & n_detected == 1) %>%   # detected in exactly 1 platform, absent in the other 3
  filter(!is.na(length_bin)) %>%
  count(length_bin, platform) %>%
  ggplot(aes(x = length_bin, y = n, color = platform, group = platform)) +
  geom_line(size = 1) +
  geom_point(size = 2) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) + 
  labs(title = "Platform-specific Detection",
       x = "Gene length (average transcript length)",
       y = "# of Gene detected") +
  scale_color_manual(values = color_palette[c("Illumina", "ONT_1", "ONT_2", "PacBio")] %>% unname())

p3 <- gene_discovery %>%
  pivot_longer(cols = c("Illumina", "ONT cDNA", "PacBio", "ONT dRNA"),
               names_to = "platform", values_to = "keep") %>%
  group_by(gene_id) %>%
  mutate(n_detected = sum(keep)) %>%
  ungroup() %>%
  filter(!keep & n_detected == 3) %>%   # absent in exactly 1 platform, present in the other 3
  filter(!is.na(length_bin)) %>% 
  count(length_bin, platform) %>%
  ggplot(aes(x = length_bin, y = n, color = platform, group = platform)) +
  geom_line(size = 1) +
  geom_point(size = 2) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) + 
  labs(title = "Platform-specific Absence",
       x = "Gene length (average transcript length)",
       y = "# of Gene absent") +
  scale_color_manual(values = color_palette[c("Illumina", "ONT_1", "ONT_2", "PacBio")] %>% unname())

(p1 | p2 | p3) + plot_layout(guides = "collect")
```

# Tx discovery analysis
## Get `tx_discovery` 
```{r tx filtering}
all_tx <- Reduce(union, lapply(list(ill_bulk.dge, ont_bulk.dge.oarfish, pb_bulk.dge.oarfish, ont_drna.dge.oarfish), rownames))


ill_bulk.dge <- pad_dge(ill_bulk.dge, all_tx)
ont_bulk.dge.oarfish <- pad_dge(ont_bulk.dge.oarfish, all_tx)
pb_bulk.dge.oarfish <- pad_dge(pb_bulk.dge.oarfish, all_tx)
ont_drna.dge.oarfish <- pad_dge(ont_drna.dge.oarfish, all_tx)

# Filter by expression
tx_discovery <- tibble(
  tx_id = all_tx,
  "Illumina" = filterByExpr(ill_bulk.dge, group = ill_bulk.dge$samples$group, min.count=5),
  "ONT cDNA" = filterByExpr(ont_bulk.dge.oarfish, group = ont_bulk.dge.oarfish$samples$group, min.count=5),
  "PacBio" = filterByExpr(pb_bulk.dge.oarfish, group = pb_bulk.dge.oarfish$samples$group, min.count=5),
  "ONT dRNA" = filterByExpr(ont_drna.dge.oarfish, group = ont_drna.dge.oarfish$samples$group, min.count=5)
) %>% filter(
  Illumina | `ONT cDNA` | PacBio | `ONT dRNA`
)
# split and get biotype (last element of tx_id)
tx_discovery <- tx_discovery %>%  mutate(
    biotype = strsplit(tx_id, "\\|") %>% sapply(function(x) x[8]),
    length = strsplit(tx_id, "\\|") %>% sapply(function(x) x[7]) %>% as.integer(),
    gene_id = strsplit(tx_id, "\\|") %>% sapply(function(x) x[2]),
    tx_id = strsplit(tx_id, "\\|") %>% sapply(function(x) x[1])
  )

tx_discovery <- tx_discovery %>% mutate(biotype = ifelse(biotype %in% 
            names(sort(table(biotype), decreasing = TRUE)[1:6]), 
            biotype, "Others"))
```

### Add overdispersion per platform

```{r}
get_overdispersion <- function(dge) {
  dge$genes %>%
    rownames_to_column("tx_id") %>%
    select(tx_id, Overdispersion) %>%
    mutate(tx_id = sapply(strsplit(tx_id, "\\|"), `[`, 1))
}

overdispersion.ill  <- get_overdispersion(ill_bulk.dge)
overdispersion.ont  <- get_overdispersion(ont_bulk.dge.oarfish)
overdispersion.pb   <- get_overdispersion(pb_bulk.dge.oarfish)
overdispersion.drna <- get_overdispersion(ont_drna.dge.oarfish)

# Merge overdispersion into tx_discovery
tx_discovery <- tx_discovery %>%
  left_join(overdispersion.ill, by = "tx_id") %>%
  rename(Overdispersion_Illumina = Overdispersion) %>%
  left_join(overdispersion.ont, by = "tx_id") %>%
  rename(Overdispersion_ONT_cDNA = Overdispersion) %>%
  left_join(overdispersion.pb, by = "tx_id") %>%
  rename(Overdispersion_PacBio = Overdispersion) %>%
  left_join(overdispersion.drna, by = "tx_id") %>%
  rename(Overdispersion_ONT_dRNA = Overdispersion)
```
## Tx discovery plots

### Main upset plot
```{r Transcript: Main upset plot, fig.height=4, fig.width=8}
library(ComplexUpset)
ComplexUpset::upset(
  tx_discovery %>% mutate(
    biotype = case_when(
      biotype =="protein_coding" ~ "Protein-coding",
      TRUE ~ "Non-protein-coding"
    )
  ),
  intersect = c('Illumina', "ONT cDNA", "PacBio", "ONT dRNA"),
  annotations = list(
        # 1st method - passing list:
        "Transcript length" = list(
          aes = aes(x = intersection, y = `length`),
          # provide a list if you wish to add several geoms
          geom = geom_boxplot(na.rm = TRUE, outliers = FALSE)
        )
        # ,
        # "BioType" = list(
        #   aes = aes(x = intersection, fill = biotype),
        #   geom = geom_bar(position = "fill")
        # )
  ),
  base_annotations=list(
        'Intersection size'=ComplexUpset::intersection_size(
            counts=TRUE,
            mapping=aes(fill=biotype),
            text=element_text(size = 3, vjust=-0.2)
        ) + scale_fill_manual(values = c("Non-protein-coding"="#BAC5CB",
                              "Protein-coding"='#B5C952'))
    ),
  set_sizes=(
        ComplexUpset::upset_set_size(
            geom=geom_bar(
                aes(fill=biotype),
                width=0.8
            ) ,
            position='right'
        )+ scale_fill_manual(values = c("Non-protein-coding"="#BAC5CB",
                              "Protein-coding"='#B5C952')) + 
          theme(legend.position = "none")
  ),
  guides='over',
  wrap = TRUE
)
```
### Upset plot with addition feature
```{r Transcript: upset plot more feature, fig.height=10, fig.width=8}
library(ComplexUpset)
ComplexUpset::upset(
  tx_discovery %>% mutate(
    biotype = case_when(
      biotype =="protein_coding" ~ "Protein-coding",
      TRUE ~ "Non-protein-coding"
    )
  ),
  intersect = c('Illumina', "ONT cDNA", "PacBio", "ONT dRNA"),
  annotations = list(
        # 1st method - passing list:
        "Transcript length" = list(
          aes = aes(x = intersection, y = `length`),
          geom = geom_boxplot(na.rm = TRUE, outliers = FALSE)
        ),
        "Illumina" = list(
          aes = aes(x = intersection, y = Overdispersion_Illumina),
          geom = geom_boxplot(na.rm = TRUE, outliers = FALSE)
        ),
        "ONT cDNA" = list(
          aes = aes(x = intersection, y = Overdispersion_ONT_cDNA),
          geom = geom_boxplot(na.rm = TRUE, outliers = FALSE)
        ),
        "PacBio" = list(
          aes = aes(x = intersection, y = Overdispersion_PacBio),
          geom = geom_boxplot(na.rm = TRUE, outliers = FALSE)
        ),
        "ONT dRNA" = list(
          aes = aes(x = intersection, y = Overdispersion_ONT_dRNA),
          geom = geom_boxplot(na.rm = TRUE, outliers = FALSE)
        )
  ),
  base_annotations=list(
        'Intersection size'=ComplexUpset::intersection_size(
            counts=TRUE,
            mapping=aes(fill=biotype),
            text=element_text(size = 3, vjust=-0.2)
        ) + scale_fill_manual(values = c("Non-protein-coding"="#BAC5CB",
                              "Protein-coding"='#B5C952'))
    ),
  set_sizes=(
        ComplexUpset::upset_set_size(
            geom=geom_bar(
                aes(fill=biotype),
                width=0.8
            ) ,
            position='right'
        )+ scale_fill_manual(values = c("Non-protein-coding"="#BAC5CB",
                              "Protein-coding"='#B5C952')) + 
          theme(legend.position = "none")
  ),
  guides='over',
  wrap = TRUE
)
```


## Jaccard similary
```{r transcript identification Jaccard similarity, fig.height=3, fig.width=4.5}

# Jaccard similarity
library(ComplexHeatmap)
binary_matrix <- tx_discovery %>% 
  select( Illumina, `ONT cDNA`, PacBio ,`ONT dRNA`) %>% 
  mutate(across(everything(), as.integer))

# Compute Jaccard similarity
jaccard_sim <- function(x, y) {
  sum(x & y) / sum(x | y)
}

platforms <- colnames(binary_matrix)
jaccard_matrix <- outer(platforms, platforms, Vectorize(function(i, j) {
  jaccard_sim(binary_matrix[[i]], binary_matrix[[j]])
}))

# Convert to matrix format
colnames(jaccard_matrix) <- rownames(jaccard_matrix) <- platforms

label_func <- function(j, i, x, y, width, height, fill) {
    grid.text(sprintf("%.2f", jaccard_matrix[i, j]), x, y, gp = gpar(fontsize = 10))
}
Heatmap(jaccard_matrix, 
        name = "Jaccard Similarity", 
        col = colorRampPalette(c("grey", "#06317F"))(50),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(round(jaccard_matrix[i, j], 2), x, y, 
                    gp = gpar(col = "white"))  # Set font color to white
        },
        na_col = "white")  # Keeps upper triangle blank
```

## Detected and Undetected transcript length
```{r Discovered transcript length dist, fig.height=3, fig.width=5}
tx_discovery$length_bin <- cut(tx_discovery$length, breaks = c(0, 500, 1000, 1500, 2000, 2500,Inf), labels = c("0-0.5k", "0.5-1k", "1-1.5k", "1.5-2k", "2-2.5k", "2.5k+"))


p1 <- tx_discovery %>%
  pivot_longer(cols = c("Illumina", "ONT cDNA", "PacBio", "ONT dRNA"), 
               names_to = "platform", values_to = "keep") %>%
  filter(keep) %>%
  count(length_bin, platform) %>%  # Summarize counts per category
  ggplot(aes(x = length_bin, y = n, color = platform, group = platform)) +
  # platform-specific lines
  geom_line(linewidth = 1) +
  geom_point(size = 2) +
  theme_minimal() +  
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) + 
  labs(title = "Detected Transcript", 
       x = "Transcript length", 
       y = "# of transcripts detected") + 
  scale_color_manual(values = color_palette[c("Illumina", "ONT_1", "ONT_2", "PacBio")] %>% unname())

```
```{r Plarform specific absense tx, fig.height=3, fig.width=8}

# Platform-specific detection
p2 <- tx_discovery %>%
  pivot_longer(cols = c("Illumina", "ONT cDNA", "PacBio", "ONT dRNA"),
               names_to = "platform", values_to = "keep") %>%
  group_by(tx_id) %>%
  mutate(n_detected = sum(keep)) %>%
  ungroup() %>%
  filter(keep & n_detected == 1) %>%   # detected in exactly 1 platform, absent in the other 3
  filter(!is.na(length_bin)) %>%
  count(length_bin, platform) %>%
  ggplot(aes(x = length_bin, y = n, color = platform, group = platform)) +
  geom_line(size = 1) +
  geom_point(size = 2) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) + 
  labs(title = "Platform-specific Detection",
       x = "Transcript length",
       y = "# of transcripts detected") +
  scale_color_manual(values = color_palette[c("Illumina", "ONT_1", "ONT_2", "PacBio")] %>% unname())


p3 <- tx_discovery %>%
  pivot_longer(cols = c("Illumina", "ONT cDNA", "PacBio", "ONT dRNA"),
               names_to = "platform", values_to = "keep") %>%
  group_by(tx_id) %>%
  mutate(n_detected = sum(keep)) %>%
  ungroup() %>%
  filter(!keep & n_detected == 3) %>%   # absent in exactly 1 platform, present in the other 3
  count(length_bin, platform) %>%
  ggplot(aes(x = length_bin, y = n, color = platform, group = platform)) +
  geom_line(size = 1) +
  geom_point(size = 2) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) + 
  labs(title = "Platform-specific Absence",
       x = "Transcript length",
       y = "# of transcripts absent") +
  scale_color_manual(values = color_palette[c("Illumina", "ONT_1", "ONT_2", "PacBio")] %>% unname())


(p1 | p2 | p3) + plot_layout(guides = "collect")
```


```

