---
title: "Bulk gene and transcript identification analysis (20M reads)"
author: "Yupei & Ash"
version: "v1.0"
output:
  html_document:
    code_folding: hide
    toc: yes
    toc_float: yes
    number_sections: yes
    toc_depth: 4
  pdf_document:
    toc: yes
    toc_depth: '4'
params:
  random_seed: 2024
  sirv_ercc_gtf: '/vast/projects/LongBench/reference_files/SIRV_Set4_Norm_Sequences_20210507/SIRV_ERCC_longSIRV_multi-fasta_20210507.gtf'
  sequins_gtf: '/vast/projects/LongBench/reference_files/Sequin_resources/v2.4/rnasequin_annotation_2.4.gtf'
  sequins_tsv: '/vast/projects/LongBench/reference_files/Sequin_resources/v2.4/rnasequin_isoforms_2.4.tsv'
  human_gtf: '/vast/projects/LongBench/reference_files/GRCh38/gencode.v44.annotation.gtf'
  sirv_csv: '/vast/projects/LongBench/reference_files/SIRV_Set4_Norm_Sequences_20210507/SIRV_set4_concentration.csv'
  ont_bulk_oarfish_dir: '/vast/projects/LongBench/analysis/main_workflow/result/rarefraction_analysis/oarfish/ont_bulk/20M'
  pb_bulk_oarfish_dir: '/vast/projects/LongBench/analysis/main_workflow/result/rarefraction_analysis/oarfish/pb_bulk/20M'
  ont_drnd_oarfish_dir: '/vast/projects/LongBench/analysis/main_workflow/result/rarefraction_analysis/oarfish/dRNA_bulk/20M'
  ill_bulk_salmon_dir: '/vast/projects/LongBench/analysis/main_workflow/result/rarefraction_analysis/salmon/20M'
  bulk_meta: '/vast/projects/LongBench/sequencing_data/illumina_bulk/metadata.txt'
  cache_dir: NULL
  fig.path: 'figures/'
  out.rds: '/vast/projects/LongBench/analysis/reports/RDS/bulk_de_human_summary.rds'
---


# Setup:
```{r Input filename setup, include=FALSE, eval=F}
# This chunk is used only when running the file chunk by chunk manually
params <- list(
  random_seed = 2024,
  cache_dir = NULL,
  sirv_ercc_gtf = '/vast/projects/LongBench/reference_files/SIRV_Set4_Norm_Sequences_20210507/SIRV_ERCC_longSIRV_multi-fasta_20210507.gtf',
  sequins_gtf = '/vast/projects/LongBench/reference_files/Sequin_resources/v2.4/rnasequin_annotation_2.4.gtf',
  sequins_tsv = '/vast/projects/LongBench/reference_files/Sequin_resources/v2.4/rnasequin_isoforms_2.4.tsv',
  human_gtf = '/vast/projects/LongBench/reference_files/GRCh38/gencode.v44.annotation.gtf',
  sirv_csv = '/vast/projects/LongBench/reference_files/SIRV_Set4_Norm_Sequences_20210507/SIRV_set4_concentration.csv',
  ont_bulk_oarfish_dir <- '/vast/projects/LongBench/analysis/main_workflow/result/rarefraction_analysis/oarfish/ont_bulk/20M'
  pb_bulk_oarfish_dir <- '/vast/projects/LongBench/analysis/main_workflow/result/rarefraction_analysis/oarfish/pb_bulk/20M'
  ont_drnd_oarfish_dir <- '/vast/projects/LongBench/analysis/main_workflow/result/rarefraction_analysis/oarfish/dRNA_bulk/20M'
  ill_bulk_salmon_dir <- '/vast/projects/LongBench/analysis/main_workflow/result/rarefraction_analysis/salmon/20M'
  bulk_meta = "/vast/projects/LongBench/sequencing_data/illumina_bulk/metadata.txt",
  out.rds= '/vast/projects/LongBench/analysis/reports/RDS/bulk_de_human_summary.rds'
)
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  cache = TRUE,
  fig.width = 15, 
  fig.height = 10, 
  warning = F, 
  fig.path = params$fig.path, 
  dev = c("png", "svg"),
  cache.path = ifelse(is.null(params$cache_dir), 
                      paste0(tools::file_path_sans_ext(knitr::current_input()), "_cache/"), 
                      params$cache_dir))
```


```{r Libraries,include=FALSE}
library(limma)
library(edgeR)
library(tidyverse)
library(RColorBrewer)
library(dplyr)
library(tidyr)
library(ggplot2)
library(ggrepel)
library(glue)
library(ggplot2)
library(gridExtra)
library(ggExtra)
library(knitr)
library(kableExtra)
library(rtracklayer)
library(scales)
library(UpSetR)
library(ggtree)
library(aplot)
source("/vast/projects/LongBench/analysis/workflow/scripts/Rfunctions.R")

# setup some conflict functions
rename <- dplyr::rename
select <- dplyr::select

# Set color palette
color_palette <- c(
  PacBio = "#df1995",
  ONT = "#00789b",
  ONT_1 = "#04476c",
  ONT_2 = "#24cdcd",
  Illumina = "#e88b20"
)
```

```{r function for splitting the transcript origin}
is.sequins <- function(dge) {grepl("^R", rownames(dge))}
is.human <- function(dge) {grepl("^ENS", rownames(dge))}
is.sirv_or_ercc <- function(dge) {!is.human(dge) & !is.sequins(dge)} # thie is the whole SIRV set4 (ERCC + SIRV E0 + Long SIRV)
is.sirv_E0 <- function(dge) {grepl("^SIRV\\d{3}$", rownames(dge))} 
is.sirv_all <- function(dge) {grepl("^SIRV", rownames(dge))} # including all SIRV E0 and Long SIRV
is.long_sirv <- function(dge) {grepl("^SIRV\\d{4}", rownames(dge))}
is.ercc <- function(dge) {!is.human(dge) & !is.sequins(dge) & !is.sirv_all(dge)}
```

```{r, get number of transcripts gene from gtf}
calcTxNum <- function(G.Tx.map){
  geneid = rep(NA, nrow(G.Tx.map))
  txcount = table(G.Tx.map$gene_id)
  txnum = as.numeric(txcount[match(G.Tx.map$gene_id, names(txcount))])
}
# SIRV
SIRV_tx <- GenomicFeatures::makeTxDbFromGFF(params$sirv_ercc_gtf, format = "gtf") %>% GenomicFeatures::transcripts(columns = c("tx_id", "tx_name", "gene_id"))
SIRV.G.Tx.map <- tibble(
  gene_id = SIRV_tx$gene_id %>% unlist,
  tx_name = SIRV_tx$tx_name
)
rm(SIRV_tx)
SIRV.G.Tx.map$tx_count <- SIRV.G.Tx.map %>% calcTxNum

# Sequins
sequins_tx  <- GenomicFeatures::makeTxDbFromGFF(params$sequins_gtf, format = "gtf") %>% GenomicFeatures::transcripts(columns = c("tx_id", "tx_name", "gene_id"))
sequins.G.Tx.map <- tibble(
  gene_id = sequins_tx$gene_id %>% unlist,
  tx_name = sequins_tx$tx_name
)
rm(sequins_tx)
sequins.G.Tx.map$tx_count <- sequins.G.Tx.map %>% calcTxNum

# Human
human_tx  <- GenomicFeatures::makeTxDbFromGFF(params$human_gtf, format = "gtf") %>% GenomicFeatures::transcripts(columns = c("tx_id", "tx_name", "gene_id"))
human.G.Tx.map <- tibble(
  gene_id = human_tx$gene_id %>% unlist,
  tx_name = human_tx$tx_name
)
rm(human_tx)
human.G.Tx.map$tx_count <- human.G.Tx.map %>% calcTxNum
```


# Gene discovery analysis
```{r Create DGE from input, echo=F}
bulk.meta <- read.csv(params$bulk_meta)
rownames(bulk.meta) <- bulk.meta$sample

# SR salmon
ill_bulk.dge <- get_dge_from_salmon(params$ill_bulk_salmon_dir, ".*/")
# LR oarfish
ont_bulk.dge.oarfish <- get_dge_from_oarfish(params$ont_bulk_oarfish_dir, ".*/")
pb_bulk.dge.oarfish <- get_dge_from_oarfish(params$pb_bulk_oarfish_dir, ".*/")
ont_drna.dge.oarfish <- get_dge_from_oarfish(params$ont_drnd_oarfish_dir, ".*/")

# Extract human transcript counts
ill_bulk.dge <- ill_bulk.dge[is.human(ill_bulk.dge), ]
ont_bulk.dge.oarfish <- ont_bulk.dge.oarfish[is.human(ont_bulk.dge.oarfish), ]
pb_bulk.dge.oarfish <- pb_bulk.dge.oarfish[is.human(pb_bulk.dge.oarfish), ]
ont_drna.dge.oarfish <- ont_drna.dge.oarfish[is.human(ont_drna.dge.oarfish), ]
```
### 1. Aggregate transcript counts to gene level using `tximport` with unfiltered transcript counts
```{r, echo=F}
library(tximport) # install latest version from github

# Extract tx_name and gene_id columns from human and sequins
human_tx2gene <- human.G.Tx.map[, c("tx_name", "gene_id")]
sequins_tx2gene <- sequins.G.Tx.map[, c("tx_name", "gene_id")]
SIRV_tx2gene <- SIRV.G.Tx.map[, c("tx_name", "gene_id")]
combined_tx2gene <- rbind(human_tx2gene, sequins_tx2gene,SIRV_tx2gene)

# Get quant.sf file paths
get_quant_paths <- function(dirs, file_name) {
  # Generate file paths for quantification files
  quant_paths <- file.path(dirs, file_name)
  return(quant_paths)
}

# Define directories with their corresponding expected file names
oarfish_dirs <- list(
  ill_bulk = list(dir = params$ill_bulk_salmon_dir, file_name = "quant.sf"),
  ont_bulk = list(dir = params$ont_bulk_oarfish_dir, file_name = ".quant"),
  pb_bulk = list(dir = params$pb_bulk_oarfish_dir, file_name = ".quant"),
  drna_bulk = list(dir = params$ont_drnd_oarfish_dir, file_name = ".quant")
)

# Apply function to each directory to get all quant file paths
quant_paths <- lapply(oarfish_dirs, function(dir_info) {
  # Get the directory and file name from dir_info
  dir <- dir_info$dir
  file_name <- dir_info$file_name
  
  # List all subdirectories and get the quant file paths
  sub_dirs <- list.dirs(dir, full.names = TRUE, recursive = FALSE)
  get_quant_paths(sub_dirs, file_name)
})

# Flatten the list of quant_paths into a named list
quant_paths <- lapply(quant_paths, unlist)

# Function to import DGE from tximport
get_dge_from_txi <- function(quant_dir, sample_prefix_regex, type) {
  # Import data using tximport
  txi <- tximport(quant_dir,
                  type = type,
                  tx2gene = combined_tx2gene,
                  ignoreAfterBar = TRUE,
                  countsFromAbundance = "no") # raw counts
  counts <- as.data.frame(txi$counts, stringAsFactors = FALSE)
  sample_names <- gsub(sample_prefix_regex, '', basename(dirname(quant_dir)))
  colnames(counts) <- sample_names
  rst.dge <- DGEList(counts = counts)
  rst.dge$genes <- data.frame(Length=txi$length %>% rowMeans())
  rownames(rst.dge$samples) <- sub(sample_prefix_regex, '', rownames(rst.dge$samples))
  colnames(rst.dge$counts) <- sub(sample_prefix_regex, "", colnames(rst.dge$counts))
  # colnames(rst.dge$gene$Length) <- sub(sample_prefix_regex, "", colnames(rst.dge$gene$Length))
  
  rst.dge$samples <- merge(rst.dge$samples %>% select(-group), bulk.meta, by = "row.names", all.x = TRUE) %>% select(-Row.names)
  rst.dge$samples <- rst.dge$samples[match(colnames(rst.dge$counts), rst.dge$samples$sample),]
  return(rst.dge)
}

# Process each quant path for different bulk types
ill_bulk.gene.dge <- get_dge_from_txi(quant_paths$ill_bulk, ".*/", "salmon")
ont_bulk.gene.dge <- get_dge_from_txi(quant_paths$ont_bulk, ".*/", "oarfish")
pb_bulk.gene.dge <- get_dge_from_txi(quant_paths$pb_bulk, ".*/", "oarfish")
drna_bulk.gene.dge <- get_dge_from_txi(quant_paths$drna_bulk, ".*/", "oarfish")

# Extract human gene counts
ill_bulk.gene.dge <- ill_bulk.gene.dge[is.human(ill_bulk.gene.dge), ]
ont_bulk.gene.dge <- ont_bulk.gene.dge[is.human(ont_bulk.gene.dge), ]
pb_bulk.gene.dge <- pb_bulk.gene.dge[is.human(pb_bulk.gene.dge), ]
drna_bulk.gene.dge <- drna_bulk.gene.dge[is.human(drna_bulk.gene.dge), ]

# remove .* 
rownames(ill_bulk.gene.dge) <- ill_bulk.gene.dge %>% rownames %>% sub("\\..*", "", .)
rownames(ont_bulk.gene.dge) <- ont_bulk.gene.dge %>% rownames %>% sub("\\..*", "", .)
rownames(pb_bulk.gene.dge) <- pb_bulk.gene.dge %>% rownames %>% sub("\\..*", "", .)
rownames(drna_bulk.gene.dge) <- drna_bulk.gene.dge %>% rownames %>% sub("\\..*", "", .)
```

### 2. Define "Gene discovery" based on6 `filterByExpr` for each dataset, using default parameters
```{r gene filtering for DE}
#GENE 
## 1. get common genes
common_genes <- Reduce(intersect, lapply(list(ill_bulk.gene.dge, ont_bulk.gene.dge, pb_bulk.gene.dge, drna_bulk.gene.dge), rownames))

ill_bulk.gene.dge <- ill_bulk.gene.dge[common_genes, ]
ont_bulk.gene.dge <- ont_bulk.gene.dge[common_genes, ]
pb_bulk.gene.dge <- pb_bulk.gene.dge[common_genes, ]
drna_bulk.gene.dge <- drna_bulk.gene.dge[common_genes, ]

common_tx <- Reduce(intersect, lapply(list(ill_bulk.dge, ont_bulk.dge.oarfish, pb_bulk.dge.oarfish, ont_drna.dge.oarfish), rownames))
ill_bulk.dge <- ill_bulk.dge[common_tx, ]
ont_bulk.dge.oarfish <- ont_bulk.dge.oarfish[common_tx, ]
pb_bulk.dge.oarfish <- pb_bulk.dge.oarfish[common_tx, ]
ont_drna.dge.oarfish <- ont_drna.dge.oarfish[common_tx, ]

# save all workpalce as Rdata
# save.image("/vast/scratch/users/you.yu/test/Bulk_DE_line269.Rdata")
```

### Gene discovery plots
```{r Define Gene discovery, fig.height=6, fig.width=10}
# Filter by expression
gene_discovery <- tibble(
  gene_id = common_genes,
  "Illumina" = filterByExpr(ill_bulk.gene.dge, group = ill_bulk.gene.dge$samples$group),
  "ONT cDNA" = filterByExpr(ont_bulk.gene.dge, group = ont_bulk.gene.dge$samples$group),
  "PacBio" = filterByExpr(pb_bulk.gene.dge, group = pb_bulk.gene.dge$samples$group),
  "ONT dRNA" = filterByExpr(drna_bulk.gene.dge, group = drna_bulk.gene.dge$samples$group)
) %>% filter(
  Illumina | `ONT cDNA` | PacBio | `ONT dRNA`
)

# Getting the gene biotype and gene id, using illumina data here but will be identical for all datasets
tx_ref_df <- ill_bulk.dge$genes %>%
  rownames_to_column("tx_id") %>%
  select(tx_id, Length) %>%
  tibble() %>%
  mutate(
    biotype = strsplit(tx_id, "\\|") %>% sapply(function(x) x[8]),
    gene_id = strsplit(tx_id, "\\|") %>% sapply(function(x) x[2]),
    tx_id = strsplit(tx_id, "\\|") %>% sapply(function(x) x[1])
  )

gene_length_lookup <- tx_ref_df %>%
    group_by(gene_id) %>%
    summarise(mean_length = mean(Length, na.rm = TRUE), .groups = "drop") %>%
    mutate(gene_id= sub("\\..*", "", gene_id))

# Connect to Ensembl
library(biomaRt)
ensembl <- useEnsembl(biomart = "genes", dataset = "hsapiens_gene_ensembl")

gene_discovery <- gene_discovery %>%
    left_join(gene_length_lookup, by = c("gene_id")) %>%
    rename(`Gene length` = mean_length)
gene_discovery$gene_id <- gene_discovery$gene_id %>% sub("\\..*", "", .)
biotypes <- getBM(
  attributes = c("ensembl_gene_id", "gene_biotype"),
  filters = "ensembl_gene_id",
  values = gene_discovery$gene_id,
  mart = ensembl
)


gene_discovery <- gene_discovery %>%
  left_join(biotypes, by = c("gene_id" = "ensembl_gene_id")) %>%
  rename(biotype = gene_biotype) %>%  
  mutate(biotype = ifelse(biotype %in% 
            names(sort(table(biotype), decreasing = TRUE)[1:6]), 
            biotype, "Others"))


gene_discovery$biotype <- factor(gene_discovery$biotype, levels = c(sort(unique(gene_discovery$biotype)) %>% setdiff("Others"), "Others"))
```

####  Main Upset plot
```{r GENE: Main Upset plot, fig.height=4, fig.width=8}
library(ComplexUpset)
ComplexUpset::upset(
  gene_discovery %>% mutate(
    biotype = case_when(
      biotype =="protein_coding" ~ "Protein-coding",
      TRUE ~ "Non-protein-coding"
    )
  ),
  intersect = c('Illumina', "ONT cDNA", "PacBio", "ONT dRNA"),
  annotations = list(
        # 1st method - passing list:
        "Gene length" = list(
          aes = aes(x = intersection, y = `Gene length`),
          # provide a list if you wish to add several geoms
          geom = geom_boxplot(na.rm = TRUE, outliers = FALSE)
        )
        # ,
        # "BioType" = list(
        #   aes = aes(x = intersection, fill = biotype),
        #   geom = geom_bar(position = "fill")
        # )
  ),
  base_annotations=list(
        'Intersection size'=ComplexUpset::intersection_size(
            counts=TRUE,
            mapping=aes(fill=biotype),
            text=element_text(size = 3, vjust=-0.2)
        ) + scale_fill_manual(values = c("Non-protein-coding"="#BAC5CB",
                              "Protein-coding"='#B5C952'))
    ),
  set_sizes=(
        ComplexUpset::upset_set_size(
            geom=geom_bar(
                aes(fill=biotype),
                width=0.8
            ) ,
            position='right'
        )+ scale_fill_manual(values = c("Non-protein-coding"="#BAC5CB",
                              "Protein-coding"='#B5C952')) + 
          theme(legend.position = "none")
  ),
  guides='over',
  wrap = TRUE
)
```

#### Gene identification Jaccard similarity
```{r gene identification Jaccard similarity, fig.height=3, fig.width=4.5}

# Jaccard similarity
library(ComplexHeatmap)
binary_matrix <- gene_discovery %>% 
  select( Illumina, `ONT cDNA`, PacBio ,`ONT dRNA`) %>% 
  mutate(across(everything(), as.integer))

# Compute Jaccard similarity
jaccard_sim <- function(x, y) {
  sum(x & y) / sum(x | y)
}

platforms <- colnames(binary_matrix)
jaccard_matrix <- outer(platforms, platforms, Vectorize(function(i, j) {
  jaccard_sim(binary_matrix[[i]], binary_matrix[[j]])
}))

# Convert to matrix format
colnames(jaccard_matrix) <- rownames(jaccard_matrix) <- platforms

label_func <- function(j, i, x, y, width, height, fill) {
    grid.text(sprintf("%.2f", jaccard_matrix[i, j]), x, y, gp = gpar(fontsize = 10))
}
Heatmap(jaccard_matrix, 
        name = "Jaccard Similarity", 
        col = colorRampPalette(c("grey", "#06317F"))(50),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(round(jaccard_matrix[i, j], 2), x, y, 
                    gp = gpar(col = "white"))  # Set font color to white
        },
        na_col = "white")  # Keeps upper triangle blank
```

#### Sup Upset plot Protein coding gene only
```{r GENE: Sup Upset plot Protein coding gene, fig.height=5, fig.width=8}
# Protein coding only
ComplexUpset::upset(
  gene_discovery %>% filter(biotype=="protein_coding"),
  intersect = c('Illumina', "ONT cDNA", "PacBio", "ONT dRNA"),
  annotations = list(
        # 1st method - passing list:
        "Average transcript length" = list(
          aes = aes(x = intersection, y = `Gene length`),
          # provide a list if you wish to add several geoms
          geom = geom_boxplot(na.rm = TRUE, outliers = FALSE)
        )
  ),
  set_sizes=(
        ComplexUpset::upset_set_size(
            geom=geom_bar(
                width=0.8
            ) ,
            position='right'
        )+ theme(legend.position = "none")
  ),
   base_annotations=list(
        'Intersection size'=ComplexUpset::intersection_size(
            counts=TRUE,
            text=element_text(size = 3, vjust=-0.2)
        )
    ),
  guides='over',
  wrap = TRUE
)
```

#### Discovered gene length distribution
```{r GENE: discovered gene distribution, fig.height=3, fig.width=8}
gene_discovery$length_bin <- cut(gene_discovery$`Gene length`, 
                                 breaks = c(0, 500, 1000, 1500, 2000, 2500, Inf), 
                                 labels = c("0-0.5k", "0.5-1k", "1-1.5k", "1.5-2k", "2-2.5k", "2.5k+"))

p1 <- gene_discovery %>%
  pivot_longer(cols = c("Illumina", "ONT cDNA", "PacBio", "ONT dRNA"), 
               names_to = "platform", values_to = "keep") %>%
  filter(keep) %>%
  count(length_bin, platform) %>%  # Summarize counts per category
  ggplot(aes(x = length_bin, y = n, color = platform, group = platform)) +
  geom_line(size = 1) +  # Use lines instead of bars
  geom_point(size = 2) +  # Add points for visibility
  theme_minimal() +
  labs(title = "Detected genes", 
       x = "Gene length (average transcript length)", 
       y = "# of genes detected") + 
  scale_color_manual(values = color_palette[c("Illumina", "ONT_1", "ONT_2", "PacBio")] %>% unname())

p2 <- gene_discovery %>%
  filter(biotype=="protein_coding") %>%
  pivot_longer(cols = c("Illumina", "ONT cDNA", "PacBio", "ONT dRNA"), 
               names_to = "platform", values_to = "keep") %>%
  filter(keep) %>%
  count(length_bin, platform) %>%  # Summarize counts per category
  ggplot(aes(x = length_bin, y = n, color = platform, group = platform)) +
  geom_line(size = 1) +  # Use lines instead of bars
  geom_point(size = 2) +  # Add points for visibility
  theme_minimal() +
  labs(title = "Detected protein-coding genes", 
       x = "Gene length (average transcript length)", 
       y = "# of genes detected") + 
  scale_color_manual(values = color_palette[c("Illumina", "ONT_1", "ONT_2", "PacBio")] %>% unname())

patchwork::wrap_plots(p1, p2) + 
  patchwork::plot_layout(guides = 'collect')
```

#### Gene discovery analysis (per cell line)
```{r Cell-line-Specific Upset plot, fig.height=30, fig.width=50}

cell_lines <- ill_bulk.gene.dge%>% colnames()
map(cell_lines,
    function(cell_line){
      plot_df <- gene_discovery %>% mutate(
        "Illumina" = filterByExpr(ill_bulk.gene.dge[gene_discovery$gene_id,cell_line]),
        "ONT cDNA" = filterByExpr(ont_bulk.gene.dge[gene_discovery$gene_id,cell_line]),
        "PacBio" = filterByExpr(pb_bulk.gene.dge[gene_discovery$gene_id,cell_line]),
        "ONT dRNA" = filterByExpr(drna_bulk.gene.dge[gene_discovery$gene_id,cell_line])
      )
      
      ComplexUpset::upset(
        plot_df,
        intersect = c('Illumina', "ONT cDNA", "PacBio", "ONT dRNA"),
        annotations = list(
              # 1st method - passing list:
              "Average transcript length" = list(
                aes = aes(x = intersection, y = `Gene length`),
                # provide a list if you wish to add several geoms
                geom = geom_boxplot(na.rm = TRUE, outliers = FALSE)
              ),
              "BioType" = list(
                aes = aes(x = intersection, fill = biotype),
                geom = geom_bar(position = "fill")
              )
        ),
        set_sizes=(
              ComplexUpset::upset_set_size(
                  geom=geom_bar(
                      aes(fill=biotype),
                      width=0.8
                  ) 
              )+ theme(legend.position = "none")
          ),
        wrap = TRUE
      ) + ggtitle(cell_line)
    }
) %>% patchwork::wrap_plots()

```

### Expression of the intronic gene and single-exon genes
```{r}
library(GenomicRanges)
library(rtracklayer)
library(dplyr)
library(purrr)
library(pbapply)
gtf_file <- "/vast/projects/LongBench/reference_files/GRCh38/gencode.v44.annotation.gtf"
annotations <- rtracklayer::import(gtf_file)

# Filter for gene and exon annotations
genes <- annotations[annotations$type == "gene"]
exons <- annotations[annotations$type == "exon"]

# for each gene, get the introns
exon_split <- split(exons, exons$gene_id)
library(furrr)
# Set up parallel processing with future_map (e.g., using all available cores)
plan(multisession)
options(future.globals.maxSize = 1024 * 1024 * 1024)  # 1 GB

introns  <- exon_split %>% names %>% 
    future_map(~ {
        intr <- gaps(exon_split[[.x]])
        if (length(intr) == 1) {
            return(NULL)
        } else {
            intr$gene_name <- .x
            return(intr[-1])
        }
    }, .progress = TRUE)

merged_introns <- do.call(c, introns)
overlaps <- findOverlaps(exons, merged_introns, ignore.strand=TRUE, type='within',  select="all", minoverlap =50)
overlaps_tibble <- tibble(
    exon_gene = exons$gene_id[queryHits(overlaps)],
    intron_gene = merged_introns$gene_name[subjectHits(overlaps)],
) %>%
    filter(exon_gene != intron_gene) %>%
    distinct()

# output the results
write.table(overlaps_tibble, file = "gene_introns.tsv", sep = "\t", quote = FALSE, row.names = FALSE)

overlaps_tibble$exon_gene <- sub("\\..*","", overlaps_tibble$exon_gene)
overlaps_tibble$intron_gene <- sub("\\..*","", overlaps_tibble$intron_gene)

# Sanity check:"ENSG00000271204.1" %in% overlaps_tibble$exon_gene
# 'ENSG00000231721.8' %in%  overlaps_tibble$intron_gene
```


```{r exon count per gene}
exons_count <-  exon_split %>% future_map(length,  .progress = TRUE) %>% unlist %>% enframe(name = "Gene", value = "Exon_Count")
exons_count <- exons_count %>% mutate(Gene = sub("\\..*", "", Gene))
gene_discovery<- gene_discovery %>% 
  left_join(exons_count, by = c("gene_id" = "Gene"))
```


```{r,  Expression of the intronic and single-exon gene, fig.height=6, fig.width=8}
plot_df <- gene_discovery %>% mutate(
  "Intronic Gene" = gene_discovery$gene_id %in% overlaps_tibble[overlaps_tibble$intron_gene %in% gene_discovery$gene_id,]$exon_gene,
  "Single exon" = gene_discovery$Exon_Count == 1
)

ComplexUpset::upset(
  plot_df,
  intersect = c('Illumina', "ONT cDNA", "PacBio", "ONT dRNA"),
  annotations = list(
        # 1st method - passing list:
        "Exon counts" = list(
          aes = aes(x = intersection, y = Exon_Count),
          # provide a list if you wish to add several geoms
          geom = geom_boxplot(na.rm = TRUE, outliers = FALSE)
        ),
        "Single-exon Genes" = (
          ggplot(mapping=aes(fill = `Single exon`)) +
            geom_bar(position = "fill")+
            scale_fill_manual(values = c("#BAC5CB",'#B5C952')) + 
            ylab('Single-exon gene proportion')
        ),
        "Intronic Genes" = (
          ggplot(mapping=aes(fill = `Intronic Gene`)) +
            geom_bar(position = "fill")+
            scale_fill_manual(values = c("#BAC5CB",'#B5C952')) + 
            ylab('Intronic gene proportion')
        )
  ),
  set_sizes=(
        ComplexUpset::upset_set_size(
            geom=geom_bar(
                width=0.8
            ) 
        )+ theme(legend.position = "none")
    ),
  wrap = TRUE
)

```

### Additional Analysis Not Included
#### Inspect Pacbio specific genes
Listing top expressed Platform specific identification on H1975 (randomly picked one as the trend is similar across cell lines)
```{r}
cell_line <- 'H1975'

pb_specific_gene <-  gene_discovery %>% mutate(
        "Illumina" = filterByExpr(ill_bulk.gene.dge[gene_discovery$gene_id,cell_line]),
        "ONT cDNA" = filterByExpr(ont_bulk.gene.dge[gene_discovery$gene_id,cell_line]),
        "PacBio" = filterByExpr(pb_bulk.gene.dge[gene_discovery$gene_id,cell_line]),
        "ONT dRNA" = filterByExpr(drna_bulk.gene.dge[gene_discovery$gene_id,cell_line])
      ) %>%
      filter(
        !Illumina,
        !`ONT cDNA` &
        PacBio &
        !`ONT dRNA`
      ) %>% pull(gene_id)

pb_bulk.gene.dge$counts[pb_specific_gene,] %>%
  as_tibble(rownames="gene_id") %>% 
  arrange(desc(H1975))
```

```{r, include=F, eval=F}
inspect_gene_list <- pb_bulk.gene.dge$counts[pb_specific_gene,] %>%
  as_tibble(rownames="gene_id") %>% 
  arrange(desc(H1975)) %>% pull(gene_id) %>% .[1:50]

plot_df %>% filter(
        !Illumina,
        !`ONT cDNA` &
        PacBio &
        !`ONT dRNA`
      )  %>% filter(gene_id %in% inspect_gene_list)
```


```{r, include=F, eval=F}
pb_bulk.gene.dge$counts[pb_specific_gene,] %>%
  as_tibble(rownames="gene_id") %>% 
  arrange(desc(H1975)) %>% left_join(plot_df, by="gene_id") %>% filter(`Intronic Gene`==F)

```
#### Coexpression of paired genes
```{r, include=F, eval=F}
library(dplyr)
library(ggplot2)
library(ggpubr)
library(tidyr)

# Define the specific sample of interest
specific_sample <- "H1975"

# Create a list of datasets
datasets <- list(
  ill = ill_bulk.gene.dge$counts,
  ont = ont_bulk.gene.dge$counts,
  pb = pb_bulk.gene.dge$counts,
  drna = drna_bulk.gene.dge$counts
)

# Create an empty list to store results
results <- list()

# Iterate through the datasets
for (dataset_name in names(datasets)) {
  counts_data <- datasets[[dataset_name]]
  
  # Extract expression values for each gene in overlaps_tibble for the specific sample
  overlaps_expr <- overlaps_tibble %>%
    filter(exon_gene %in% rownames(counts_data),
           intron_gene %in% rownames(counts_data)) %>%
    mutate(exon_expr = log1p(counts_data[exon_gene, specific_sample]),
           intron_expr = log1p(counts_data[intron_gene, specific_sample])) %>%
    mutate(dataset = dataset_name)  # Add the dataset column
  
  # Filter out rows where intron expression is below 5 for the specific sample
  overlaps_expr_filtered <- overlaps_expr %>%
    filter(counts_data[intron_gene, specific_sample] >= 5)
  
  # Compute Pearson correlation for the filtered data
  cor_value <- cor(overlaps_expr_filtered$exon_expr, overlaps_expr_filtered$intron_expr, method = "pearson")
  
  # Store the result in the list
  results[[dataset_name]] <- list(
    data = overlaps_expr_filtered,
    cor_value = cor_value
  )
}

# Combine all results into one data frame
combined_results <- bind_rows(lapply(results, function(x) x$data), .id = "dataset")

# Create the hex plot with regression line and correlation annotation
ggplot(combined_results, aes(x = exon_expr, y = intron_expr)) +
  geom_hex(bins = 30, aes(fill = ..count..), color = "white") +  # Use hexagonal bins with 30 bins
  scale_fill_viridis_c(option = "C") +  # Use a colorblind-friendly and distinct palette (Viridis)
  labs(x = paste("Log Exon Gene Expression (", specific_sample, ")", sep = ""),
       y = paste("Log Intron Gene Expression (", specific_sample, ")", sep = ""),
       title = paste("Correlation of Log Expression for", specific_sample)) +
  theme_minimal() +
  geom_smooth(method = "lm", color = "darkred", se = FALSE) +  # Add regression line in dark red
  facet_wrap(~dataset) +  # Facet by dataset
  # Annotate the Pearson correlation coefficient on each plot
  geom_text(data = combined_results %>%
              group_by(dataset) %>%
              summarise(cor = cor(exon_expr, intron_expr, method = "pearson"), .groups = "drop"),
            aes(x = Inf, y = Inf, label = paste("Pearson Correlation = ", round(cor, 2))),
            hjust = 1.1, vjust = 1.1, size = 4, color = "black")  # Adjust annotation position
```

```{r, include=F, eval=F}
library(dplyr)
library(ggplot2)
library(ggpubr)
library(tidyr)

# Define the specific sample of interest
specific_sample <- "H1975"

# Create a list of datasets
datasets <- list(
  ill = ill_bulk.gene.dge$counts,
  ont = ont_bulk.gene.dge$counts,
  pb = pb_bulk.gene.dge$counts,
  drna = drna_bulk.gene.dge$counts
)

# Create an empty list to store results
results <- list()

# Iterate through the datasets
for (dataset_name in names(datasets)) {
  counts_data <- datasets[[dataset_name]]
  
  # Extract expression values for each gene in overlaps_tibble for the specific sample
  overlaps_expr <- overlaps_tibble %>%
    filter(exon_gene %in% pb_specific_gene,
           intron_gene %in% rownames(counts_data)) %>%
    mutate(exon_expr = log1p(counts_data[exon_gene, specific_sample]),
           intron_expr = log1p(counts_data[intron_gene, specific_sample])) %>%
    mutate(dataset = dataset_name)  # Add the dataset column
  
  # Filter out rows where intron expression is below 5 for the specific sample
  overlaps_expr_filtered <- overlaps_expr %>%
    filter(counts_data[intron_gene, specific_sample] >= 5)
  
  # Compute Spearman correlation for the filtered data
  cor_value <- cor(overlaps_expr_filtered$exon_expr, overlaps_expr_filtered$intron_expr, method = "spearman")
  
  # Store the result in the list
  results[[dataset_name]] <- list(
    data = overlaps_expr_filtered,
    cor_value = cor_value
  )
}

# Combine all results into one data frame
combined_results <- bind_rows(lapply(results, function(x) x$data), .id = "dataset")

# Create the hex plot with regression line and correlation annotation
ggplot(combined_results, aes(x = exon_expr, y = intron_expr)) +
  geom_hex(bins = 30, aes(fill = ..count..), color = "white") +  # Use hexagonal bins with 30 bins
  scale_fill_viridis_c(option = "C") +  # Use a colorblind-friendly and distinct palette (Viridis)
  labs(x = paste("Log Exon Gene Expression (", specific_sample, ")", sep = ""),
       y = paste("Log Intron Gene Expression (", specific_sample, ")", sep = ""),
       title = paste("Correlation of Log Expression for", specific_sample)) +
  theme_minimal() +
  geom_smooth(method = "lm", color = "darkred", se = FALSE) +  # Add regression line in dark red
  facet_wrap(~dataset) +  # Facet by dataset
  # Annotate the Spearman correlation coefficient on each plot
  geom_text(data = combined_results %>%
              group_by(dataset) %>%
              summarise(cor = cor(exon_expr, intron_expr, method = "spearman"), .groups = "drop"),
            aes(x = Inf, y = Inf, label = paste("Spearman Correlation = ", round(cor, 2))),
            hjust = 1.1, vjust = 1.1, size = 4, color = "black")  # Adjust annotation position
```


# Transcript discovery analysis
```{r, fig.height=6, fig.width=10}
# Filter by expression
tx_discovery <- tibble(
  tx_id = common_tx,
  "Illumina" = filterByExpr(ill_bulk.dge, group = ill_bulk.dge$samples$group, min.count=5),
  "ONT cDNA" = filterByExpr(ont_bulk.dge.oarfish, group = ont_bulk.dge.oarfish$samples$group, min.count=5),
  "PacBio" = filterByExpr(pb_bulk.dge.oarfish, group = pb_bulk.dge.oarfish$samples$group, min.count=5),
  "ONT dRNA" = filterByExpr(ont_drna.dge.oarfish, group = ont_drna.dge.oarfish$samples$group, min.count=5)
) %>% filter(
  Illumina | `ONT cDNA` | PacBio | `ONT dRNA`
)
# split and get biotype (last element of tx_id)
tx_discovery <- tx_discovery %>%  mutate(
    biotype = strsplit(tx_id, "\\|") %>% sapply(function(x) x[8]),
    length = strsplit(tx_id, "\\|") %>% sapply(function(x) x[7]) %>% as.integer(),
    gene_id = strsplit(tx_id, "\\|") %>% sapply(function(x) x[2]),
    tx_id = strsplit(tx_id, "\\|") %>% sapply(function(x) x[1])
  )

tx_discovery <- tx_discovery %>% mutate(biotype = ifelse(biotype %in% 
            names(sort(table(biotype), decreasing = TRUE)[1:6]), 
            biotype, "Others"))
```

## Main upset plot
```{r Transcript: Main upset plot, fig.height=4, fig.width=8}
library(ComplexUpset)
ComplexUpset::upset(
  tx_discovery %>% mutate(
    biotype = case_when(
      biotype =="protein_coding" ~ "Protein-coding",
      TRUE ~ "Non-protein-coding"
    )
  ),
  intersect = c('Illumina', "ONT cDNA", "PacBio", "ONT dRNA"),
  annotations = list(
        # 1st method - passing list:
        "Transcript length" = list(
          aes = aes(x = intersection, y = `length`),
          # provide a list if you wish to add several geoms
          geom = geom_boxplot(na.rm = TRUE, outliers = FALSE)
        )
        # ,
        # "BioType" = list(
        #   aes = aes(x = intersection, fill = biotype),
        #   geom = geom_bar(position = "fill")
        # )
  ),
  base_annotations=list(
        'Intersection size'=ComplexUpset::intersection_size(
            counts=TRUE,
            mapping=aes(fill=biotype),
            text=element_text(size = 3, vjust=-0.2)
        ) + scale_fill_manual(values = c("Non-protein-coding"="#BAC5CB",
                              "Protein-coding"='#B5C952'))
    ),
  set_sizes=(
        ComplexUpset::upset_set_size(
            geom=geom_bar(
                aes(fill=biotype),
                width=0.8
            ) ,
            position='right'
        )+ scale_fill_manual(values = c("Non-protein-coding"="#BAC5CB",
                              "Protein-coding"='#B5C952')) + 
          theme(legend.position = "none")
  ),
  guides='over',
  wrap = TRUE
)
```
## Jaccard similary
```{r transcript identification Jaccard similarity, fig.height=3, fig.width=4.5}

# Jaccard similarity
library(ComplexHeatmap)
binary_matrix <- tx_discovery %>% 
  select( Illumina, `ONT cDNA`, PacBio ,`ONT dRNA`) %>% 
  mutate(across(everything(), as.integer))

# Compute Jaccard similarity
jaccard_sim <- function(x, y) {
  sum(x & y) / sum(x | y)
}

platforms <- colnames(binary_matrix)
jaccard_matrix <- outer(platforms, platforms, Vectorize(function(i, j) {
  jaccard_sim(binary_matrix[[i]], binary_matrix[[j]])
}))

# Convert to matrix format
colnames(jaccard_matrix) <- rownames(jaccard_matrix) <- platforms

label_func <- function(j, i, x, y, width, height, fill) {
    grid.text(sprintf("%.2f", jaccard_matrix[i, j]), x, y, gp = gpar(fontsize = 10))
}
Heatmap(jaccard_matrix, 
        name = "Jaccard Similarity", 
        col = colorRampPalette(c("grey", "#06317F"))(50),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(round(jaccard_matrix[i, j], 2), x, y, 
                    gp = gpar(col = "white"))  # Set font color to white
        },
        na_col = "white")  # Keeps upper triangle blank
```
## Discovered transcript length
```{r Discovered transcript length dist, fig.height=3, fig.width=8}
tx_discovery$length_bin <- cut(tx_discovery$length, breaks = c(0, 500, 1000, 1500, 2000, 2500,Inf), labels = c("0-0.5k", "0.5-1k", "1-1.5k", "1.5-2k", "2-2.5k", "2.5k+"))

p1 <- tx_discovery %>%
  pivot_longer(cols = c("Illumina", "ONT cDNA", "PacBio", "ONT dRNA"), 
               names_to = "platform", values_to = "keep") %>%
  filter(keep) %>%
  count(length_bin, platform) %>%  # Summarize counts per category
  ggplot(aes(x = length_bin, y = n, color = platform, group = platform)) +
  geom_line(size = 1) +  # Use lines instead of bars
  geom_point(size = 2) +  # Add points for visibility
  theme_minimal() +
  labs(title = "Detected Transcript", 
       x = "Transcript length", 
       y = "# of transcript detected") + 
  scale_color_manual(values = color_palette[c("Illumina", "ONT_1", "ONT_2", "PacBio")] %>% unname())

p2 <- tx_discovery %>%
  filter(biotype=="protein_coding") %>%
  pivot_longer(cols = c("Illumina", "ONT cDNA", "PacBio", "ONT dRNA"), 
               names_to = "platform", values_to = "keep") %>%
  filter(keep) %>%
  count(length_bin, platform) %>%  # Summarize counts per category
  ggplot(aes(x = length_bin, y = n, color = platform, group = platform)) +
  geom_line(size = 1) +  # Use lines instead of bars
  geom_point(size = 2) +  # Add points for visibility
  theme_minimal() +
  labs(title = "Detected protein-coding Transcript", 
       x = "Transcript length", 
       y = "# of transcript detected") + 
  scale_color_manual(values = color_palette[c("Illumina", "ONT_1", "ONT_2", "PacBio")] %>% unname())

patchwork::wrap_plots(p1, p2) + 
  patchwork::plot_layout(guides = 'collect')


```


## Additional Analysis Not Included
### Exclude Illumina
```{r, fig.height=6, fig.width=8, echo=F, include=F, eval=F}
ComplexUpset::upset(
  tx_discovery,
  intersect = c("ONT cDNA", "PacBio", "ONT dRNA"),
  annotations = list(
        # 1st method - passing list:
        "Average transcript length" = list(
          aes = aes(x = intersection, y = length),
          # provide a list if you wish to add several geoms
          geom = geom_boxplot(na.rm = TRUE, outliers = FALSE)
        ),
        "BioType" = list(
          aes = aes(x = intersection, fill = biotype),
          geom = geom_bar(position = "fill")
        )
  ),
  set_sizes=(
        ComplexUpset::upset_set_size(
            geom=geom_bar(
                aes(fill=biotype),
                width=0.8
            ) , position='right'
        )+ theme(legend.position = "none")
    ),
  wrap = TRUE, guides='over'
)

# protein coding
ComplexUpset::upset(
  tx_discovery %>% filter(grepl("protein_coding*",biotype)),
  intersect = c("ONT cDNA", "PacBio", "ONT dRNA"),
  annotations = list(
        # 1st method - passing list:
        "Average transcript length" = list(
          aes = aes(x = intersection, y = length),
          # provide a list if you wish to add several geoms
          geom = geom_boxplot(na.rm = TRUE, outliers = FALSE)
        ),
        "BioType" = list(
          aes = aes(x = intersection, fill = biotype),
          geom = geom_bar(position = "fill")
        )
  ),
  set_sizes=(
        ComplexUpset::upset_set_size(
            geom=geom_bar(
                aes(fill=biotype),
                width=0.8
            ) , position='right'
        )+ theme(legend.position = "none")
    ),
  wrap = TRUE, guides='over'
)

# protein coding
ComplexUpset::upset(
  tx_discovery %>% filter(!grepl("protein_coding*",biotype)),
  intersect = c("ONT cDNA", "PacBio", "ONT dRNA"),
  annotations = list(
        # 1st method - passing list:
        "Average transcript length" = list(
          aes = aes(x = intersection, y = length),
          # provide a list if you wish to add several geoms
          geom = geom_boxplot(na.rm = TRUE, outliers = FALSE)
        ),
        "BioType" = list(
          aes = aes(x = intersection, fill = biotype),
          geom = geom_bar(position = "fill") 
        )
  ),
  set_sizes=(
        ComplexUpset::upset_set_size(
            geom=geom_bar(
                aes(fill=biotype),
                width=0.8
            ), position='right'
        )+ theme(legend.position = "none")
    ),
  guides='over',
  wrap = TRUE
)
```
