---
title: "Spike in DE analysis"
author: "Yupei & Ash"
version: "v2.0"
output:
  html_document:
    code_folding: hide
    toc: yes
    toc_float: yes
    number_sections: yes
    toc_depth: 4
  pdf_document:
    toc: yes
    toc_depth: '4'
params:
  random_seed: 2024
  cache_dir: NULL
  sequins_tsv: '/vast/projects/LongBench/reference_files/Sequin_resources/v2.4/rnasequin_isoforms_2.4.tsv'
  sirv_csv: '/home/users/allstaff/you.yu/LongBench/reference_files/SIRV_Set4_Norm_Sequences_20210507/SIRV_set4_concentration.csv'
  bulk_meta: "/vast/projects/LongBench/sequencing_data/illumina_bulk/metadata.txt"
  input_DGEobject_rds: "/vast/projects/LongBench/analysis/workflow/rmarkdown/RDS/bulk_DGE.obj.rds"
  Tx2Gene.map.rds: "/vast/projects/LongBench/analysis/workflow/rmarkdown/RDS/Tx2Gene.map.rds"
  fig.path: NULL
---

# Setup: Create DGE from input
```{r setup, include=FALSE}
knitr::opts_chunk$set(cache = TRUE, fig.width = 7, fig.height=5, warning = F)

```

```{r Libraries,include=FALSE}
library(limma)
library(edgeR)
library(tidyverse)
library(RColorBrewer)
library(dplyr)
library(tidyr)
library(ggplot2)
library(ggrepel)
library(glue)
library(ggplot2)
library(gridExtra)
library(ggExtra)
library(knitr)
library(kableExtra)
library(rtracklayer)
library(scales)
library(UpSetR)
library(ggtree)
library(aplot)
library(patchwork)

# setup some conflict functions
rename <- dplyr::rename
select <- dplyr::select


knitr::opts_chunk$set(
  cache = !is.null(params$cache_dir),  # Enable caching only if cache_dir is provided
  fig.width = 7,                     # Set default figure width
  fig.height = 5,                    # Set default figure height
  warning = FALSE,                    # Suppress warnings in the output
  fig.path = params$fig.path,         # Set figure path (NULL if not provided)
  dev = if (!is.null(params$fig.path)) c("png", "svg") else NULL  # Set output devices if fig.path is provided
)

# Set cache.path only if params$cache_dir is not NULL
if (!is.null(params$cache_dir)) {
  knitr::opts_chunk$set(cache.path = params$cache_dir)
}
```

```{r color code, include=FALSE, eval=T}
color_palette <- c(
  PacBio = "#df1995",
  ONT = "#00789b",
  ONT_1 = "#04476c", #cDNA
  ONT_2 = "#00789b", # dRNA
  Illumina = "#e88b20"
)
```

```{r function for splitting the transcript origin}
is.sequins <- function(dge) {grepl("^R", rownames(dge))}
is.human <- function(dge) {grepl("^ENS", rownames(dge))}
is.sirv_or_ercc <- function(dge) {!is.human(dge) & !is.sequins(dge)} # thie is the whole SIRV set4 (ERCC + SIRV E0 + Long SIRV)
is.sirv_E0 <- function(dge) {grepl("^SIRV\\d{3}$", rownames(dge))} 
is.sirv_all <- function(dge) {grepl("^SIRV", rownames(dge))} # including all SIRV E0 and Long SIRV
is.long_sirv <- function(dge) {grepl("^SIRV\\d{4}", rownames(dge))}
is.ercc <- function(dge) {!is.human(dge) & !is.sequins(dge) & !is.sirv_all(dge)}
```

# Load the count 
## Get the count data from RDS

```{r Get the count data, echo=TRUE, message=FALSE, warning=FALSE, results='hide'}
list2env(readRDS(params$input_DGEobject_rds), .GlobalEnv)
# This will load the following DGE objects:
# ill_bulk.tx.dge, ont_bulk.tx.dge, pb_bulk.tx.dge, drna_bulk.tx.dge, for transcript level unfiltered DGE
# ill_bulk.gene.dge, ont_bulk.gene.dge, pb_bulk.gene.dge, drna_bulk.gene.dge for gene level unfiltered DGE

list2env(readRDS(params$Tx2Gene.map.rds), .GlobalEnv)
# This will load 
# "SIRV.G.Tx.map"  "sequins.G.Tx.map" "human.G.Tx.map"  
```



# Separate spike in types
```{r}
# Get seqins and sirv transcript dge
ill_bulk.txi.dge_sequins <- ill_bulk.tx.dge[is.sequins(ill_bulk.tx.dge), ]
ont_bulk.txi.dge_sequins <- ont_bulk.tx.dge[is.sequins(ont_bulk.tx.dge), ]
pb_bulk.txi.dge_sequins <- pb_bulk.tx.dge[is.sequins(pb_bulk.tx.dge), ]
drna_bulk.txi.dge_sequins <- drna_bulk.tx.dge[is.sequins(drna_bulk.tx.dge), ]

## get sirv transcript dge
ill_bulk.txi.dge_sirv <- ill_bulk.tx.dge[is.sirv_all(ill_bulk.tx.dge), ]
ont_bulk.txi.dge_sirv <- ont_bulk.tx.dge[is.sirv_all(ont_bulk.tx.dge), ]
pb_bulk.txi.dge_sirv <- pb_bulk.tx.dge[is.sirv_all(pb_bulk.tx.dge), ]
drna_bulk.txi.dge_sirv <- drna_bulk.tx.dge[is.sirv_all(drna_bulk.tx.dge), ]

## get ercc transcript dge
ill_bulk.txi.dge_ercc <- ill_bulk.tx.dge[is.ercc(ill_bulk.tx.dge), ]
ont_bulk.txi.dge_ercc <- ont_bulk.tx.dge[is.ercc(ont_bulk.tx.dge), ]
pb_bulk.txi.dge_ercc <- pb_bulk.tx.dge[is.ercc(pb_bulk.tx.dge), ]
drna_bulk.txi.dge_ercc <- drna_bulk.tx.dge[is.ercc(drna_bulk.tx.dge), ]
```

# Gene Transcript discovery analysis
Using `filterByExpr` for each dataset, using min.count = 10 for gene and min.count = 5 for transcript.
```{r Gene Transcript discovery upset plot}
# Transcript
get_discovery_upset <- function(dge_lst, min.count = 10) {
  tx_discovery <- tibble(
    "Illumina" = filterByExpr(dge_lst[["Illumina"]], group = dge_lst[["Illumina"]]$samples$sample, min.count = min.count),
    "ONT cDNA" = filterByExpr(dge_lst[["ONT cDNA"]], group = dge_lst[["ONT cDNA"]]$samples$sample, min.count = min.count),
    "PacBio" = filterByExpr(dge_lst[["PacBio"]], group = dge_lst[["PacBio"]]$samples$sample, min.count = min.count),
    "ONT dRNA" = filterByExpr(dge_lst[["ONT dRNA"]], group = dge_lst[["ONT dRNA"]]$samples$sample, min.count = min.count),
    Length = dge_lst[["Illumina"]]$genes$Length # use illumina here but they are the same
  )

ComplexUpset::upset(
  tx_discovery,
  intersect = c('Illumina', "ONT cDNA", "PacBio", "ONT dRNA"),
  annotations = list(
        # 1st method - passing list:
        "Gene length" = list(
          aes = aes(x = intersection, y = Length),
          # provide a list if you wish to add several geoms
          geom = geom_boxplot(na.rm = TRUE, outliers = FALSE)
        )
  ),
  wrap = TRUE
)}


# Transcript
list(
  "Illumina" = ill_bulk.txi.dge_sequins,
  "ONT cDNA" = ont_bulk.txi.dge_sequins,
  "PacBio" = pb_bulk.txi.dge_sequins,
  "ONT dRNA" = drna_bulk.txi.dge_sequins
) %>% get_discovery_upset(min.count = 5) + ggtitle("Sequins transcript discovery")

list(
  "Illumina" = ill_bulk.txi.dge_sirv,
  "ONT cDNA" = ont_bulk.txi.dge_sirv,
  "PacBio" = pb_bulk.txi.dge_sirv,
  "ONT dRNA" = drna_bulk.txi.dge_sirv
) %>% get_discovery_upset(min.count = 5) + ggtitle("SIRV transcript discovery")

## ERCC with true abundance
dge_lst <- list(
  "Illumina" = ill_bulk.txi.dge_ercc,
  "ONT cDNA" = ont_bulk.txi.dge_ercc,
  "PacBio" = pb_bulk.txi.dge_ercc,
  "ONT dRNA" = drna_bulk.txi.dge_ercc
)
min.count = 5
tx_discovery <- tibble(
    "Illumina" = filterByExpr(dge_lst[["Illumina"]], group = dge_lst[["Illumina"]]$samples$sample, min.count = min.count),
    "ONT cDNA" = filterByExpr(dge_lst[["ONT cDNA"]], group = dge_lst[["ONT cDNA"]]$samples$sample, min.count = min.count),
    "PacBio" = filterByExpr(dge_lst[["PacBio"]], group = dge_lst[["PacBio"]]$samples$sample, min.count = min.count),
    "ONT dRNA" = filterByExpr(dge_lst[["ONT dRNA"]], group = dge_lst[["ONT dRNA"]]$samples$sample, min.count = min.count),
    Length = dge_lst[["Illumina"]]$genes$Length, # use illumina here but they are the same,
    Name = rownames(dge_lst[["Illumina"]])
  ) %>% left_join(read_csv(params$sirv_csv) %>% filter(grepl("ERCC", txid)), by = c("Name" = "GenBank"))

  ComplexUpset::upset(
    tx_discovery,
    intersect = c("Illumina", "ONT cDNA", "PacBio", "ONT dRNA"),
    annotations = list(
      # 1st method - passing list:
      "Gene length" = list(
        aes = aes(x = intersection, y = Length),
        # provide a list if you wish to add several geoms
        geom = geom_boxplot(na.rm = TRUE, outliers = FALSE)
      ),
      "True abundance (log10 amol/ul)" = list(
        aes = aes(x = intersection, y = log10(conc)),
        geom = geom_boxplot(na.rm = TRUE, outliers = FALSE)
      )
    ),
    wrap = TRUE
  ) + ggtitle("ERCC transcript discovery")
```

# Differential Transcript expression analysis

```{r filter for common gene and transcript only}
# Transcript
# Sequin
common_tx <- Reduce(intersect, lapply(list(ill_bulk.txi.dge_sequins, ont_bulk.txi.dge_sequins, pb_bulk.txi.dge_sequins, drna_bulk.txi.dge_sequins), rownames))
ill_bulk.txi.dge_sequins <- ill_bulk.txi.dge_sequins[common_tx, ]
ont_bulk.txi.dge_sequins <- ont_bulk.txi.dge_sequins[common_tx, ]
pb_bulk.txi.dge_sequins <- pb_bulk.txi.dge_sequins[common_tx, ]
drna_bulk.txi.dge_sequins <- drna_bulk.txi.dge_sequins[common_tx, ]
keep <- filterByExpr(ill_bulk.txi.dge_sequins, group = ill_bulk.txi.dge_sequins$samples$group, min.count=5) &
          filterByExpr(ont_bulk.txi.dge_sequins, group = ont_bulk.txi.dge_sequins$samples$group, min.count=5) &
          filterByExpr(pb_bulk.txi.dge_sequins, group = pb_bulk.txi.dge_sequins$samples$group, min.count=5) &
          filterByExpr(drna_bulk.txi.dge_sequins, group = drna_bulk.txi.dge_sequins$samples$group, min.count=5)
ill_bulk.txi.dge_sequins <- ill_bulk.txi.dge_sequins[keep, ,keep.lib.sizes=FALSE]
ont_bulk.txi.dge_sequins <- ont_bulk.txi.dge_sequins[keep, ,keep.lib.sizes=FALSE]
pb_bulk.txi.dge_sequins <- pb_bulk.txi.dge_sequins[keep, ,keep.lib.sizes=FALSE]
drna_bulk.txi.dge_sequins <- drna_bulk.txi.dge_sequins[keep, ,keep.lib.sizes=FALSE]


# Sirv
common_tx <- Reduce(intersect, lapply(list(ill_bulk.txi.dge_sirv, ont_bulk.txi.dge_sirv, pb_bulk.txi.dge_sirv, drna_bulk.txi.dge_sirv), rownames))
ill_bulk.txi.dge_sirv <- ill_bulk.txi.dge_sirv[common_tx, ]
ont_bulk.txi.dge_sirv <- ont_bulk.txi.dge_sirv[common_tx, ]
pb_bulk.txi.dge_sirv <- pb_bulk.txi.dge_sirv[common_tx, ]
drna_bulk.txi.dge_sirv <- drna_bulk.txi.dge_sirv[common_tx, ]

keep <- filterByExpr(ill_bulk.txi.dge_sirv, group = ill_bulk.txi.dge_sirv$samples$group, min.count=5) &
          filterByExpr(ont_bulk.txi.dge_sirv, group = ont_bulk.txi.dge_sirv$samples$group, min.count=5) &
          filterByExpr(pb_bulk.txi.dge_sirv, group = pb_bulk.txi.dge_sirv$samples$group, min.count=5) &
          filterByExpr(drna_bulk.txi.dge_sirv, group = drna_bulk.txi.dge_sirv$samples$group, min.count=5)
ill_bulk.txi.dge_sirv <- ill_bulk.txi.dge_sirv[keep, ,keep.lib.sizes=FALSE]
ont_bulk.txi.dge_sirv <- ont_bulk.txi.dge_sirv[keep, ,keep.lib.sizes=FALSE]
pb_bulk.txi.dge_sirv <- pb_bulk.txi.dge_sirv[keep, ,keep.lib.sizes=FALSE]
drna_bulk.txi.dge_sirv <- drna_bulk.txi.dge_sirv[keep, ,keep.lib.sizes=FALSE]



# ERCC
common_tx <- Reduce(intersect, lapply(list(ill_bulk.txi.dge_ercc, ont_bulk.txi.dge_ercc, pb_bulk.txi.dge_ercc, drna_bulk.txi.dge_ercc), rownames))
ill_bulk.txi.dge_ercc <- ill_bulk.txi.dge_ercc[common_tx, ]
ont_bulk.txi.dge_ercc <- ont_bulk.txi.dge_ercc[common_tx, ]
pb_bulk.txi.dge_ercc <- pb_bulk.txi.dge_ercc[common_tx, ]
drna_bulk.txi.dge_ercc <- drna_bulk.txi.dge_ercc[common_tx, ]

keep <- filterByExpr(ill_bulk.txi.dge_ercc, group = ill_bulk.txi.dge_ercc$samples$group, min.count=5) &
          filterByExpr(ont_bulk.txi.dge_ercc, group = ont_bulk.txi.dge_ercc$samples$group, min.count=5) &
          filterByExpr(pb_bulk.txi.dge_ercc, group = pb_bulk.txi.dge_ercc$samples$group, min.count=5) &
          filterByExpr(drna_bulk.txi.dge_ercc, group = drna_bulk.txi.dge_ercc$samples$group, min.count=5)
ill_bulk.txi.dge_ercc <- ill_bulk.txi.dge_ercc[keep, ,keep.lib.sizes=FALSE]
ont_bulk.txi.dge_ercc <- ont_bulk.txi.dge_ercc[keep, ,keep.lib.sizes=FALSE]
pb_bulk.txi.dge_ercc <- pb_bulk.txi.dge_ercc[keep, ,keep.lib.sizes=FALSE]
drna_bulk.txi.dge_ercc <- drna_bulk.txi.dge_ercc[keep, ,keep.lib.sizes=FALSE]

```


## Transcript level
### Run DTE edgeR+scaling (Salmon and Oarfish) 
```{r}
DTE.edgeR_sequins <- function(dge, norm.method = "TMM") {
  
  if (!is.null(norm.method)) {
    dge <- calcNormFactors(dge, method = norm.method)
  }
  
  design <- model.matrix(~0 + sequins, data = dge$samples)
  dge <- estimateDisp(dge, design)
  qlfit <- glmQLFit(dge, design)
  
  contrasts <- makeContrasts(
    AvsB = sequinsA - sequinsB,
    levels = design
  )
  
  AvsB.QLF.test <- glmQLFTest(qlfit, contrast = contrasts[, "AvsB"])
  
  list(
    AvsB.DE = AvsB.QLF.test
  )
}
```

```{r get dte scaling}
ill_bulk.dte_sequins.list <- DTE.edgeR_sequins(ill_bulk.txi.dge_sequins)
ont_bulk.dte_sequins.list <- DTE.edgeR_sequins(ont_bulk.txi.dge_sequins)
drna_bulk.dte_sequins.list <- DTE.edgeR_sequins(drna_bulk.txi.dge_sequins)
pb_bulk.dte_sequins.list <- DTE.edgeR_sequins(pb_bulk.txi.dge_sequins)


ill_bulk.dte_sirv.list <- DTE.edgeR_sequins(rbind(ill_bulk.txi.dge_sirv, ill_bulk.txi.dge_ercc))
ont_bulk.dte_sirv.list <- DTE.edgeR_sequins(rbind(ont_bulk.txi.dge_sirv, ont_bulk.txi.dge_ercc))
drna_bulk.dte_sirv.list <- DTE.edgeR_sequins(rbind(drna_bulk.txi.dge_sirv, drna_bulk.txi.dge_ercc))
pb_bulk.dte_sirv.list <- DTE.edgeR_sequins(rbind(pb_bulk.txi.dge_sirv, pb_bulk.txi.dge_ercc))
```

### Upset plot
### DTE Upset plot (sequins transcript)
```{r upset plot dte sequins edgeR scaling, fig.width=8, fig.height=6}
# Upset plot
## A vs B (AvsB)
lst <- list(
  "Illumina" = ill_bulk.dte_sequins.list$AvsB.DE,
  "ONT cDNA" = ont_bulk.dte_sequins.list$AvsB.DE,
  "PacBio" = pb_bulk.dte_sequins.list$AvsB.DE,
  "ONT dRNA" = drna_bulk.dte_sequins.list$AvsB.DE
)
complex_upset_plot(lst)$plot + ggtitle("DTE Sequins MIX_A vs MIX_B (with scaling)") + theme(plot.title = element_text(size = 20))
```

# Spike in fold change analysis
```{r}
compare_sequins_txi_logFC <- function(glm.test.rst, fig.title){
  sequins.gt <- read_tsv(params$sequins_tsv)
  sequins.gt$gt.logFC <- log2(sequins.gt$MIX_A/sequins.gt$MIX_B)

  # glm.test result
  df <- glm.test.rst %>% 
    topTags(adjust.method = 'BH', n = nrow(glm.test.rst))  %>% 
    data.frame() %>% tibble::rownames_to_column(var="Transcript") %>% 
    select(Transcript, logFC, FDR) %>% 
    left_join(sequins.gt %>% select(NAME, gt.logFC, LENGTH), by = c("Transcript"="NAME"))

  # add a column for significant DE, "down" for downregulated, "up" for upregulated
  df$DE <- ifelse(df$logFC > 0 & df$FDR < 0.05, "up", ifelse(df$logFC < 0 & df$FDR < 0.05, "down", "not significant"))

  # plot gt.logFC vs logFC, add correlation
  p <- ggplot(df, aes(x = logFC, y = gt.logFC, color = DE, size = LENGTH, label=Transcript)) +
    geom_point() +
    # geom_text_repel(size = 3)    +
    scale_size(limits = c(min(sequins.gt$LENGTH), max(sequins.gt$LENGTH))) + 
    geom_abline(intercept = 0, slope = 1, linetype = "dashed") +
    scale_color_viridis_d() +
    theme_minimal() +
    labs(x = "Estimated logFC", y = " Ground truth logFC", color = "DE") +
    annotate("text",
      x = min(df$logFC), y = max(df$gt.logFC),
      label = paste(
        "\nMAE: ", round(mean(abs(df$logFC - df$gt.logFC)), 2),
        "\nMSE: ", round(mean((df$logFC - df$gt.logFC)^2), 2)
      ),
      hjust = 0, vjust = 1,
      color = "black", size = 4,
      fontface = "italic"
    ) +
    ggtitle(fig.title) 
  return(p)
}



```


```{r spike in logFC scater edgeR + scaling, fig.width=8, fig.height=6}
# Sequins
# open a pdf
lst <- list(
  "Illumina" = ill_bulk.dte_sequins.list$AvsB.DE,
  "ONT cDNA" = ont_bulk.dte_sequins.list$AvsB.DE,
  "PacBio" = pb_bulk.dte_sequins.list$AvsB.DE,
  "ONT dRNA" = drna_bulk.dte_sequins.list$AvsB.DE
)
lfc_plot <- lapply(1:4, function(x) compare_sequins_txi_logFC(lst[[x]], fig.title = names(lst)[x]))
(lfc_plot[[1]] | lfc_plot[[2]]) / (lfc_plot[[3]] | lfc_plot[[4]]) + patchwork::plot_annotation(
  title = "Sequins MixA vs MixB",
  theme = theme(plot.title = element_text(size =25))
  )  + patchwork::plot_layout(
    guides = "collect"
)
```


```{r spike in logFC scater edgeR + scaling SIRV, fig.width=8, fig.height=6}
# SIRV (including ERCC)

compare_sirv_txi_logFC <- function(glm.test.rst.list){
  df <- do.call(rbind, lapply(
    names(glm.test.rst.list),
    function(x){
      glm.test.rst.list[[x]] %>% 
        topTags(adjust.method = 'BH', n = nrow(glm.test.rst.list[[x]])) %>% 
        data.frame() %>% 
        tibble::rownames_to_column(var="NAME") %>% 
        select(NAME, logFC, FDR) %>% 
        mutate(gt.logFC = 0, Platform = x)
    }
  ))
  
  df$DE <- ifelse(df$logFC > 0 & df$FDR < 0.05, "up", 
                  ifelse(df$logFC < 0 & df$FDR < 0.05, "down", "not_sig"))

  # Calculate MAE and MSE per platform
  metrics <- df %>%
    group_by(Platform) %>%
    summarise(
      MAE = mean(abs(logFC)),
      MSE = mean(logFC^2)
    )

  # Generate plot
  p <- ggplot(df, aes(x = logFC, fill = DE, y = Platform)) +
    geom_dotplot(binwidth = (max(df$logFC) - min(df$logFC))/100, stackdir = "up", color = NA, dotsize = 1.5) + 
    scale_fill_viridis_d() +
    theme_minimal() +
    geom_vline(xintercept = 0, linetype = "dashed") +
    labs(x = "Estimated logFC", fill = "DE")

  # Add per-platform MAE and MSE as annotations
  y_positions <- seq_along(unique(df$Platform))
  for (i in seq_len(nrow(metrics))) {
    p <- p + annotate("text",
                      x = min(df$logFC),
                      y = metrics$Platform[i],
                      vjust = -4,
                      hjust = -1,
                      label = paste0("MAE=", round(metrics$MAE[i], 2),
                                     " | MSE=", round(metrics$MSE[i], 2)),
                      size = 3.5,
                      fontface = "italic")
  }

  return(p)
}

list(
  "Illumina" = ill_bulk.dte_sirv.list$AvsB.DE,
  "ONT cDNA" = ont_bulk.dte_sirv.list$AvsB.DE,
  "PacBio" = pb_bulk.dte_sirv.list$AvsB.DE,
  "ONT dRNA" = drna_bulk.dte_sirv.list$AvsB.DE
) %>% compare_sirv_txi_logFC


# save Rdata
# save.image("~/vast/DE_with_salmon_oarfish.RData")
# load("~/vast/DE_with_salmon_oarfish.RData")
```