---

title: "Bulk gene transcript identification analysis"
author: "Yupei & Ash"
date: '`r Sys.Date()`'
version: "v2.0"
output:
  html_document:
    code_folding: hide
    toc: yes
    toc_float: yes
    number_sections: yes
    toc_depth: 4
  pdf_document:
    toc: yes
    toc_depth: '4'
params:
  random_seed: 2024
  input_DGEobject_rds: "/vast/projects/LongBench/analysis/workflow/rmarkdown/RDS/bulk_DGE.obj.rds"
  bulk_meta: '/vast/projects/LongBench/sequencing_data/illumina_bulk/metadata.txt'
  cache_dir: NULL
  fig.path: 'figures/'
  out.rds: '/vast/projects/LongBench/analysis/workflow/rmarkdown/RDS/bulk_identification.rds'
---

# Setup
```{r Input filename setup, include=FALSE, eval=F}
# This chunk is used only when running the file chunk by chunk manually
params <- list(
  random_seed = 2024,
  cache_dir = NULL,
  bulk_meta = "/vast/projects/LongBench/sequencing_data/illumina_bulk/metadata.txt",
  out.rds= '/vast/projects/LongBench/analysis/workflow/rmarkdown/RDS/bulk_identification.rds'
)
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  cache = TRUE,
  fig.width = 15, 
  fig.height = 10, 
  warning = F, 
  fig.path = params$fig.path, 
  dev = c("png", "svg"),
  cache.path = ifelse(is.null(params$cache_dir), 
                      paste0(tools::file_path_sans_ext(knitr::current_input()), "_cache/"), 
                      params$cache_dir))
```

```{r Libraries,include=FALSE}
library(limma)
library(edgeR)
library(tidyverse)
library(RColorBrewer)
library(dplyr)
library(tidyr)
library(ggplot2)
library(ggrepel)
library(glue)
library(ggplot2)
library(gridExtra)
library(ggExtra)
library(knitr)
library(kableExtra)
library(rtracklayer)
library(scales)
library(UpSetR)
library(ggtree)
library(aplot)
library(patchwork)
source("/home/users/allstaff/you.yu/LongBench/analysis/workflow/scripts/Rfunctions.R")

# setup some conflict functions
rename <- dplyr::rename
select <- dplyr::select

# Set color palette
color_palette <- c(
  PacBio = "#df1995",
  ONT = "#00789b",
  ONT_1 = "#04476c",
  ONT_2 = "#00789b",
  Illumina = "#e88b20"
)


output_list <- list() # used to store result for downstream analysis
```

```{r function for splitting the transcript origin}
is.human <- function(dge) {grepl("^ENS", rownames(dge))}
```


# Get the count data from RDS

```{r Get the count data, echo=TRUE, message=FALSE, warning=FALSE, results='hide'}
list2env(readRDS(params$input_DGEobject_rds), .GlobalEnv)
# This will load the following DGE objects:
# ill_bulk.tx.dge, ont_bulk.tx.dge, pb_bulk.tx.dge, drna_bulk.tx.dge, for transcript level unfiltered DGE
# ill_bulk.gene.dge, ont_bulk.gene.dge, pb_bulk.gene.dge, drna_bulk.gene.dge for gene level unfiltered DGE
```

# Gene discovery analysis
## Get `gene_discovery` 

Based on `filterByExpr` for each dataset, using default parameters
```{r gene filtering for DE}
# Filter for human Gene 
ill_bulk.gene.dge <- ill_bulk.gene.dge[is.human(ill_bulk.gene.dge),]
ont_bulk.gene.dge <- ont_bulk.gene.dge[is.human(ont_bulk.gene.dge),]
pb_bulk.gene.dge <- pb_bulk.gene.dge[is.human(pb_bulk.gene.dge),]
drna_bulk.gene.dge <- drna_bulk.gene.dge[is.human(drna_bulk.gene.dge),]

# Get union (superset) of genes
# Salmon and oarfish didn't output exactly some number of feature in the count matrix
all_genes <- Reduce(union, lapply(
  list(ill_bulk.gene.dge, ont_bulk.gene.dge, pb_bulk.gene.dge, drna_bulk.gene.dge),
  rownames
))

# Pad each DGE with missing genes (set to 0)
pad_dge <- function(dge, all_genes) {
  missing <- setdiff(all_genes, rownames(dge))
  if (length(missing) > 0) {
    pad <- matrix(0, nrow = length(missing), ncol = ncol(dge$counts),
                  dimnames = list(missing, colnames(dge$counts)))
    dge$counts <- rbind(dge$counts, pad)
  }
  dge[all_genes, ]
}

ill_bulk.gene.dge <- pad_dge(ill_bulk.gene.dge, all_genes)
ont_bulk.gene.dge <- pad_dge(ont_bulk.gene.dge, all_genes)
pb_bulk.gene.dge <- pad_dge(pb_bulk.gene.dge, all_genes)
drna_bulk.gene.dge <- pad_dge(drna_bulk.gene.dge, all_genes)


# Remove GENCODE version number
rownames(ill_bulk.gene.dge) <- ill_bulk.gene.dge %>% rownames %>% sub("\\..*", "", .)
rownames(ont_bulk.gene.dge) <- ont_bulk.gene.dge %>% rownames %>% sub("\\..*", "", .)
rownames(pb_bulk.gene.dge) <- pb_bulk.gene.dge %>% rownames %>% sub("\\..*", "", .)
rownames(drna_bulk.gene.dge) <- drna_bulk.gene.dge %>% rownames %>% sub("\\..*", "", .)

```

### Gene detection
```{r Define Gene discovery, fig.height=6, fig.width=10}
# Filter by expression
gene_discovery <- tibble(
  gene_id = all_genes,
  "Illumina" = filterByExpr(ill_bulk.gene.dge, group = ill_bulk.gene.dge$samples$group),
  "ONT cDNA" = filterByExpr(ont_bulk.gene.dge, group = ont_bulk.gene.dge$samples$group),
  "PacBio" = filterByExpr(pb_bulk.gene.dge, group = pb_bulk.gene.dge$samples$group),
  "ONT dRNA" = filterByExpr(drna_bulk.gene.dge, group = drna_bulk.gene.dge$samples$group)
) %>% filter(
  Illumina | `ONT cDNA` | PacBio | `ONT dRNA`
)

gene_discovery$gene_id <- gene_discovery$gene_id %>% sub("\\..*", "", .)
```


### Add biotype and gene length
Let's adding some more feature for plotting
```{r, fig.height=6, fig.width=10}
# Getting the gene biotype and gene id, using illumina data here but will be identical for all datasets
tx_ref_df <- ill_bulk.tx.dge$genes %>%
  rownames_to_column("tx_id") %>%
  select(tx_id, Length) %>%
  tibble() %>%
  mutate(
    biotype = strsplit(tx_id, "\\|") %>% sapply(function(x) x[8]),
    gene_id = strsplit(tx_id, "\\|") %>% sapply(function(x) x[2]),
    tx_id = strsplit(tx_id, "\\|") %>% sapply(function(x) x[1])
  )

gene_length_lookup <- tx_ref_df %>%
    group_by(gene_id) %>%
    summarise(mean_length = mean(Length, na.rm = TRUE), .groups = "drop") %>%
    mutate(gene_id= sub("\\..*", "", gene_id))

# Connect to Ensembl
library(biomaRt)
ensembl <- useEnsembl(biomart = "genes", dataset = "hsapiens_gene_ensembl")


gene_discovery <- gene_discovery %>%
    left_join(gene_length_lookup, by = c("gene_id")) %>%
    rename(`Gene length` = mean_length)

biotypes <- getBM(
  attributes = c("ensembl_gene_id", "gene_biotype"),
  filters = "ensembl_gene_id",
  values = gene_discovery$gene_id,
  mart = ensembl
)


gene_discovery <- gene_discovery %>%
  left_join(biotypes, by = c("gene_id" = "ensembl_gene_id")) %>%
  rename(biotype = gene_biotype) %>%  
  mutate(biotype = ifelse(biotype %in% 
            names(sort(table(biotype), decreasing = TRUE)[1:6]), 
            biotype, "Others"))


gene_discovery$biotype <- factor(gene_discovery$biotype, levels = c(sort(unique(gene_discovery$biotype)) %>% setdiff("Others"), "Others"))
```

### Add intronic Gene and exon count
Boolean for intronic genes. Each gene is checked if it is located inside the intron of the other gene. The specific definition: `findOverlaps(exons, merged_introns, ignore.strand=TRUE, type='within',  select="all", minoverlap =50)`
```{r Add intronic gene and exon count}

rds_list <- readRDS("/home/users/allstaff/you.yu/LongBench/analysis/workflow/rmarkdown/RDS/intronic_gene_and_exon_count.rds")
overlaps_tibble <- rds_list$exon_intron_overlaps
overlaps_tibble$exon_gene <- sub("\\..*","", overlaps_tibble$exon_gene)
overlaps_tibble$intron_gene <- sub("\\..*", "", overlaps_tibble$intron_gene)

exons_count <- rds_list$exons_count  %>% mutate(Gene = sub("\\..*", "", Gene))

rm(rds_list)

# Merge the intronic gene information into the gene_discovery tibble
gene_discovery <- gene_discovery %>% mutate(
  "Intronic Gene" = gene_discovery$gene_id %in% overlaps_tibble[overlaps_tibble$intron_gene %in% gene_discovery$gene_id,]$exon_gene
)  %>% left_join(exons_count, by = c("gene_id" = "Gene"))
```


### Save to `output_list`
```{r}
output_list[["gene_discovery"]] <- gene_discovery # save for downstream analysis
```

## Gene detection plots
###  Main Upset plot
```{r GENE: Main Upset plot, fig.height=4, fig.width=8}
library(ComplexUpset)
ComplexUpset::upset(
  gene_discovery %>% mutate(
    biotype = case_when(
      biotype =="protein_coding" ~ "Protein-coding",
      TRUE ~ "Non-protein-coding"
    )
  ),
  intersect = c('Illumina', "ONT cDNA", "PacBio", "ONT dRNA"),
  annotations = list(
        # 1st method - passing list:
        "Gene length" = list(
          aes = aes(x = intersection, y = `Gene length`),
          # provide a list if you wish to add several geoms
          geom = geom_boxplot(na.rm = TRUE, outliers = FALSE)
        )
        # ,
        # "BioType" = list(
        #   aes = aes(x = intersection, fill = biotype),
        #   geom = geom_bar(position = "fill")
        # )
  ),
  base_annotations=list(
        'Intersection size'=ComplexUpset::intersection_size(
            counts=TRUE,
            mapping=aes(fill=biotype),
            text=element_text(size = 3, vjust=-0.2)
        ) + scale_fill_manual(values = c("Non-protein-coding"="#BAC5CB",
                              "Protein-coding"='#B5C952'))
    ),
  set_sizes=(
        ComplexUpset::upset_set_size(
            geom=geom_bar(
                aes(fill=biotype),
                width=0.8
            ) ,
            position='right'
        )+ scale_fill_manual(values = c("Non-protein-coding"="#BAC5CB",
                              "Protein-coding"='#B5C952')) + 
          theme(legend.position = "none")
  ),
  guides='over',
  wrap = TRUE
)
```


```{r GENE: Agreement plot, fig.height=3, fig.width=8}
agreement_plot <- function(df){
  df_long <- df %>% select(gene_id, Illumina, `ONT cDNA`, PacBio, `ONT dRNA`) %>%
  pivot_longer(-gene_id, names_to = "Platform", values_to = "Detected") %>%
  group_by(gene_id) %>%
  mutate(TotalDetected = sum(Detected)) %>%
  ungroup() %>%
  mutate(
    Category = case_when(
      Detected == 1 & TotalDetected == 1 ~ "Platform-Specific Detection",  # Only detected by this platform
      Detected == 0 & TotalDetected == 3 ~ "Platform-Specific Absence",      # Not detected by this platform but present elsewhere
      Detected == 1 & TotalDetected > 1 ~ "Common Detection (also detected in >0 other platform)",
      Detected == 0 & TotalDetected < 3 ~ "Common Absence (also absent in >0 other platform)" # Detected by this platform + at least one other
    )
  ) %>%
  count(Platform, Category)

# Convert to factor for proper ordering in plot
df_long$Category <- factor(df_long$Category, 
                           levels = c(
                                      "Platform-Specific Absence", 
                                      "Common Absence (also absent in >0 other platform)",
                                      "Platform-Specific Detection", 
                                      "Common Detection (also detected in >0 other platform)"
                                      ))

# Stacked bar plot
df_long %>%
  mutate(n = case_when(
    Category %in% c("Platform-Specific Absence", "Common Absence (also absent in >0 other platform)") ~ -n,
    TRUE ~ n
  )) %>%
  ggplot(aes(x = Platform, y = n, fill = Category)) +
  geom_bar(stat = "identity") +
  labs(
    y = "Number of Genes",
    x = "Platform",
    fill = "Category"
  ) +
  theme_minimal() +
  scale_fill_manual(
    values = c(
      "Platform-Specific Detection" = "#0072B2",  # Blue (color-blind friendly)
      "Platform-Specific Absence" = "#E69F00",    # Orange (color-blind friendly)
      "Common Detection (also detected in >0 other platform)" = "#009E73",  # Green
      "Common Absence (also absent in >0 other platform)" = "#999999"  # Gray
    )
  ) +
  coord_flip() +  # Flip axes for horizontal bars
  theme(
    axis.text.x = element_text(angle = 0, hjust = 0.5, vjust = 0.5),  # Ensure x-axis labels are readable
    axis.title.x = element_text(size = 12),  # Adjust x-axis title size
    axis.text.y = element_text(size = 10)   # Adjust y-axis text size if needed
  )
}

gene_discovery %>% agreement_plot
```
### Sup Upset plot Protein coding gene only
```{r GENE: Sup Upset plot Protein coding gene, fig.height=5, fig.width=8}
# Protein coding only
ComplexUpset::upset(
  gene_discovery %>% filter(biotype=="protein_coding"),
  intersect = c('Illumina', "ONT cDNA", "PacBio", "ONT dRNA"),
  annotations = list(
        # 1st method - passing list:
        "Average transcript length" = list(
          aes = aes(x = intersection, y = `Gene length`),
          # provide a list if you wish to add several geoms
          geom = geom_boxplot(na.rm = TRUE, outliers = FALSE)
        )
  ),
  set_sizes=(
        ComplexUpset::upset_set_size(
            geom=geom_bar(
                width=0.8
            ) ,
            position='right'
        ) + theme(legend.position = "none")
  ),
   base_annotations=list(
        'Intersection size'=ComplexUpset::intersection_size(
            counts=TRUE,
            text=element_text(size = 3, vjust=-0.2)
        )
    ),
  guides='over',
  wrap = TRUE
)
```


```{r GENE: Platform specific detection, fig.height=3, fig.width=8}
(
  (agreement_plot(gene_discovery %>% filter(biotype=="protein_coding")) + ggtitle("Protein-coding")) |
  (agreement_plot(gene_discovery %>% filter(biotype!="protein_coding")) + ggtitle("Non-Protein-coding")) 
) + 
patchwork::plot_layout(guides = "collect")
```


### Gene identification Jaccard similarity
```{r gene identification Jaccard similarity, fig.height=3, fig.width=4.5}

# Jaccard similarity
library(ComplexHeatmap)
binary_matrix <- gene_discovery %>% 
  select( Illumina, `ONT cDNA`, PacBio ,`ONT dRNA`) %>% 
  mutate(across(everything(), as.integer))

# Compute Jaccard similarity
jaccard_sim <- function(x, y) {
  sum(x & y) / sum(x | y)
}

platforms <- colnames(binary_matrix)
jaccard_matrix <- outer(platforms, platforms, Vectorize(function(i, j) {
  jaccard_sim(binary_matrix[[i]], binary_matrix[[j]])
}))

# Convert to matrix format
colnames(jaccard_matrix) <- rownames(jaccard_matrix) <- platforms

label_func <- function(j, i, x, y, width, height, fill) {
    grid.text(sprintf("%.2f", jaccard_matrix[i, j]), x, y, gp = gpar(fontsize = 10))
}
Heatmap(jaccard_matrix, 
        name = "Jaccard Similarity", 
        col = colorRampPalette(c("grey", "#06317F"))(50),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(round(jaccard_matrix[i, j], 2), x, y, 
                    gp = gpar(col = "white"))  # Set font color to white
        },
        na_col = "white")  # Keeps upper triangle blank
```



### Detected and undetected gene length distribution
```{r GENE: discovered gene distribution, fig.height=3, fig.width=5}
gene_discovery$length_bin <- cut(gene_discovery$`Gene length`, 
                                 breaks = c(0, 500, 1000, 1500, 2000, 2500, Inf), 
                                 labels = c("0-0.5k", "0.5-1k", "1-1.5k", "1.5-2k", "2-2.5k", "2.5k+"))

bg_dist <- gene_discovery %>%
  filter(!is.na(length_bin)) %>%
  count(length_bin) %>%
  mutate(platform = "All annotated Gene")


p1 <- gene_discovery %>%
  pivot_longer(cols = c("Illumina", "ONT cDNA", "PacBio", "ONT dRNA"), 
               names_to = "platform", values_to = "keep") %>%
  filter(keep) %>%
  filter(!is.na(length_bin)) %>% 
  count(length_bin, platform) %>%  # Summarize counts per category
  ggplot(aes(x = length_bin, y = n, color = platform, group = platform)) +
  # background line
  geom_line(data = bg_dist, aes(x = length_bin, y = n, group = 1),
            color = "grey60", linewidth = 1, linetype = "dashed") +
  # platform-specific lines
  geom_line(linewidth = 1) +
  geom_point(size = 2) +
  theme_minimal() +
  labs(title = "Detected Gene", 
       x = "Gene length (average transcript length)", 
       y = "# of Gene detected") + 
  scale_color_manual(values = color_palette[c("Illumina", "ONT_1", "ONT_2", "PacBio")] %>% unname())
```


```{r Plarform specific absense gene, fig.height=3, fig.width=12}

p2 <- gene_discovery %>%
  pivot_longer(cols = c("Illumina", "ONT cDNA", "PacBio", "ONT dRNA"),
               names_to = "platform", values_to = "keep") %>%
  group_by(gene_id) %>%
  mutate(n_detected = sum(keep)) %>%
  ungroup() %>%
  filter(keep & n_detected == 1) %>%   # detected in exactly 1 platform, absent in the other 3
  filter(!is.na(length_bin)) %>%
  count(length_bin, platform) %>%
  ggplot(aes(x = length_bin, y = n, color = platform, group = platform)) +
  geom_line(size = 1) +
  geom_point(size = 2) +
  theme_minimal() +
  labs(title = "Platform-specific Detection",
       x = "Gene length (average transcript length)",
       y = "# of Gene detected") +
  scale_color_manual(values = color_palette[c("Illumina", "ONT_1", "ONT_2", "PacBio")] %>% unname())

p3 <- gene_discovery %>%
  pivot_longer(cols = c("Illumina", "ONT cDNA", "PacBio", "ONT dRNA"),
               names_to = "platform", values_to = "keep") %>%
  group_by(gene_id) %>%
  mutate(n_detected = sum(keep)) %>%
  ungroup() %>%
  filter(!keep & n_detected == 3) %>%   # absent in exactly 1 platform, present in the other 3
  filter(!is.na(length_bin)) %>% 
  count(length_bin, platform) %>%
  ggplot(aes(x = length_bin, y = n, color = platform, group = platform)) +
  geom_line(size = 1) +
  geom_point(size = 2) +
  theme_minimal() +
  labs(title = "Platform-specific Absence",
       x = "Gene length (average transcript length)",
       y = "# of Gene absent") +
  scale_color_manual(values = color_palette[c("Illumina", "ONT_1", "ONT_2", "PacBio")] %>% unname())

(p1 | p2 | p3) + plot_layout(guides = "collect")
```


## Additional Analysis Not Included
### Inspect Pacbio specific genes
Listing top expressed Platform specific identification on H146 (randomly picked one)
```{r}

pb_specific_gene <-  gene_discovery %>% 
  filter( !Illumina &  !`ONT cDNA` &  PacBio & !`ONT dRNA` & `Intronic Gene` & biotype != "protein_coding")  %>% pull(gene_id)

pb_bulk.gene.dge$counts[pb_specific_gene, ] %>%
  as_tibble(rownames = "gene_id") %>%
  rowwise() %>%
  mutate(avg_count = mean(c_across(-gene_id))) %>%
  ungroup() %>%
  arrange(desc(avg_count))

```

```{r, include=F, eval=F}
inspect_gene_list <- pb_bulk.gene.dge$counts[pb_specific_gene,] %>%
  as_tibble(rownames="gene_id") %>% 
  arrange(desc(H1975)) %>% pull(gene_id) %>% .[1:50]

plot_df %>% filter(
        !Illumina,
        !`ONT cDNA` &
        PacBio &
        !`ONT dRNA`
      )  %>% filter(gene_id %in% inspect_gene_list)
```


```{r, include=F, eval=F}
pb_bulk.gene.dge$counts[pb_specific_gene,] %>%
  as_tibble(rownames="gene_id") %>% 
  arrange(desc(H1975)) %>% left_join(plot_df, by="gene_id") %>% filter(`Intronic Gene`==F)

```
#### Coexpression of paired genes
```{r, include=F, eval=F}
library(dplyr)
library(ggplot2)
library(ggpubr)
library(tidyr)

# Define the specific sample of interest
specific_sample <- "H1975"

# Create a list of datasets
datasets <- list(
  ill = ill_bulk.gene.dge$counts,
  ont = ont_bulk.gene.dge$counts,
  pb = pb_bulk.gene.dge$counts,
  drna = drna_bulk.gene.dge$counts
)

# Create an empty list to store results
results <- list()

# Iterate through the datasets
for (dataset_name in names(datasets)) {
  counts_data <- datasets[[dataset_name]]
  
  # Extract expression values for each gene in overlaps_tibble for the specific sample
  overlaps_expr <- overlaps_tibble %>%
    filter(exon_gene %in% rownames(counts_data),
           intron_gene %in% rownames(counts_data)) %>%
    mutate(exon_expr = log1p(counts_data[exon_gene, specific_sample]),
           intron_expr = log1p(counts_data[intron_gene, specific_sample])) %>%
    mutate(dataset = dataset_name)  # Add the dataset column
  
  # Filter out rows where intron expression is below 5 for the specific sample
  overlaps_expr_filtered <- overlaps_expr %>%
    filter(counts_data[intron_gene, specific_sample] >= 5)
  
  # Compute Pearson correlation for the filtered data
  cor_value <- cor(overlaps_expr_filtered$exon_expr, overlaps_expr_filtered$intron_expr, method = "pearson")
  
  # Store the result in the list
  results[[dataset_name]] <- list(
    data = overlaps_expr_filtered,
    cor_value = cor_value
  )
}

# Combine all results into one data frame
combined_results <- bind_rows(lapply(results, function(x) x$data), .id = "dataset")

# Create the hex plot with regression line and correlation annotation
ggplot(combined_results, aes(x = exon_expr, y = intron_expr)) +
  geom_hex(bins = 30, aes(fill = ..count..), color = "white") +  # Use hexagonal bins with 30 bins
  scale_fill_viridis_c(option = "C") +  # Use a colorblind-friendly and distinct palette (Viridis)
  labs(x = paste("Log Exon Gene Expression (", specific_sample, ")", sep = ""),
       y = paste("Log Intron Gene Expression (", specific_sample, ")", sep = ""),
       title = paste("Correlation of Log Expression for", specific_sample)) +
  theme_minimal() +
  geom_smooth(method = "lm", color = "darkred", se = FALSE) +  # Add regression line in dark red
  facet_wrap(~dataset) +  # Facet by dataset
  # Annotate the Pearson correlation coefficient on each plot
  geom_text(data = combined_results %>%
              group_by(dataset) %>%
              summarise(cor = cor(exon_expr, intron_expr, method = "pearson"), .groups = "drop"),
            aes(x = Inf, y = Inf, label = paste("Pearson Correlation = ", round(cor, 2))),
            hjust = 1.1, vjust = 1.1, size = 4, color = "black")  # Adjust annotation position
```

```{r, include=F, eval=F}
library(dplyr)
library(ggplot2)
library(ggpubr)
library(tidyr)

# Define the specific sample of interest
specific_sample <- "H1975"

# Create a list of datasets
datasets <- list(
  ill = ill_bulk.gene.dge$counts,
  ont = ont_bulk.gene.dge$counts,
  pb = pb_bulk.gene.dge$counts,
  drna = drna_bulk.gene.dge$counts
)

# Create an empty list to store results
results <- list()

# Iterate through the datasets
for (dataset_name in names(datasets)) {
  counts_data <- datasets[[dataset_name]]
  
  # Extract expression values for each gene in overlaps_tibble for the specific sample
  overlaps_expr <- overlaps_tibble %>%
    filter(exon_gene %in% pb_specific_gene,
           intron_gene %in% rownames(counts_data)) %>%
    mutate(exon_expr = log1p(counts_data[exon_gene, specific_sample]),
           intron_expr = log1p(counts_data[intron_gene, specific_sample])) %>%
    mutate(dataset = dataset_name)  # Add the dataset column
  
  # Filter out rows where intron expression is below 5 for the specific sample
  overlaps_expr_filtered <- overlaps_expr %>%
    filter(counts_data[intron_gene, specific_sample] >= 5)
  
  # Compute Spearman correlation for the filtered data
  cor_value <- cor(overlaps_expr_filtered$exon_expr, overlaps_expr_filtered$intron_expr, method = "spearman")
  
  # Store the result in the list
  results[[dataset_name]] <- list(
    data = overlaps_expr_filtered,
    cor_value = cor_value
  )
}

# Combine all results into one data frame
combined_results <- bind_rows(lapply(results, function(x) x$data), .id = "dataset")

# Create the hex plot with regression line and correlation annotation
ggplot(combined_results, aes(x = exon_expr, y = intron_expr)) +
  geom_hex(bins = 30, aes(fill = ..count..), color = "white") +  # Use hexagonal bins with 30 bins
  scale_fill_viridis_c(option = "C") +  # Use a colorblind-friendly and distinct palette (Viridis)
  labs(x = paste("Log Exon Gene Expression (", specific_sample, ")", sep = ""),
       y = paste("Log Intron Gene Expression (", specific_sample, ")", sep = ""),
       title = paste("Correlation of Log Expression for", specific_sample)) +
  theme_minimal() +
  geom_smooth(method = "lm", color = "darkred", se = FALSE) +  # Add regression line in dark red
  facet_wrap(~dataset) +  # Facet by dataset
  # Annotate the Spearman correlation coefficient on each plot
  geom_text(data = combined_results %>%
              group_by(dataset) %>%
              summarise(cor = cor(exon_expr, intron_expr, method = "spearman"), .groups = "drop"),
            aes(x = Inf, y = Inf, label = paste("Spearman Correlation = ", round(cor, 2))),
            hjust = 1.1, vjust = 1.1, size = 4, color = "black")  # Adjust annotation position
```


# Transcript discovery analysis
## Get `tx_discovery` 
### Tx detection
```{r}
# Filter for human Tx 
ill_bulk.tx.dge <- ill_bulk.tx.dge[is.human(ill_bulk.tx.dge),]
ont_bulk.tx.dge <- ont_bulk.tx.dge[is.human(ont_bulk.tx.dge),]
pb_bulk.tx.dge <- pb_bulk.tx.dge[is.human(pb_bulk.tx.dge),]
drna_bulk.tx.dge <- drna_bulk.tx.dge[is.human(drna_bulk.tx.dge),]


all_tx <- Reduce(union, lapply(list(ill_bulk.tx.dge, ont_bulk.tx.dge, pb_bulk.tx.dge, drna_bulk.tx.dge), rownames))
ill_bulk.tx.dge <- pad_dge(ill_bulk.tx.dge, all_tx)
ont_bulk.tx.dge <- pad_dge(ont_bulk.tx.dge, all_tx)
pb_bulk.tx.dge <- pad_dge(pb_bulk.tx.dge, all_tx)
drna_bulk.tx.dge <- pad_dge(drna_bulk.tx.dge, all_tx)


# Filter by expression
tx_discovery <- tibble(
  tx_id = all_tx,
  "Illumina" = filterByExpr(ill_bulk.tx.dge, group = ill_bulk.tx.dge$samples$group, min.count=5),
  "ONT cDNA" = filterByExpr(ont_bulk.tx.dge, group = ont_bulk.tx.dge$samples$group, min.count=5),
  "PacBio" = filterByExpr(pb_bulk.tx.dge, group = pb_bulk.tx.dge$samples$group, min.count=5),
  "ONT dRNA" = filterByExpr(drna_bulk.tx.dge, group = drna_bulk.tx.dge$samples$group, min.count=5)
) %>% filter(
  Illumina | `ONT cDNA` | PacBio | `ONT dRNA`
)
# split and get biotype (last element of tx_id)
tx_discovery <- tx_discovery %>%  mutate(
    biotype = strsplit(tx_id, "\\|") %>% sapply(function(x) x[8]),
    length = strsplit(tx_id, "\\|") %>% sapply(function(x) x[7]) %>% as.integer(),
    gene_id = strsplit(tx_id, "\\|") %>% sapply(function(x) x[2]),
    tx_id = strsplit(tx_id, "\\|") %>% sapply(function(x) x[1])
  )
```

### Add biotype
```{r}
tx_discovery <- tx_discovery %>% mutate(biotype = ifelse(biotype %in% 
            names(sort(table(biotype), decreasing = TRUE)[1:6]), 
            biotype, "Others"))
```

### Save Gene and Tx detection
```{r}
output_list[["tx_discovery"]] <- tx_discovery # save for downstream analysis
saveRDS(output_list, params$out.rds)
```

## Tx detection plots
### Main upset plot
```{r Transcript: Main upset plot, fig.height=4, fig.width=8}
library(ComplexUpset)
ComplexUpset::upset(
  tx_discovery %>% mutate(
    biotype = case_when(
      biotype =="protein_coding" ~ "Protein-coding",
      TRUE ~ "Non-protein-coding"
    )
  ),
  intersect = c('Illumina', "ONT cDNA", "PacBio", "ONT dRNA"),
  annotations = list(
        # 1st method - passing list:
        "Transcript length" = list(
          aes = aes(x = intersection, y = `length`),
          # provide a list if you wish to add several geoms
          geom = geom_boxplot(na.rm = TRUE, outliers = FALSE)
        )
        # ,
        # "BioType" = list(
        #   aes = aes(x = intersection, fill = biotype),
        #   geom = geom_bar(position = "fill")
        # )
  ),
  base_annotations=list(
        'Intersection size'=ComplexUpset::intersection_size(
            counts=TRUE,
            mapping=aes(fill=biotype),
            text=element_text(size = 3, vjust=-0.2)
        ) + scale_fill_manual(values = c("Non-protein-coding"="#BAC5CB",
                              "Protein-coding"='#B5C952'))
    ),
  set_sizes=(
        ComplexUpset::upset_set_size(
            geom=geom_bar(
                aes(fill=biotype),
                width=0.8
            ) ,
            position='right'
        )+ scale_fill_manual(values = c("Non-protein-coding"="#BAC5CB",
                              "Protein-coding"='#B5C952')) + 
          theme(legend.position = "none")
  ),
  guides='over',
  wrap = TRUE
)
```


```{r Transcript: Main agreement plot, fig.height=3, fig.width=8}
agreement_plot_tx <- function(d){
  # the agreement_plot function take gene_id columns by default. Modify to work on tx_discovery
  p <- d  %>% mutate(gene_id=tx_id) %>% agreement_plot()
  p + labs(y = "Number of Transcripts")
}

 tx_discovery %>% mutate(
    biotype = case_when(
      biotype =="protein_coding" ~ "Protein-coding",
      TRUE ~ "Non-protein-coding"
    )) %>% agreement_plot_tx
```


```{r Transcript: Sup agreement plot, fig.height=3, fig.width=12}
(
  ( tx_discovery %>% filter(biotype =="protein_coding") %>% agreement_plot_tx + ggtitle("Protein-coding") ) |
  ( tx_discovery %>% filter(biotype !="protein_coding") %>% agreement_plot_tx+ ggtitle("Non-protein-coding")) 
) + patchwork::plot_layout(guides = "collect")
```
## Jaccard similary
```{r transcript identification Jaccard similarity, fig.height=3, fig.width=4.5}

# Jaccard similarity
library(ComplexHeatmap)
binary_matrix <- tx_discovery %>% 
  select( Illumina, `ONT cDNA`, PacBio ,`ONT dRNA`) %>% 
  mutate(across(everything(), as.integer))

# Compute Jaccard similarity
jaccard_sim <- function(x, y) {
  sum(x & y) / sum(x | y)
}

platforms <- colnames(binary_matrix)
jaccard_matrix <- outer(platforms, platforms, Vectorize(function(i, j) {
  jaccard_sim(binary_matrix[[i]], binary_matrix[[j]])
}))

# Convert to matrix format
colnames(jaccard_matrix) <- rownames(jaccard_matrix) <- platforms

label_func <- function(j, i, x, y, width, height, fill) {
    grid.text(sprintf("%.2f", jaccard_matrix[i, j]), x, y, gp = gpar(fontsize = 10))
}
Heatmap(jaccard_matrix, 
        name = "Jaccard Similarity", 
        col = colorRampPalette(c("grey", "#06317F"))(50),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(round(jaccard_matrix[i, j], 2), x, y, 
                    gp = gpar(col = "white"))  # Set font color to white
        },
        na_col = "white")  # Keeps upper triangle blank
```
## Discovered transcript length
```{r Discovered transcript length dist, fig.height=3, fig.width=8}
tx_discovery$length_bin <- cut(tx_discovery$length, breaks = c(0, 500, 1000, 1500, 2000, 2500,Inf), labels = c("0-0.5k", "0.5-1k", "1-1.5k", "1.5-2k", "2-2.5k", "2.5k+"))

p1 <- tx_discovery %>%
  pivot_longer(cols = c("Illumina", "ONT cDNA", "PacBio", "ONT dRNA"), 
               names_to = "platform", values_to = "keep") %>%
  filter(keep) %>%
  count(length_bin, platform) %>%  # Summarize counts per category
  ggplot(aes(x = length_bin, y = n, color = platform, group = platform)) +
  geom_line(size = 1) +  # Use lines instead of bars
  geom_point(size = 2) +  # Add points for visibility
  theme_minimal() +
  labs(title = "Detected Transcript", 
       x = "Transcript length", 
       y = "# of transcript detected") + 
  scale_color_manual(values = color_palette[c("Illumina", "ONT_1", "ONT_2", "PacBio")] %>% unname())

p2 <- tx_discovery %>%
  filter(biotype=="protein_coding") %>%
  pivot_longer(cols = c("Illumina", "ONT cDNA", "PacBio", "ONT dRNA"), 
               names_to = "platform", values_to = "keep") %>%
  filter(keep) %>%
  count(length_bin, platform) %>%  # Summarize counts per category
  ggplot(aes(x = length_bin, y = n, color = platform, group = platform)) +
  geom_line(size = 1) +  # Use lines instead of bars
  geom_point(size = 2) +  # Add points for visibility
  theme_minimal() +
  labs(title = "Detected protein-coding Transcript", 
       x = "Transcript length", 
       y = "# of transcript detected") + 
  scale_color_manual(values = color_palette[c("Illumina", "ONT_1", "ONT_2", "PacBio")] %>% unname())

patchwork::wrap_plots(p1, p2) + 
  patchwork::plot_layout(guides = 'collect')


```


## Additional Analysis Not Included
### Exclude Illumina
```{r, fig.height=6, fig.width=8, echo=F, include=F, eval=F}
ComplexUpset::upset(
  tx_discovery,
  intersect = c("ONT cDNA", "PacBio", "ONT dRNA"),
  annotations = list(
        # 1st method - passing list:
        "Average transcript length" = list(
          aes = aes(x = intersection, y = length),
          # provide a list if you wish to add several geoms
          geom = geom_boxplot(na.rm = TRUE, outliers = FALSE)
        ),
        "BioType" = list(
          aes = aes(x = intersection, fill = biotype),
          geom = geom_bar(position = "fill")
        )
  ),
  set_sizes=(
        ComplexUpset::upset_set_size(
            geom=geom_bar(
                aes(fill=biotype),
                width=0.8
            ) , position='right'
        )+ theme(legend.position = "none")
    ),
  wrap = TRUE, guides='over'
)

# protein coding
ComplexUpset::upset(
  tx_discovery %>% filter(grepl("protein_coding*",biotype)),
  intersect = c("ONT cDNA", "PacBio", "ONT dRNA"),
  annotations = list(
        # 1st method - passing list:
        "Average transcript length" = list(
          aes = aes(x = intersection, y = length),
          # provide a list if you wish to add several geoms
          geom = geom_boxplot(na.rm = TRUE, outliers = FALSE)
        ),
        "BioType" = list(
          aes = aes(x = intersection, fill = biotype),
          geom = geom_bar(position = "fill")
        )
  ),
  set_sizes=(
        ComplexUpset::upset_set_size(
            geom=geom_bar(
                aes(fill=biotype),
                width=0.8
            ) , position='right'
        )+ theme(legend.position = "none")
    ),
  wrap = TRUE, guides='over'
)

# protein coding
ComplexUpset::upset(
  tx_discovery %>% filter(!grepl("protein_coding*",biotype)),
  intersect = c("ONT cDNA", "PacBio", "ONT dRNA"),
  annotations = list(
        # 1st method - passing list:
        "Average transcript length" = list(
          aes = aes(x = intersection, y = length),
          # provide a list if you wish to add several geoms
          geom = geom_boxplot(na.rm = TRUE, outliers = FALSE)
        ),
        "BioType" = list(
          aes = aes(x = intersection, fill = biotype),
          geom = geom_bar(position = "fill") 
        )
  ),
  set_sizes=(
        ComplexUpset::upset_set_size(
            geom=geom_bar(
                aes(fill=biotype),
                width=0.8
            ), position='right'
        )+ theme(legend.position = "none")
    ),
  guides='over',
  wrap = TRUE
)
```

