---
title: "DE analysis"
author: "Yupei & Ash"
output:
  html_document:
    code_folding: hide
    toc: yes
    toc_float: yes
    number_sections: yes
    toc_depth: 4
  pdf_document:
    toc: yes
    toc_depth: '4'
params:
  random_seed: 2024
  sirv_ercc_gtf: '/vast/projects/LongBench/reference_files/SIRV_Set4_Norm_Sequences_20210507/SIRV_ERCC_longSIRV_multi-fasta_20210507.gtf'
  sequins_gtf: '/vast/projects/LongBench/reference_files/Sequin_resources/v2.4/rnasequin_annotation_2.4.gtf'
  sequins_tsv: '/vast/projects/LongBench/reference_files/Sequin_resources/v2.4/rnasequin_isoforms_2.4.tsv'
  human_gtf: '/vast/projects/LongBench/reference_files/GRCh38/gencode.v44.annotation.gtf'
  sirv_csv: '/home/users/allstaff/you.yu/LongBench/reference_files/SIRV_Set4_Norm_Sequences_20210507/SIRV_set4_concentration.csv'
  ont_bulk_oarfish_dir: '/vast/projects/LongBench/analysis/lr_bulk/result/oarfish_cov_output/ont_bulk'
  pb_bulk_oarfish_dir: '/vast/projects/LongBench/analysis/lr_bulk/result/oarfish_cov_output/pb_bulk'
  ont_drnd_oarfish_dir: '/vast/projects/LongBench/analysis/lr_bulk/result/oarfish_cov_output/dRNA_bulk'
  ill_bulk_salmon_dir: '/vast/projects/LongBench//analysis/sr_bulk/result/salmon/salmon_quant'
  bulk_meta: '/vast/projects/LongBench/sequencing_data/illumina_bulk/metadata.txt'
  cache_dir: NULL
  fig.path: NULL
  out.rds: '/vast/projects/LongBench/analysis/reports/RDS/bulk_de_human_summary.rds'
---
# Setup:
```{r Input filename setup, include=FALSE, eval=F}
# This chunk is used only when running the file chunk by chunk manually
params <- list(
  random_seed = 2024,
  cache_dir = NULL,
  sirv_ercc_gtf = '/vast/projects/LongBench/reference_files/SIRV_Set4_Norm_Sequences_20210507/SIRV_ERCC_longSIRV_multi-fasta_20210507.gtf',
  sequins_gtf = '/vast/projects/LongBench/reference_files/Sequin_resources/v2.4/rnasequin_annotation_2.4.gtf',
  sequins_tsv = '/vast/projects/LongBench/reference_files/Sequin_resources/v2.4/rnasequin_isoforms_2.4.tsv',
  human_gtf = '/vast/projects/LongBench/reference_files/GRCh38/gencode.v44.annotation.gtf',
  sirv_csv = '/home/users/allstaff/you.yu/LongBench/reference_files/SIRV_Set4_Norm_Sequences_20210507/SIRV_set4_concentration.csv',
  ont_bulk_oarfish_dir = "/vast/projects/LongBench/analysis/lr_bulk/result/oarfish_cov_output/ont_bulk",
  pb_bulk_oarfish_dir = "/vast/projects/LongBench/analysis/lr_bulk/result/oarfish_cov_output/pb_bulk",
  ont_drnd_oarfish_dir = "/vast/projects/LongBench/analysis/lr_bulk/result/oarfish_cov_output/dRNA_bulk",
  ill_bulk_salmon_dir = "/vast/projects/LongBench//analysis/sr_bulk/result/salmon/salmon_quant",
  bulk_meta = "/vast/projects/LongBench/sequencing_data/illumina_bulk/metadata.txt",
  out.rds= '/vast/projects/LongBench/analysis/reports/RDS/bulk_de_human_summary.rds'
)
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  cache = !is.null(params$cache_dir),  # Enable caching only if cache_dir is provided
  fig.width = 15,                     # Set default figure width
  fig.height = 10,                    # Set default figure height
  warning = FALSE,                    # Suppress warnings in the output
  fig.path = params$fig.path,         # Set figure path (NULL if not provided)
  dev = if (!is.null(params$fig.path)) c("png", "svg") else NULL  # Set output devices if fig.path is provided
)

# Set cache.path only if params$cache_dir is not NULL
if (!is.null(params$cache_dir)) {
  knitr::opts_chunk$set(cache.path = params$cache_dir)
}
```

```{r Libraries,include=FALSE}
library(limma)
library(edgeR)
library(tidyverse)
library(RColorBrewer)
library(dplyr)
library(tidyr)
library(ggplot2)
library(ggrepel)
library(glue)
library(ggplot2)
library(gridExtra)
library(ggExtra)
library(knitr)
library(kableExtra)
library(rtracklayer)
library(scales)
library(UpSetR)
library(ggtree)
library(aplot)
source("/home/users/allstaff/you.yu/LongBench/analysis/workflow/scripts/Rfunctions.R")

# setup some conflict functions
rename <- dplyr::rename
select <- dplyr::select

# Set color palette
color_palette <- c(
  PacBio = "#df1995",
  ONT = "#00789b",
  ONT_1 = "#04476c",
  ONT_2 = "#24cdcd",
  Illumina = "#e88b20"
)
```

```{r function for splitting the transcript origin}
is.sequins <- function(dge) {grepl("^R", rownames(dge))}
is.human <- function(dge) {grepl("^ENS", rownames(dge))}
is.sirv_or_ercc <- function(dge) {!is.human(dge) & !is.sequins(dge)} # thie is the whole SIRV set4 (ERCC + SIRV E0 + Long SIRV)
is.sirv_E0 <- function(dge) {grepl("^SIRV\\d{3}$", rownames(dge))} 
is.sirv_all <- function(dge) {grepl("^SIRV", rownames(dge))} # including all SIRV E0 and Long SIRV
is.long_sirv <- function(dge) {grepl("^SIRV\\d{4}", rownames(dge))}
is.ercc <- function(dge) {!is.human(dge) & !is.sequins(dge) & !is.sirv_all(dge)}
```
# Create DGE from input
```{r, get number of transcripts gene from gtf}
calcTxNum <- function(G.Tx.map){
  geneid = rep(NA, nrow(G.Tx.map))
  txcount = table(G.Tx.map$gene_id)
  txnum = as.numeric(txcount[match(G.Tx.map$gene_id, names(txcount))])
}
# SIRV
SIRV_tx <- GenomicFeatures::makeTxDbFromGFF(params$sirv_ercc_gtf, format = "gtf") %>% GenomicFeatures::transcripts(columns = c("tx_id", "tx_name", "gene_id"))
SIRV.G.Tx.map <- tibble(
  gene_id = SIRV_tx$gene_id %>% unlist,
  tx_name = SIRV_tx$tx_name
)
rm(SIRV_tx)
SIRV.G.Tx.map$tx_count <- SIRV.G.Tx.map %>% calcTxNum

# Sequins
sequins_tx  <- GenomicFeatures::makeTxDbFromGFF(params$sequins_gtf, format = "gtf") %>% GenomicFeatures::transcripts(columns = c("tx_id", "tx_name", "gene_id"))
sequins.G.Tx.map <- tibble(
  gene_id = sequins_tx$gene_id %>% unlist,
  tx_name = sequins_tx$tx_name
)
rm(sequins_tx)
sequins.G.Tx.map$tx_count <- sequins.G.Tx.map %>% calcTxNum

# Human
human_tx  <- GenomicFeatures::makeTxDbFromGFF(params$human_gtf, format = "gtf") %>% GenomicFeatures::transcripts(columns = c("tx_id", "tx_name", "gene_id"))
human.G.Tx.map <- tibble(
  gene_id = human_tx$gene_id %>% unlist,
  tx_name = human_tx$tx_name
)
rm(human_tx)
human.G.Tx.map$tx_count <- human.G.Tx.map %>% calcTxNum
```

```{r Create DGE from input, echo=T}
bulk.meta <- read.csv(params$bulk_meta)
rownames(bulk.meta) <- bulk.meta$sample

# SR salmon
ill_bulk.dge <- get_dge_from_salmon(params$ill_bulk_salmon_dir, ".*/")
# LR oarfish
ont_bulk.dge.oarfish <- get_dge_from_oarfish(params$ont_bulk_oarfish_dir, ".*/")
pb_bulk.dge.oarfish <- get_dge_from_oarfish(params$pb_bulk_oarfish_dir, ".*/")
ont_drna.dge.oarfish <- get_dge_from_oarfish(params$ont_drnd_oarfish_dir, ".*/")

# Extract human transcript counts
ill_bulk.dge <- ill_bulk.dge[is.human(ill_bulk.dge), ]
ont_bulk.dge.oarfish <- ont_bulk.dge.oarfish[is.human(ont_bulk.dge.oarfish), ]
pb_bulk.dge.oarfish <- pb_bulk.dge.oarfish[is.human(pb_bulk.dge.oarfish), ]
ont_drna.dge.oarfish <- ont_drna.dge.oarfish[is.human(ont_drna.dge.oarfish), ]
```

# Count data Preprocessing
1. Aggregate transcript counts to gene level using `tximport` with unfiltered transcript counts
2. Filter out lowly expressed genes and transcript using `filterByExpr` for each dataset, using default parameters 
  - The filtering will be done on single dataset level, and the **intersect** of the filtered genes will be used for the downstream analysis
  - Note: The starting number of gene/transcript is smaller for the salmon pseudoalignment method compared to the kallisto and oarfish methods. This is due to the fact that the genecode transcript FASTA has more transcript/gene than their annotation GTF file. Given the proportion is tiny (<0.2% transcripts), to keep the analysis consistent, we will remove the genes (either expressed or not) that are not present in salmon output from the kallisto and oarfish datasets.
3. Normalise the gene counts using `calcNormFactors` with `TMM` methods for each dataset 

## 1. Aggregate transcript counts to gene level using `tximport` with unfiltered transcript counts
```{r}
library(tximport) # install latest version from github

# Extract tx_name and gene_id columns from human and sequins
human_tx2gene <- human.G.Tx.map[, c("tx_name", "gene_id")]
sequins_tx2gene <- sequins.G.Tx.map[, c("tx_name", "gene_id")]
SIRV_tx2gene <- SIRV.G.Tx.map[, c("tx_name", "gene_id")]
combined_tx2gene <- rbind(human_tx2gene, sequins_tx2gene,SIRV_tx2gene)

# Get quant.sf file paths
get_quant_paths <- function(dirs, file_name) {
  # Generate file paths for quantification files
  quant_paths <- file.path(dirs, file_name)
  return(quant_paths)
}

# Define directories with their corresponding expected file names
oarfish_dirs <- list(
  ill_bulk = list(dir = params$ill_bulk_salmon_dir, file_name = "quant.sf"),
  ont_bulk = list(dir = params$ont_bulk_oarfish_dir, file_name = ".quant"),
  pb_bulk = list(dir = params$pb_bulk_oarfish_dir, file_name = ".quant"),
  drna_bulk = list(dir = params$ont_drnd_oarfish_dir, file_name = ".quant")
)

# Apply function to each directory to get all quant file paths
quant_paths <- lapply(oarfish_dirs, function(dir_info) {
  # Get the directory and file name from dir_info
  dir <- dir_info$dir
  file_name <- dir_info$file_name
  
  # List all subdirectories and get the quant file paths
  sub_dirs <- list.dirs(dir, full.names = TRUE, recursive = FALSE)
  get_quant_paths(sub_dirs, file_name)
})

# Flatten the list of quant_paths into a named list
quant_paths <- lapply(quant_paths, unlist)

# Function to import DGE from tximport
get_dge_from_txi <- function(quant_dir, sample_prefix_regex, type) {
  # Import data using tximport
  txi <- tximport(quant_dir,
                  type = type,
                  tx2gene = combined_tx2gene,
                  ignoreAfterBar = TRUE,
                  countsFromAbundance = "no") # raw counts
  counts <- as.data.frame(txi$counts, stringAsFactors = FALSE)
  sample_names <- gsub(sample_prefix_regex, '', basename(dirname(quant_dir)))
  colnames(counts) <- sample_names
  rst.dge <- DGEList(counts = counts)
  rst.dge$genes <- data.frame(Length=txi$length %>% rowMeans())
  rownames(rst.dge$samples) <- sub(sample_prefix_regex, '', rownames(rst.dge$samples))
  colnames(rst.dge$counts) <- sub(sample_prefix_regex, "", colnames(rst.dge$counts))
  # colnames(rst.dge$gene$Length) <- sub(sample_prefix_regex, "", colnames(rst.dge$gene$Length))
  
  rst.dge$samples <- merge(rst.dge$samples %>% select(-group), bulk.meta, by = "row.names", all.x = TRUE) %>% select(-Row.names)
  rst.dge$samples <- rst.dge$samples[match(colnames(rst.dge$counts), rst.dge$samples$sample),]
  return(rst.dge)
}

# Process each quant path for different bulk types
ill_bulk.gene.dge <- get_dge_from_txi(quant_paths$ill_bulk, ".*/", "salmon")
ont_bulk.gene.dge <- get_dge_from_txi(quant_paths$ont_bulk, ".*/", "oarfish")
pb_bulk.gene.dge <- get_dge_from_txi(quant_paths$pb_bulk, ".*/", "oarfish")
drna_bulk.gene.dge <- get_dge_from_txi(quant_paths$drna_bulk, ".*/", "oarfish")

# Extract human gene counts
ill_bulk.gene.dge <- ill_bulk.gene.dge[is.human(ill_bulk.gene.dge), ]
ont_bulk.gene.dge <- ont_bulk.gene.dge[is.human(ont_bulk.gene.dge), ]
pb_bulk.gene.dge <- pb_bulk.gene.dge[is.human(pb_bulk.gene.dge), ]
drna_bulk.gene.dge <- drna_bulk.gene.dge[is.human(drna_bulk.gene.dge), ]
```

## 2. Filter out lowly expressed genes and transcript using `filterByExpr` for each dataset, using default parameters
```{r gene filtering for DE}
#GENE 
## 1. get common genes
common_genes <- Reduce(intersect, lapply(list(ill_bulk.gene.dge, ont_bulk.gene.dge, pb_bulk.gene.dge, drna_bulk.gene.dge), rownames))

ill_bulk.gene.dge <- ill_bulk.gene.dge[common_genes, ]
ont_bulk.gene.dge <- ont_bulk.gene.dge[common_genes, ]
pb_bulk.gene.dge <- pb_bulk.gene.dge[common_genes, ]
drna_bulk.gene.dge <- drna_bulk.gene.dge[common_genes, ]

# save all workpalce as Rdata
# save.image("/vast/scratch/users/you.yu/test/Bulk_DE_line269.Rdata")
```

### Gene discovery analysis
```{r}
# Filter by expression
gene_discovery <- tibble(
  gene_id = common_genes,
  "Illumina" = filterByExpr(ill_bulk.gene.dge, group = ill_bulk.gene.dge$samples$sample),
  "ONT cDNA" = filterByExpr(ont_bulk.gene.dge, group = ont_bulk.gene.dge$samples$sample),
  "PacBio" = filterByExpr(pb_bulk.gene.dge, group = pb_bulk.gene.dge$samples$sample),
  "ONT dRNA" = filterByExpr(drna_bulk.gene.dge, group = drna_bulk.gene.dge$samples$sample)
) %>% filter(
  Illumina | `ONT cDNA` | PacBio | `ONT dRNA`
)

# Getting the gene biotype and gene id, using illumina data here but will be identical for all datasets
tx_ref_df <- ill_bulk.dge$genes %>%
  rownames_to_column("tx_id") %>%
  select(tx_id, Length) %>%
  tibble() %>%
  mutate(
    biotype = strsplit(tx_id, "\\|") %>% sapply(function(x) x[8]),
    gene_id = strsplit(tx_id, "\\|") %>% sapply(function(x) x[2]),
    tx_id = strsplit(tx_id, "\\|") %>% sapply(function(x) x[1])
  )

gene_length_lookup <- tx_ref_df %>%
    group_by(gene_id) %>%
    summarise(mean_length = mean(Length, na.rm = TRUE), .groups = "drop")

# Connect to Ensembl
library(biomaRt)
ensembl <- useEnsembl(biomart = "genes", dataset = "hsapiens_gene_ensembl")

gene_discovery <- gene_discovery %>%
    left_join(gene_length_lookup, by = c("gene_id")) %>%
    rename(`Gene length` = mean_length)
gene_discovery$gene_id <- gene_discovery$gene_id %>% sub("\\..*", "", .)
biotypes <- getBM(
  attributes = c("ensembl_gene_id", "gene_biotype"),
  filters = "ensembl_gene_id",
  values = gene_discovery$gene_id,
  mart = ensembl
)


gene_discovery <- gene_discovery %>%
  left_join(biotypes, by = c("gene_id" = "ensembl_gene_id")) %>%
  rename(biotype = gene_biotype) %>%  
  mutate(biotype = ifelse(biotype %in% 
            names(sort(table(biotype), decreasing = TRUE)[1:3]), 
            biotype, "Others"))


gene_discovery$biotype <- factor(gene_discovery$biotype, levels = c(sort(unique(gene_discovery$biotype)) %>% setdiff("Others"), "Others"))

ComplexUpset::upset(
  gene_discovery,
  intersect = c('Illumina', "ONT cDNA", "PacBio", "ONT dRNA"),
  annotations = list(
        # 1st method - passing list:
        "Average transcript length" = list(
          aes = aes(x = intersection, y = `Gene length`),
          # provide a list if you wish to add several geoms
          geom = geom_boxplot(na.rm = TRUE, outliers = FALSE)
        ),
        "BioType" = list(
          aes = aes(x = intersection, fill = biotype),
          geom = geom_bar(position = "fill")
        )
  ),
  set_sizes=(
        ComplexUpset::upset_set_size(
            geom=geom_bar(
                aes(fill=biotype),
                width=0.8
            ) 
        )+ theme(legend.position = "none")
    ),
  wrap = TRUE
)


gene_discovery$length_bin <- cut(gene_discovery$`Gene length`, breaks = c(0, 500, 1000, 1500, 2000, 2500,Inf), labels = c("0-0.5k", "0.5-1k", "1-1.5k", "1.5-2k", "2-2.5k", "2.5k+"))
gene_discovery %>%
  pivot_longer(cols = c("Illumina", "ONT cDNA", "PacBio", "ONT dRNA"), names_to = "platform", values_to = "keep") %>%
  filter(keep) %>%
  ggplot(aes(x = length_bin, fill = platform)) +
  geom_bar(position = "dodge") +
  theme_minimal() +
  #theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  labs(title = "Expressed genes", x = "Gene length (average transcript length)", y = "# of Genes") + 
            scale_fill_manual(values = color_palette[c("Illumina", "ONT_1", "ONT_2", "PacBio")] %>% unname())

gene_discovery %>% 
  pivot_longer(cols = c("Illumina", "ONT cDNA", "PacBio", "ONT dRNA"), names_to = "platform", values_to = "keep") %>% filter(keep) %>%
  ggplot(aes(x = biotype, fill = platform)) +
  geom_bar(position = "dodge") + 
  theme_minimal() +
  #theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  labs(title = "Expressed genes", x = "Biotype", y = "# of Genes")+ 
  scale_fill_manual(values = color_palette[c("Illumina", "ONT_1", "ONT_2", "PacBio")] %>% unname())


# distribution of gene length stratified by biotype
# gene_discovery %>% ggplot(aes(x = `Gene length`, fill = NULL, color=biotype)) +
#   geom_density(alpha = 0.5) +
#   theme_minimal() +
#   labs(title = "Gene length distribution", x = "Gene length", y = "Density") +
#   scale_fill_viridis_d() +
#   xlim(0, 5000)
```


```{r}
## 2. filter by expression
keep <- filterByExpr(ill_bulk.gene.dge, group = ill_bulk.gene.dge$samples$sample) &
          filterByExpr(ont_bulk.gene.dge, group = ont_bulk.gene.dge$samples$sample) &
          filterByExpr(pb_bulk.gene.dge, group = pb_bulk.gene.dge$samples$sample) &
          filterByExpr(drna_bulk.gene.dge, group = drna_bulk.gene.dge$samples$sample)

table(keep)
ill_bulk.gene.dge <- ill_bulk.gene.dge[keep, ]
ont_bulk.gene.dge <- ont_bulk.gene.dge[keep, ]
pb_bulk.gene.dge <- pb_bulk.gene.dge[keep, ]
drna_bulk.gene.dge <- drna_bulk.gene.dge[keep, ]
```

```{r transcript filtering for DE}
# Transcript
## 1. get common transcripts
common_tx <- Reduce(intersect, lapply(list(ill_bulk.dge, ont_bulk.dge.oarfish, pb_bulk.dge.oarfish, ont_drna.dge.oarfish), rownames))

ill_bulk.dge <- ill_bulk.dge[common_tx, ]
ont_bulk.dge.oarfish <- ont_bulk.dge.oarfish[common_tx, ]
pb_bulk.dge.oarfish <- pb_bulk.dge.oarfish[common_tx, ]
ont_drna.dge.oarfish <- ont_drna.dge.oarfish[common_tx, ]

# Filter by expression
txi_discovery <- tibble(
  tx_id = common_tx,
  "Illumina" = filterByExpr(ill_bulk.dge, group = ill_bulk.dge$samples$sample, min.count = 5),
  "ONT cDNA" = filterByExpr(ont_bulk.dge.oarfish, group = ont_bulk.dge.oarfish$samples$sample, min.count = 5),
  "PacBio" = filterByExpr(pb_bulk.dge.oarfish, group = pb_bulk.dge.oarfish$samples$sample, min.count = 5),
  "ONT dRNA" = filterByExpr(ont_drna.dge.oarfish, group = ont_drna.dge.oarfish$samples$sample, min.count = 5)
) %>% filter(
  Illumina | `ONT cDNA` | PacBio | `ONT dRNA`
) %>%
  mutate(
    biotype = strsplit(tx_id, "\\|") %>% sapply(function(x) x[8]),
    gene_id = strsplit(tx_id, "\\|") %>% sapply(function(x) x[2]),
    Length = strsplit(tx_id, "\\|") %>% sapply(function(x) x[7]),
    tx_id = strsplit(tx_id, "\\|") %>% sapply(function(x) x[1])
  )

txi_discovery <- txi_discovery %>%
  mutate(top_biotype = ifelse(biotype %in% 
            names(sort(table(biotype), decreasing = TRUE)[1:3]), 
            biotype, "Others"))
txi_discovery$top_biotype <- factor(txi_discovery$biotype, levels = c(sort(unique(txi_discovery$biotype)) %>% setdiff("Others"), "Others"))

ComplexUpset::upset(
  txi_discovery,
  intersect = c('Illumina', "ONT cDNA", "PacBio", "ONT dRNA"),
  annotations = list(
        # 1st method - passing list:
        "Average transcript length" = list(
          aes = aes(x = intersection, y = Length),
          # provide a list if you wish to add several geoms
          geom = geom_boxplot(na.rm = TRUE, outliers = FALSE)
        ),
        "BioType" = list(
          aes = aes(x = intersection, fill = top_biotype),
          geom = geom_bar(position = "fill")
        )
  ),
  set_sizes=(
        ComplexUpset::upset_set_size(
            geom=geom_bar(
                aes(fill=top_biotype),
                width=0.8
            ) 
        )+ theme(legend.position = "none")
    ),
  wrap = TRUE
)
# 
# 
# txi_discovery$length_bin <- cut(txi_discovery$Length, breaks = c(0, 500, 1000, 1500, 2000, 2500,Inf), labels = c("0-0.5k", "0.5-1k", "1-1.5k", "1.5-2k", "2-2.5k", "2.5k+"))
# txi_discovery %>%
#   pivot_longer(cols = c("Illumina", "ONT cDNA", "PacBio", "ONT dRNA"), names_to = "platform", values_to = "keep") %>%
#   filter(keep) %>%
#   ggplot(aes(x = length_bin, fill = platform)) +
#   geom_bar(position = "dodge") +
#   theme_minimal() +
#   #theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
#   labs(title = "Expressed genes", x = "Gene length (average transcript length)", y = "# of Genes") + 
#             scale_fill_manual(values = color_palette[c("Illumina", "ONT_1", "ONT_2", "PacBio")] %>% unname())
# 
# txi_discovery %>%
#   pivot_longer(cols = c("Illumina", "ONT cDNA", "PacBio", "ONT dRNA"), names_to = "platform", values_to = "keep") %>%
#   filter(keep) %>%
#   ggplot(aes(x = biotype, fill = platform)) +
#   geom_bar(position = "dodge") +
#   theme_minimal() +
#   # theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
#   labs(title = "Expressed genes", x = "Biotype", y = "# of Genes") +
#   scale_fill_manual(values = color_palette[c("Illumina", "ONT_1", "ONT_2", "PacBio")] %>% unname())
# 
# saveRDS(list(
#   gene_discovery = gene_discovery,
#   txi_discovery = txi_discovery
# ), "/vast/projects/LongBench/analysis/workflow/rmarkdown/Bulk_DE_discovery.RDS")

```



```{r}
## 2. filter by expression
keep <- filterByExpr(ill_bulk.dge, group = ill_bulk.dge$samples$sample) &
          filterByExpr(ont_bulk.dge.oarfish, group = ont_bulk.dge.oarfish$samples$sample) &
          filterByExpr(pb_bulk.dge.oarfish, group = pb_bulk.dge.oarfish$samples$sample) &
          filterByExpr(ont_drna.dge.oarfish, group = ont_drna.dge.oarfish$samples$sample)
table(keep)

ill_bulk.dge <- ill_bulk.dge[keep, ]
ont_bulk.dge.oarfish <- ont_bulk.dge.oarfish[keep, ]
pb_bulk.dge.oarfish <- pb_bulk.dge.oarfish[keep, ]
ont_drna.dge.oarfish <- ont_drna.dge.oarfish[keep, ]
```

## 3. Normalise the gene and transcript counts using `calcNormFactors` with `TMM` methods for each dataset, separately for human

```{r TMM normalisation}
### 4. Normalise the gene counts using `calcNormFactors` with `TMM` methods for each dataset
ill_bulk.gene.dge <- calcNormFactors(ill_bulk.gene.dge, method = "TMM")
ont_bulk.gene.dge <- calcNormFactors(ont_bulk.gene.dge, method = "TMM")
pb_bulk.gene.dge <- calcNormFactors(pb_bulk.gene.dge, method = "TMM")
drna_bulk.gene.dge <- calcNormFactors(drna_bulk.gene.dge, method = "TMM")


### 4. Normalise the transcript counts using `calcNormFactors` with `TMM` methods for each dataset
ill_bulk.dge <- calcNormFactors(ill_bulk.dge, method = "TMM")
ont_bulk.dge.oarfish <- calcNormFactors(ont_bulk.dge.oarfish, method = "TMM")
pb_bulk.dge.oarfish <- calcNormFactors(pb_bulk.dge.oarfish, method = "TMM")
ont_drna.dge.oarfish <- calcNormFactors(ont_drna.dge.oarfish, method = "TMM")

```




# MDS (Merged)
## Merge DGElist 
NOTE: The merged DGElist will only be used for the MDS plot. The DE analysis will be done separately for each dataset.
```{r Merge DGEs}
# Gene
ill_bulk.gene.dge$samples$platform <- "Illumina"
ont_bulk.gene.dge$samples$platform <- "ONT cDNA"
pb_bulk.gene.dge$samples$platform <- "PacBio"
drna_bulk.gene.dge$samples$platform <- "ONT dRNA"
bulk.gene.dge.merge <- cbind(
  ill_bulk.gene.dge,
  ont_bulk.gene.dge,
  pb_bulk.gene.dge,
  drna_bulk.gene.dge
)

# Transcript
ill_bulk.dge$samples$platform <- "Illumina"
ill_bulk.dge.copy <- ill_bulk.dge
ill_bulk.dge.copy$genes$EffectiveLength <- NULL
ont_bulk.dge.oarfish$samples$platform <- "ONT cDNA"
pb_bulk.dge.oarfish$samples$platform <- "PacBio"
ont_drna.dge.oarfish$samples$platform <- "ONT dRNA"

bulk.tx.dge.merge <- cbind(
  ill_bulk.dge.copy,
  ont_bulk.dge.oarfish,
  pb_bulk.dge.oarfish,
  ont_drna.dge.oarfish
)
```

```{r}
norm_and_mds <- function(dge) {
  dge <- calcNormFactors(dge, method = "TMM")
  mds <- plotMDS(dge, plot = FALSE)
  return(mds)
}
myMDS <- function(dge, mds) {
  # normalisation
  mds_data <- tibble(
    Dim1 = mds$x,
    Dim2 = mds$y,
    "Cell lines" = dge$sample$sample,
    "Cancer types" = dge$sample$group,
    Platforms = dge$sample$platform
  )

  p <- ggplot(mds_data, aes(x = Dim1, y = Dim2, color =  `Cancer types`, shape = `Cell lines`)) +
    geom_point(size = 3, alpha = 0.6) + # Plot points +
    labs(title = "MDS plot", 
        x = paste0("Dimension 1 (", round(mds$var.explained[1]*100, 2), "%)" ) , 
        y = paste0("Dimension 2 (", round(mds$var.explained[2]*100, 2), "%)" )) +
    scale_color_viridis_d() + # Set color scale
    theme_minimal() 
  return(p)
}
```
## Gene count MDS
### Human
```{r}
mds <- norm_and_mds(bulk.gene.dge.merge)
p1 <- myMDS(bulk.gene.dge.merge, mds) + ggtitle("Human gene counts") +
    scale_shape_manual(values = c(7, 8, 9, 15, 16, 17, 18, 1))
p2 <- myMDS(bulk.gene.dge.merge, mds) + ggtitle("Human gene counts")  + 
            aes(color = Platforms, shape = `Cancer types`) + 
            scale_color_manual(values = color_palette[c("Illumina", "ONT_1", "ONT_2", "PacBio")] %>% unname())
p1 | p2 
```

## Transcript count MDS
### Human
```{r}
mds <- norm_and_mds(bulk.tx.dge.merge)
p1 <- myMDS(bulk.tx.dge.merge, mds) + ggtitle("Human transcript counts") + 
  scale_shape_manual(values = c(7, 8, 9, 15, 16, 17, 18, 1))

p2 <- myMDS(bulk.tx.dge.merge, mds) + ggtitle("Human transcript counts") + 
            aes(color = Platforms, shape = `Cancer types`) + 
            scale_color_manual(values = color_palette[c("Illumina", "ONT_1", "ONT_2", "PacBio")] %>% unname())

p1 | p2
```


### Heatmap （placeholder for potential plot）
```{r}

library(pheatmap)
dge <- bulk.gene.dge.merge
logCPM <- cpm(dge, log = TRUE)
colnames(logCPM) <- paste0(dge$samples$platform, "_", dge$samples$sample)
cor_matrix <- cor(logCPM, method = "pearson")

sample_annotations <- data.frame(
  Group = dge$samples$group,
  Platform = dge$samples$platform,
  Cellline = dge$samples$sample
)
rownames(sample_annotations) <- colnames(logCPM)
# Create the heatmap with sample annotations
pheatmap(cor_matrix, 
         cluster_rows = TRUE, 
         cluster_cols = TRUE, 
         color = colorRampPalette(c("navy", "white", "red"))(100), 
         display_numbers = TRUE,
         annotation_col = sample_annotations,
         main = "Pairwise Correlation Heatmap with Group Annotations")


```


```{r}

library(pheatmap)
dge <- bulk.tx.dge.merge
logCPM <- cpm(dge, log = TRUE)
colnames(logCPM) <- paste0(dge$samples$platform, "_", dge$samples$sample)
cor_matrix <- cor(logCPM, method = "pearson")

sample_annotations <- data.frame(Group = dge$samples$group, Platform = dge$samples$platform)
rownames(sample_annotations) <- colnames(logCPM)
# Create the heatmap with sample annotations
pheatmap(cor_matrix, 
         cluster_rows = TRUE, 
         cluster_cols = TRUE, 
         color = colorRampPalette(c("navy", "white", "red"))(100), 
         display_numbers = TRUE,
         annotation_col = sample_annotations,
         main = "Pairwise Correlation Heatmap with Group Annotations")

```

# Differential gene expression
## DGE edgeR (Salmon and Oarfish)
```{r}
DGE.edgeR_human <- function(dge, norm.method = "TMM") {
  # QC and normalization
  filter <- filterByExpr(dge, group = dge$samples$sample)
  dge <- dge[filter, , keep.lib.sizes = FALSE]
  
  if (!is.null(norm.method)) {
    dge <- calcNormFactors(dge, method = norm.method)
  }
  
  design <- model.matrix(~0 + group, data = dge$samples)
  dge <- estimateDisp(dge, design)
  qlfit <- glmQLFit(dge, design)
  
  contrasts <- makeContrasts(
    AvsL = groupsclc_a - groupluad,
    PvsL = groupsclc_p - groupluad,
    AvsP = groupsclc_a - groupsclc_p,
    levels = design
  )
  
  AvsL.QLF.test <- glmQLFTest(qlfit, contrast = contrasts[, "AvsL"])
  PvsL.QLF.test <- glmQLFTest(qlfit, contrast = contrasts[, "PvsL"])
  AvsP.QLF.test <- glmQLFTest(qlfit, contrast = contrasts[, "AvsP"])
  
  list(
    AvsL.DE = AvsL.QLF.test,
    PvsL.DE = PvsL.QLF.test,
    AvsP.DE = AvsP.QLF.test 
  )
}

ill_bulk.dge_human.list <- DGE.edgeR_human(ill_bulk.gene.dge[is.human(ill_bulk.gene.dge),])
ont_bulk.dge_human.list <- DGE.edgeR_human(ont_bulk.gene.dge[is.human(ont_bulk.gene.dge),])
drna_bulk.deg_human.list <- DGE.edgeR_human(drna_bulk.gene.dge[is.human(drna_bulk.gene.dge),])
pb_bulk.dge_human.list <- DGE.edgeR_human(pb_bulk.gene.dge[is.human(pb_bulk.gene.dge), ])
```

### Cross-platform comparison (upset)
```{r upset plot function}

complex_upset_plot <- function(lst) {
  # apply decideTests to each list element
  DEGs <- lapply(lst, function(x) {
    x <- x %>% decideTests()
    x <- rownames(x)[x != 0]
    return(x)
  })

  long_df <- stack(DEGs)

  # Make the binary membership matrix
  binary_matrix <- long_df %>%
    mutate(present = 1) %>%
    pivot_wider(names_from = ind, values_from = present, values_fill = 0)

  p <- ComplexUpset::upset(
    binary_matrix %>% select(-values) %>% as.data.frame(),
    intersect = names(lst),
    wrap = TRUE
  )

  return(list(
    plot = p,
    matrix = binary_matrix %>% rename(Gene = values)
  ))
}
```

#### DGE human 
```{r upset plot dge human, fig.width=8, fig.height=6}
# Upset plot
lst <- list(
  "Illumina" = ill_bulk.dge_human.list$AvsL.DE,
  "ONT cDNA" = ont_bulk.dge_human.list$AvsL.DE,
  "PacBio" = pb_bulk.dge_human.list$AvsL.DE,
  "ONT dRNA" = drna_bulk.deg_human.list$AvsL.DE
)
p1 <- complex_upset_plot(lst)$plot + ggtitle("SCLC_A vs LUAD") + theme(plot.title = element_text(size =20))
mat.AvsL.DE <- complex_upset_plot(lst)$matrix

lst <- list(
  "Illumina" = ill_bulk.dge_human.list$PvsL.DE,
  "ONT cDNA" = ont_bulk.dge_human.list$PvsL.DE,
  "PacBio" = pb_bulk.dge_human.list$PvsL.DE,
  "ONT dRNA" = drna_bulk.deg_human.list$PvsL.DE
)
p2 <- complex_upset_plot(lst)$plot + ggtitle("SCLC_P vs LUAD") + theme(plot.title = element_text(size = 20))
mat.PvsL.DE <- complex_upset_plot(lst)$matrix

lst <- list(
  "Illumina" = ill_bulk.dge_human.list$AvsP.DE,
  "ONT cDNA" = ont_bulk.dge_human.list$AvsP.DE,
  "PacBio" = pb_bulk.dge_human.list$AvsP.DE,
  "ONT dRNA" = drna_bulk.deg_human.list$AvsP.DE
)

p3 <- complex_upset_plot(lst)$plot + ggtitle("SCLC_A vs SCLC_P") + theme(plot.title = element_text(size = 20))
mat.AvsP.DE <- complex_upset_plot(lst)$matrix
(p1 / p2 / p3) + patchwork::plot_annotation(
  title = "DGE Human",
  theme = theme(plot.title = element_text(size =25))
)
```

## More inspection on DEG human
```{r Inspect DEG human}
# Getting the gene biotype and gene id, using illumina data here but will be identical for all datasets
tx_ref_df <- ill_bulk.dge$genes %>%
  rownames_to_column("tx_id") %>%
  select(tx_id, Length) %>%
  tibble() %>%
  mutate(
    biotype = strsplit(tx_id, "\\|") %>% sapply(function(x) x[8]),
    gene_id = strsplit(tx_id, "\\|") %>% sapply(function(x) x[2]),
    tx_id = strsplit(tx_id, "\\|") %>% sapply(function(x) x[1])
  )

gene_length_lookup <- tx_ref_df %>%
    group_by(gene_id) %>%
    summarise(mean_length = mean(Length, na.rm = TRUE), .groups = "drop")

# Connect to Ensembl
library(biomaRt)
ensembl <- useEnsembl(biomart = "genes", dataset = "hsapiens_gene_ensembl")

# SCLC_A vs LUAD
plot_d <- mat.AvsL.DE %>%
    left_join(gene_length_lookup, by = c("Gene" = "gene_id")) %>%
    rename(`Gene length` = mean_length)
  plot_d$Gene <- plot_d$Gene %>% sub("\\..*", "", .)
  biotypes <- getBM(
    attributes = c("ensembl_gene_id", "gene_biotype"),
    filters = "ensembl_gene_id",
    values = plot_d$Gene,
    mart = ensembl
  )

plot_d <- plot_d %>%
  left_join(biotypes, by = c("Gene" = "ensembl_gene_id")) %>%
  rename(biotype = gene_biotype) %>%  
  mutate(biotype = ifelse(biotype %in% 
            names(sort(table(biotype), decreasing = TRUE)[1:3]), 
            biotype, "Others"))
# set Others as the last level
plot_d$biotype <- factor(plot_d$biotype, levels = c(sort(unique(plot_d$biotype)) %>% setdiff("Others"), "Others"))


ComplexUpset::upset(
  plot_d,
  intersect = c('Illumina', "ONT cDNA", "PacBio", "ONT dRNA"),
  annotations = list(
        # 1st method - passing list:
        "Average transcript length" = list(
          aes = aes(x = intersection, y = `Gene length`),
          # provide a list if you wish to add several geoms
          geom = geom_boxplot(na.rm = TRUE, outliers = FALSE)
        ),
        "BioType" = list(
          aes = aes(x = intersection, fill = biotype),
          geom = geom_bar(position = "fill")
        )
  ),
  set_sizes=(
        ComplexUpset::upset_set_size(
            geom=geom_bar(
                aes(fill=biotype),
                width=0.8
            ) 
        )+ theme(legend.position = "none")
    ),
  wrap = TRUE
)
plot_d1 <- plot_d
```

```{r}
# SCLC_P vs LUAD
plot_d <- mat.PvsL.DE %>%
    left_join(gene_length_lookup, by = c("Gene" = "gene_id")) %>%
    rename(`Gene length` = mean_length)
  plot_d$Gene <- plot_d$Gene %>% sub("\\..*", "", .)
  biotypes <- getBM(
    attributes = c("ensembl_gene_id", "gene_biotype"),
    filters = "ensembl_gene_id",
    values = plot_d$Gene,
    mart = ensembl
  )

plot_d <- plot_d %>%
  left_join(biotypes, by = c("Gene" = "ensembl_gene_id")) %>%
  rename(biotype = gene_biotype) %>%  
  mutate(biotype = ifelse(biotype %in% 
            names(sort(table(biotype), decreasing = TRUE)[1:3]), 
            biotype, "Others"))
# set Others as the last level
plot_d$biotype <- factor(plot_d$biotype, levels = c(sort(unique(plot_d$biotype)) %>% setdiff("Others"), "Others"))
  
ComplexUpset::upset(
  plot_d,
  intersect = c('Illumina', "ONT cDNA", "PacBio", "ONT dRNA"),
  annotations = list(
        # 1st method - passing list:
        "Average transcript length" = list(
          aes = aes(x = intersection, y = `Gene length`),
          # provide a list if you wish to add several geoms
          geom = geom_boxplot(na.rm = TRUE, outliers = FALSE)
        ),
        "BioType" = list(
          aes = aes(x = intersection, fill = biotype),
          geom = geom_bar(position = "fill")
        )
  ),
  set_sizes=(
        ComplexUpset::upset_set_size(
            geom=geom_bar(
                aes(fill=biotype),
                width=0.8
            ) 
        )+ theme(legend.position = "none")
    ),
  wrap = TRUE
)
plot_d2 <- plot_d
```

```{r}
# SCLC_P vs LUAD
plot_d <- mat.AvsP.DE %>%
    left_join(gene_length_lookup, by = c("Gene" = "gene_id")) %>%
    rename(`Gene length` = mean_length)
  plot_d$Gene <- plot_d$Gene %>% sub("\\..*", "", .)
  biotypes <- getBM(
    attributes = c("ensembl_gene_id", "gene_biotype"),
    filters = "ensembl_gene_id",
    values = plot_d$Gene,
    mart = ensembl
  )

plot_d <- plot_d %>%
  left_join(biotypes, by = c("Gene" = "ensembl_gene_id")) %>%
  rename(biotype = gene_biotype) %>%
  mutate(biotype = ifelse(biotype %in%
    names(sort(table(biotype), decreasing = TRUE)[1:3]),
  biotype, "Others"
  ))
# set Others as the last level
plot_d$biotype <- factor(plot_d$biotype, levels = c(sort(unique(plot_d$biotype)) %>% setdiff("Others"), "Others"))

ComplexUpset::upset(
  plot_d,
  intersect = c('Illumina', "ONT cDNA", "PacBio", "ONT dRNA"),
  annotations = list(
        # 1st method - passing list:
        "Average transcript length" = list(
          aes = aes(x = intersection, y = `Gene length`),
          # provide a list if you wish to add several geoms
          geom = geom_boxplot(na.rm = TRUE, outliers = FALSE)
        ),
        "BioType" = list(
          aes = aes(x = intersection, fill = biotype),
          geom = geom_bar(position = "fill")
        )
  ),
  set_sizes=(
        ComplexUpset::upset_set_size(
            geom=geom_bar(
                aes(fill=biotype),
                width=0.8
            ) 
        )+ theme(legend.position = "none")
    ),
  wrap = TRUE
)
plot_d3 <- plot_d
```

### Clusterprofiler (Haven't finished)
```{r, include =F, eval=F, echo=F}
library(clusterProfiler)

# SCLC_A vs LUAD
gene_list <- plot_d1 %>%
  filter(`Illumina` == 1 & `ONT cDNA`==1 &  PacBio==1 &  `ONT dRNA`==1 ) %>%
  pull(Gene)
## GO enrich of DEG shared across all platforms
# gene_list <- bitr(gene_list, fromType = "ENSEMBL", toType = "ENTREZID", OrgDb = 'org.Hs.eg.db')
go_AvsL <- enrichGO(gene_list, "org.Hs.eg.db", pvalueCutoff = 0.05, keyType = "ENSEMBL", pAdjustMethod = "BH", qvalueCutoff = 0.05, ont="BP")
dotplot(go_AvsL, showCategory = 15)

## GO enrich of DEG from pacbio specifc
gene_list <- plot_d1 %>%
  filter(PacBio == 1 & `ONT cDNA` == 0 & Illumina == 0 & `ONT dRNA` == 0) %>%
  pull(Gene)
# gene_list <- bitr(gene_list, fromType = "ENSEMBL", toType = "ENTREZID", OrgDb = 'org.Hs.eg.db')
go_AvsL_pacbio <- enrichGO(gene_list, "org.Hs.eg.db", pvalueCutoff = 0.05, keyType = "ENSEMBL", pAdjustMethod = "BH", qvalueCutoff = 0.05, ont="BP")
dotplot(go_AvsL_pacbio, showCategory = 15)

# SCLC_P vs LUAD
gene_list <- plot_d2 %>%
  filter(`Illumina` == 1 & `ONT cDNA`==1 &  PacBio==1 &  `ONT dRNA`==1 ) %>%
  pull(Gene)
## GO enrich of DEG shared across all platforms
# gene_list <- bitr(gene_list, fromType = "ENSEMBL", toType = "ENTREZID", OrgDb = 'org.Hs.eg.db')
go_AvsL <- enrichGO(gene_list, "org.Hs.eg.db", pvalueCutoff = 0.05, keyType = "ENSEMBL", pAdjustMethod = "BH", qvalueCutoff = 0.05, ont="BP")
dotplot(go_AvsL, showCategory = 15)

## GO enrich of DEG from pacbio specifc
gene_list <- plot_d2 %>%
  filter(PacBio == 1 & `ONT cDNA` == 0 & Illumina == 0 & `ONT dRNA` == 0) %>%
  pull(Gene)
# gene_list <- bitr(gene_list, fromType = "ENSEMBL", toType = "ENTREZID", OrgDb = 'org.Hs.eg.db')
go_AvsL_pacbio <- enrichGO(gene_list, "org.Hs.eg.db", pvalueCutoff = 0.05, keyType = "ENSEMBL", pAdjustMethod = "BH", qvalueCutoff = 0.05, ont="BP")
dotplot(go_AvsL_pacbio, showCategory = 15)

```


# Differential transcript expression
```{r}
DTE.edgeR_human <- function(dge, norm.method="TMM") {

  if (!is.null(norm.method)) {
     dge <- calcNormFactors(dge, method = norm.method)
  }

  design <- model.matrix(~0+group, data = dge$samples)
  dge <- estimateDisp(dge, design)
  qlfit <- glmQLFit(dge, design)
  contrasts <- makeContrasts(AvsL=groupsclc_a-groupluad, PvsL=groupsclc_p-groupluad, AvsP = groupsclc_a-groupsclc_p, levels=design)
  # SCLC_A vs LUAD
  AvsL.QLF.test <- glmQLFTest(qlfit, contrast=contrasts[,"AvsL"]) 
  # SCLC_P vs LUAD
  PvsL.QLF.test <- glmQLFTest(qlfit, contrast=contrasts[,"PvsL"])
  # SCLC_A vs SCLC_P
  AvsP.QLF.test <- glmQLFTest(qlfit, contrast=contrasts[,"AvsP"]) 

 list(
    AvsL.DE = AvsL.QLF.test,
    PvsL.DE = PvsL.QLF.test,
    AvsP.DE = AvsP.QLF.test
  )
}
```

## DTE edgeR+scaling
```{r get dte scaling}
# scale
ill_bulk.scale_dge <- ill_bulk.dge
ont_bulk.scale_dge <- ont_bulk.dge.oarfish
pb_bulk.scale_dge <- pb_bulk.dge.oarfish
ont_drna.scale_dge <- ont_drna.dge.oarfish

ill_bulk.scale_dge$counts  <- ill_bulk.scale_dge$counts / ill_bulk.scale_dge$genes$Overdispersion
ont_bulk.scale_dge$counts  <- ont_bulk.scale_dge$counts / ont_bulk.scale_dge$genes$Overdispersion
pb_bulk.scale_dge$counts  <- pb_bulk.scale_dge$counts / pb_bulk.scale_dge$genes$Overdispersion
ont_drna.scale_dge$counts  <- ont_drna.scale_dge$counts / ont_drna.scale_dge$genes$Overdispersion

# get human list
ill_bulk.scl_dte_human.list <- DTE.edgeR_human(ill_bulk.scale_dge)
ont_bulk.scl_dte_human.list <- DTE.edgeR_human(ont_bulk.scale_dge)
pb_bulk.scl_dte_human.list <- DTE.edgeR_human(pb_bulk.scale_dge)
drna_bulk.scl_dte_human.list <- DTE.edgeR_human(ont_drna.scale_dge)
```

### Upset plot
```{r upset plot dte human edgeR with scaling, fig.width=8, fig.height=6}

# Upset plot
## SCLC_A vs LUAD (AvsL)
lst <- list(
  "Illumina" = ill_bulk.scl_dte_human.list$AvsL.DE,
  "ONT cDNA" = ont_bulk.scl_dte_human.list$AvsL.DE,
  "PacBio" = pb_bulk.scl_dte_human.list$AvsL.DE,
  "ONT dRNA" = drna_bulk.scl_dte_human.list$AvsL.DE
)
p1 <- complex_upset_plot(lst)$plot + ggtitle("SCLC_A vs LUAD") + theme(plot.title = element_text(size = 20))
mat.AvsL.DTE <- complex_upset_plot(lst)$matrix


lst <- list(
  "Illumina" = ill_bulk.scl_dte_human.list$PvsL.DE,
  "ONT cDNA" = ont_bulk.scl_dte_human.list$PvsL.DE,
  "PacBio" = pb_bulk.scl_dte_human.list$PvsL.DE,
  "ONT dRNA" = drna_bulk.scl_dte_human.list$PvsL.DE
)
p2 <- complex_upset_plot(lst)$plot + ggtitle("SCLC_P vs LUAD") + theme(plot.title = element_text(size = 20))
mat.PvsL.DTE <- complex_upset_plot(lst)$matrix

lst <- list(
  "Illumina" = ill_bulk.scl_dte_human.list$AvsP.DE,
  "ONT cDNA" = ont_bulk.scl_dte_human.list$AvsP.DE,
  "PacBio" = pb_bulk.scl_dte_human.list$AvsP.DE,
  "ONT dRNA" = drna_bulk.scl_dte_human.list$AvsP.DE
)

p3 <- complex_upset_plot(lst)$plot + ggtitle("SCLC_A vs SCLC_P") + theme(plot.title = element_text(size = 20))
mat.AvsP.DTE <- complex_upset_plot(lst)$matrix

p1 / p2 / p3 + patchwork::plot_annotation(
  title = "DTE Human (with scaling)",
  theme = theme(plot.title = element_text(size =25))
)
```

### Output the DEG and DTE results
```{r}
DE.rst <- list(
  gene_discovery = gene_discovery,
  txi_discovery = txi_discovery,
  AvsL.DTE.all = mat.AvsL.DTE %>% filter(Illumina == 1 & `ONT cDNA` == 1 & PacBio == 1 & `ONT dRNA` == 1) %>% .$Gene,
  PvsL.DTE.all = mat.PvsL.DTE %>% filter(Illumina == 1 & `ONT cDNA` == 1 & PacBio == 1 & `ONT dRNA` == 1) %>% .$Gene,
  AvsP.DTE.all = mat.AvsP.DTE %>% filter(Illumina == 1 & `ONT cDNA` == 1 & PacBio == 1 & `ONT dRNA` == 1) %>% .$Gene,
  AvsL.DEG.all = mat.AvsL.DE %>% filter(Illumina == 1 & `ONT cDNA` == 1 & PacBio==1 & `ONT dRNA`==1) %>% .$Gene,
  PvsL.DEG.all = mat.PvsL.DE %>% filter(Illumina == 1 & `ONT cDNA` == 1 & PacBio==1 & `ONT dRNA`==1) %>% .$Gene,
  AvsP.DEG.all = mat.AvsP.DE %>% filter(Illumina == 1 & `ONT cDNA` == 1 & PacBio==1 & `ONT dRNA`==1) %>% .$Gene,
  AvsL.DTE.any = mat.AvsL.DTE %>% filter(Illumina == 1 | `ONT cDNA` == 1 | PacBio == 1 | `ONT dRNA` == 1) %>% .$Gene,
  PvsL.DTE.any = mat.PvsL.DTE %>% filter(Illumina == 1 | `ONT cDNA` == 1 | PacBio == 1 | `ONT dRNA` == 1) %>% .$Gene,
  AvsP.DTE.any = mat.AvsP.DTE %>% filter(Illumina == 1 | `ONT cDNA` == 1 | PacBio == 1 | `ONT dRNA` == 1) %>% .$Gene,
  AvsL.DEG.any = mat.AvsL.DE %>% filter(Illumina == 1 | `ONT cDNA` == 1 | PacBio==1 | `ONT dRNA`==1) %>% .$Gene,
  PvsL.DEG.any = mat.PvsL.DE %>% filter(Illumina == 1 | `ONT cDNA` == 1 | PacBio==1 | `ONT dRNA`==1) %>% .$Gene,
  AvsP.DEG.any = mat.AvsP.DE %>% filter(Illumina == 1 | `ONT cDNA` == 1 | PacBio==1 | `ONT dRNA`==1) %>% .$Gene
)
saveRDS(DE.rst, file =params$out.rds)



myMDS(ill_bulk.gene.dge, norm_and_mds(ill_bulk.gene.dge)) + aes(label = `Cell lines`, shape = NULL) +
      geom_blank() +
      geom_text() +
      ggtitle("Illumina")
```
