---
title: "BCV analysis"
author: "Yupei"
version: "v1.0"
output:
  html_document:
    code_folding: hide
    toc: yes
    toc_float: yes
    number_sections: yes
    toc_depth: 4
  pdf_document:
    toc: yes
    toc_depth: '4'
params:
  random_seed: 2024
  cache_dir: NULL
  sirv_ercc_gtf: '/vast/projects/LongBench/reference_files/SIRV_Set4_Norm_Sequences_20210507/SIRV_ERCC_longSIRV_multi-fasta_20210507.gtf'
  sequins_gtf: '/vast/projects/LongBench/reference_files/Sequin_resources/v2.4/rnasequin_annotation_2.4.gtf'
  sequins_tsv: '/vast/projects/LongBench/reference_files/Sequin_resources/v2.4/rnasequin_isoforms_2.4.tsv'
  human_gtf: '/vast/projects/LongBench/reference_files/GRCh38/gencode.v44.annotation.gtf'
  sirv_csv: '/home/users/allstaff/you.yu/LongBench/reference_files/SIRV_Set4_Norm_Sequences_20210507/SIRV_set4_concentration.csv'
  ill_bulk_salmon_dir: "/vast/projects/LongBench//analysis/sr_bulk/result/salmon/salmon_quant"
  longbench1_ill: "/home/users/allstaff/you.yu/LongBench/ext_data/Xueyi_benchmarking"
  bulk_meta: "/vast/projects/LongBench/sequencing_data/illumina_bulk/metadata.txt"
  fig.path: NULL
---

# Setup: Create DGE from input
```{r setup, include=FALSE}
if (!is.null(params$cache_dir)) {
  knitr::opts_chunk$set(cache = TRUE, fig.width = 7, fig.height=5, warning = F, cache.path = params$cache_dir)
} else {
  knitr::opts_chunk$set(cache = FALSE, fig.width = 7, fig.height=5, warning = F)
}

knitr::opts_chunk$set(
  cache = !is.null(params$cache_dir),  # Enable caching only if cache_dir is provided
  fig.width = 15,                     # Set default figure width
  fig.height = 10,                    # Set default figure height
  warning = FALSE,                    # Suppress warnings in the output
  fig.path = params$fig.path,         # Set figure path (NULL if not provided)
  dev = if (!is.null(params$fig.path)) c("png", "svg") else NULL  # Set output devices if fig.path is provided
)

# Set cache.path only if params$cache_dir is not NULL
if (!is.null(params$cache_dir)) {
  knitr::opts_chunk$set(cache.path = params$cache_dir)
}
```

```{r Libraries,include=FALSE}
library(limma)
library(edgeR)
library(tidyverse)
library(RColorBrewer)
library(dplyr)
library(tidyr)
library(ggplot2)
library(ggrepel)
library(glue)
library(ggplot2)
library(gridExtra)
library(ggExtra)
library(knitr)
library(kableExtra)
library(rtracklayer)
library(scales)
library(UpSetR)
library(ggtree)
library(aplot)
library(viridis)
library(tximport)
source("/home/users/allstaff/you.yu/LongBench/analysis/workflow/scripts/Rfunctions.R")

# setup some conflict functions
rename <- dplyr::rename
select <- dplyr::select
```


```{r}
is.human <- function(dge) {grepl("^ENS", rownames(dge))}
human_tx  <- GenomicFeatures::makeTxDbFromGFF(params$human_gtf, format = "gtf") %>% GenomicFeatures::transcripts(columns = c("tx_id", "tx_name", "gene_id"))
human.G.Tx.map <- tibble(
  gene_id = human_tx$gene_id %>% unlist,
  tx_name = human_tx$tx_name
)
rm(human_tx)
combined_tx2gene <- human.G.Tx.map[, c("tx_name", "gene_id")]

bulk.meta <- read.csv(params$bulk_meta)
rownames(bulk.meta) <- bulk.meta$sample
```


```{r}
# Define file names
files <- paste(params$longbench1_ill, c(
    "GSM5255695_R1_H1975_Illumina.txt", "GSM5255696_R2_H1975_Illumina.txt", "GSM5255697_R3_H1975_Illumina.txt",
    "GSM5255704_R1_HCC827_Illumina.txt", "GSM5255705_R2_HCC827_Illumina.txt", "GSM5255706_R3_HCC827_Illumina.txt"
)           , sep = "/")

sample_ids <- sub("^.*/(GSM\\d+)_(R[1-3]_(H1975|HCC827)).*$", "\\2", files)

# Read data into a list, skipping the first line and ignoring empty lines
count_list <- lapply(files, function(f) {
  read.table(f, header = FALSE, row.names = 1, col.names = c("Gene", f), skip = 1, blank.lines.skip = TRUE)
})

# Merge all count tables by gene ID
common_genes <- Reduce(intersect, lapply(count_list, rownames))

counts <- do.call(cbind, lapply(count_list, function(x) x[common_genes, ]))
rownames(counts) <- common_genes
colnames(counts) <- sample_ids
# Convert to matrix
counts <- as.matrix(counts)

# Define sample metadata
samples <- data.frame(
  row.names = colnames(counts),
  group = rep(c("H1975", "HCC827"), each = 3)
)

# Create DGEList
# Define directories with their corresponding expected file names


# Apply function to each directory to get all quant file paths

dir_info <- list(dir = params$ill_bulk_salmon_dir, file_name = "quant.sf")
dir <- dir_info$dir
file_name <- dir_info$file_name

# List all subdirectories and get the quant file paths
sub_dirs <- list.dirs(dir, full.names = TRUE, recursive = FALSE)
quant_paths <- file.path(sub_dirs, file_name)


# Process each quant path for different bulk types
ill_bulk.gene.dge <- get_dge_from_txi(quant_paths, ".*/", "salmon")

dge1 <- DGEList(counts = counts, group = samples$group)
dge2 <- get_dge_from_txi(quant_paths, ".*/", "salmon")
#overlap the genes before filtering
common_genes <- intersect(dge1 %>% rownames, dge2 %>% rownames)
dge1 <- dge1[common_genes,]
dge2 <- dge2[common_genes,]

# Dong et al
dge1 <- dge1[filterByExpr(dge1) & is.human(dge1),] %>% 
                calcNormFactors() %>% 
                estimateDisp()
# LongBench2
dge2 <- dge2[filterByExpr(dge2) & is.human(dge2),] %>% 
                calcNormFactors() %>% 
                estimateDisp()
```


```{r, fig.height=4, fig.width=4, include=F, eval=F, echo=F}
plotBCV(dge1[,dge1$samples$group == "H1975"])
plotBCV(dge1[,dge1$samples$group == "HCC827"])
plotBCV(dge2[,dge2$samples$group == "luad"])
plotBCV(dge2[,dge2$samples$group == "sclc_a"])
plotBCV(dge2[,dge2$samples$group == "sclc_p"])
```

```{r BCV Dong et al, fig.height=4, fig.width=4}
dispersion <- data.frame(
  AveLogCPM = dge1$AveLogCPM,
  common.dispersion = dge1$common.dispersion,
  trended.dispersion = dge1$trended.dispersion,
  tagwise.dispersion = dge1$tagwise.dispersion
)

rownames(dispersion) <- rownames(dge1)
dispersion <- dispersion[order(dispersion$AveLogCPM), ]

plot.bcv.dong <- ggplot() +
  stat_binhex(data = dispersion[grep("^ENSG", rownames(dispersion)), ],
              aes(x = AveLogCPM, y = sqrt(tagwise.dispersion)), bins = 100) +
  scale_fill_viridis(option = "magma") +
  geom_line(data = dispersion[grep("^ENSG", rownames(dispersion)), ],
            aes(x = AveLogCPM, y = sqrt(common.dispersion), colour = "Common")) +
  geom_line(data = dispersion[grep("^ENSG", rownames(dispersion)), ],
            aes(x = AveLogCPM, y = sqrt(trended.dispersion), colour = "Trend")) +
  scale_colour_manual(name = "BCV",
                      breaks = c("Common", "Trend"),
                      values = c("Common" = "red", "Trend" = "blue")) +
  scale_y_continuous(limits = c(0, 0.75)) +
  theme_bw() +
  labs(x = "Average log CPM", y = "Biological coefficient of variation",
       fill = "# of Genes",
       title = "Dong et al.") +
  theme(text = element_text(size = 16))


plot.bcv.dong
```

```{r BCV LongBench, fig.height=4, fig.width=4}
dispersion <- data.frame(
  AveLogCPM = dge2$AveLogCPM,
  common.dispersion = dge2$common.dispersion,
  trended.dispersion = dge2$trended.dispersion,
  tagwise.dispersion = dge2$tagwise.dispersion
)

rownames(dispersion) <- rownames(dge2)
dispersion <- dispersion[order(dispersion$AveLogCPM), ]

plot.bcv.longbench <- ggplot() +
  stat_binhex(data = dispersion[grep("^ENSG", rownames(dispersion)), ],
              aes(x = AveLogCPM, y = sqrt(tagwise.dispersion)), bins = 100) +
  scale_fill_viridis(option = "magma") +
  geom_line(data = dispersion[grep("^ENSG", rownames(dispersion)), ],
            aes(x = AveLogCPM, y = sqrt(common.dispersion), colour = "Common")) +
  geom_line(data = dispersion[grep("^ENSG", rownames(dispersion)), ],
            aes(x = AveLogCPM, y = sqrt(trended.dispersion), colour = "Trend")) +
  # geom_point(data = dispersion[grep("^R", rownames(dispersion)), ],
  #            aes(x = AveLogCPM, y = sqrt(tagwise.dispersion),
  #                colour = "Tagwise_sequins")) +
  scale_colour_manual(name = "BCV",
                      breaks = c("Common", "Trend"),
                      values = c("Common" = "red", "Trend" = "blue")) +
  scale_y_continuous(limits = c(0, 3.5)) +
  theme_bw() +
  labs(x = "Average log CPM", y = "Biological coefficient of variation",
       fill = "# of Genes",
       title = "Longbench") +
  theme(text = element_text(size = 16))


plot.bcv.longbench
```

```{r LongBench individual cell line BCV, fig.height=10, fig.width=10}
individual_bcv <- function(cell_line){
  dge <-dge2[,dge2$samples$group==cell_line] %>% 
                calcNormFactors() %>% 
                estimateDisp()
  dispersion <- data.frame(
  AveLogCPM = dge$AveLogCPM,
  common.dispersion = dge$common.dispersion,
  trended.dispersion = dge$trended.dispersion,
  tagwise.dispersion = dge$tagwise.dispersion
  )
  
  rownames(dispersion) <- rownames(dge)
  dispersion <- dispersion[order(dispersion$AveLogCPM), ]
  
  plot.bcv <- ggplot() +
    stat_binhex(data = dispersion[grep("^ENSG", rownames(dispersion)), ],
                aes(x = AveLogCPM, y = sqrt(tagwise.dispersion)), bins = 100) +
    scale_fill_viridis(option = "magma") +
    geom_line(data = dispersion[grep("^ENSG", rownames(dispersion)), ],
              aes(x = AveLogCPM, y = sqrt(common.dispersion), colour = "Common")) +
    geom_line(data = dispersion[grep("^ENSG", rownames(dispersion)), ],
              aes(x = AveLogCPM, y = sqrt(trended.dispersion), colour = "Trend")) +
    # geom_point(data = dispersion[grep("^R", rownames(dispersion)), ],
    #            aes(x = AveLogCPM, y = sqrt(tagwise.dispersion),
    #                colour = "Tagwise_sequins")) +
    scale_colour_manual(name = "BCV",
                        breaks = c("Common", "Trend", "Tagwise_sequins"),
                        values = c("Common" = "red", "Trend" = "blue", "Tagwise_sequins" = "green")) +
    # scale_y_continuous(limits = c(0, 0.75)) +
    theme_bw() +
    labs(x = "Average log CPM", y = "Biological coefficient of variation",
         fill = "Tagwise BCV\ncount",
         title = cell_line) +
    theme(text = element_text(size = 16))
  
  
  plot.bcv
}

lapply(dge2$samples$group %>% unique, individual_bcv) %>% patchwork::wrap_plots(ncol=2)
```


```{r MDS longbench, fig.height=4, fig.width=4}
mds <- plotMDS(dge1, plot = FALSE)

mds_data <- data.frame(
    Dim1 = mds$x,
    Dim2 = mds$y,
    cell.line = dge1$sample$group
  )

ggplot(mds_data, aes(x = Dim1, y = Dim2, color = cell.line)) +
    geom_point(size = 3, alpha = 0.6) + # Plot points
    labs(title = "MDS plot", 
        x = paste0("Dimension 1 (", round(mds$var.explained[1]*100, 2), "%)" ) , 
        y = paste0("Dimension 2 (", round(mds$var.explained[2]*100, 2), "%)" )) +
    scale_color_viridis_d() + # Set color scale
    theme_minimal() 



mds <- plotMDS(dge2, plot = FALSE)

mds_data <- data.frame(
    Dim1 = mds$x,
    Dim2 = mds$y,
    cell.line = dge2$sample$sample,
    Cancer.type = dge2$sample$group
  )

ggplot(mds_data, aes(x = Dim1, y = Dim2, color = Cancer.type, label=cell.line)) +
    geom_point(size = 3, alpha = 0.6) + # Plot points
    geom_text_repel(size = 3) +
    labs(title = "MDS plot", 
        x = paste0("Dimension 1 (", round(mds$var.explained[1]*100, 2), "%)" ) , 
        y = paste0("Dimension 2 (", round(mds$var.explained[2]*100, 2), "%)" )) +
    scale_color_viridis_d() + # Set color scale
    theme_minimal() 
```

