---
title: "DE analysis"
author: "Yupei & Ash"
output:
  html_document:
    code_folding: hide
    toc: yes
    toc_float: yes
    number_sections: yes
    toc_depth: 4
  pdf_document:
    toc: yes
    toc_depth: '4'
params:
  random_seed: 2024
  cache_dir: NULL
  sirv_ercc_gtf: '/vast/projects/LongBench/reference_files/SIRV_Set4_Norm_Sequences_20210507/SIRV_ERCC_longSIRV_multi-fasta_20210507.gtf'
  sequins_gtf: '/vast/projects/LongBench/reference_files/Sequin_resources/v2.4/rnasequin_annotation_2.4.gtf'
  sequins_tsv: '/vast/projects/LongBench/reference_files/Sequin_resources/v2.4/rnasequin_isoforms_2.4.tsv'
  human_gtf: '/vast/projects/LongBench/reference_files/GRCh38/gencode.v44.annotation.gtf'
  ont_bulk_salmon_dir: "/vast/projects/LongBench/analysis/lr_bulk/result/salmon_output/ont_bulk"
  pb_bulk_salmon_dir: "/vast/projects/LongBench/analysis/lr_bulk/result/salmon_output/pb_bulk"
  ont_sc_clean_salmon_dir: "/vast/projects/LongBench/analysis/lr_sc_sn/result/PseudoBulkSalmon/ont_sc_clean"
  ont_sn_clean_salmon_dir: "/vast/projects/LongBench/analysis/lr_sc_sn/result/PseudoBulkSalmon/ont_sn_clean"
  ont_sc_salmon_dir: "/vast/projects/LongBench/analysis/lr_sc_sn/result/PseudoBulkSalmon/ont_sc"
  ont_sn_salmon_dir: "/vast/projects/LongBench/analysis/lr_sc_sn/result/PseudoBulkSalmon/ont_sn"
  ill_bulk_salmon_dir: "/vast/projects/LongBench/analysis/illumina_bulk/tx_de_analysis/salmon_bs"
  bulk_meta: '/vast/projects/LongBench/sequencing_data/illumina_bulk/metadata.txt'
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(cache = TRUE, fig.width = 20, fig.height=15, warning = F)
```

```{r, include=FALSE, eval=F}
params <- list(
  random_seed = 2024,
  cache_dir = NULL,
  sirv_ercc_gtf = '/vast/projects/LongBench/reference_files/SIRV_Set4_Norm_Sequences_20210507/SIRV_ERCC_longSIRV_multi-fasta_20210507.gtf',
  sequins_gtf = '/vast/projects/LongBench/reference_files/Sequin_resources/v2.4/rnasequin_annotation_2.4.gtf',
  sequins_tsv = '/vast/projects/LongBench/reference_files/Sequin_resources/v2.4/rnasequin_isoforms_2.4.tsv',
  human_gtf = '/vast/projects/LongBench/reference_files/GRCh38/gencode.v44.annotation.gtf',
  ont_bulk_salmon_dir = "/vast/projects/LongBench/analysis/lr_bulk/result/salmon_output/ont_bulk",
  pb_bulk_salmon_dir = "/vast/projects/LongBench/analysis/lr_bulk/result/salmon_output/pb_bulk",
  ont_sc_clean_salmon_dir = "/vast/projects/LongBench/analysis/lr_sc_sn/result/PseudoBulkSalmon/ont_sc_clean",
  ont_sn_clean_salmon_dir = "/vast/projects/LongBench/analysis/lr_sc_sn/result/PseudoBulkSalmon/ont_sn_clean",
  ont_sc_salmon_dir = "/vast/projects/LongBench/analysis/lr_sc_sn/result/PseudoBulkSalmon/ont_sc",
  ont_sn_salmon_dir = "/vast/projects/LongBench/analysis/lr_sc_sn/result/PseudoBulkSalmon/ont_sn",
  ill_bulk_salmon_dir = "/vast/projects/LongBench/analysis/illumina_bulk/tx_de_analysis/salmon_bs",
  bulk_meta = '/vast/projects/LongBench/sequencing_data/illumina_bulk/metadata.txt'
)
```


```{r}
library(limma)
library(edgeR)
library(tidyverse)
library(RColorBrewer)
library(dplyr)
library(ggplot2)
library(ggrepel)
library(glue)
library(ggplot2)
library(gridExtra)
library(ggExtra)
library(knitr)
library(kableExtra)
library(rtracklayer)
library(scales)
library(UpSetR)
```

# Read the data
```{r}
bulk.meta <- read.csv(params$bulk_meta)
rownames(bulk.meta) <- bulk.meta$sample


get_dge_from_salmon <- function(dir, sample_prefix_regex) {
  salmon.dirs <- file.path(dir, 
                               list.dirs(dir, full.names = FALSE, recursive = FALSE))
  catched.salmon <- edgeR::catchSalmon(salmon.dirs, verbose = TRUE)
  rst.dge <- DGEList(counts=catched.salmon$counts, gene=catched.salmon$annotation)
  rownames(rst.dge$samples) <- rownames(rst.dge$samples) %>% sub(sample_prefix_regex, '', .)
  colnames(rst.dge$counts) <- colnames(rst.dge$counts) %>% sub(sample_prefix_regex, '', .)
  rst.dge$samples <- merge(rst.dge$samples %>% select(-group), bulk.meta, by = "row.names", all.x = TRUE) %>% select(-Row.names)
  rst.dge$samples <- rst.dge$samples[match(colnames(rst.dge$counts), rst.dge$samples$sample),]
  return(rst.dge)
}

ill_bulk.dge <- get_dge_from_salmon(params$ill_bulk_salmon_dir, ".*_S")
ont_bulk.dge <- get_dge_from_salmon(params$ont_bulk_salmon_dir, ".*/")
pb_bulk.dge <- get_dge_from_salmon(params$pb_bulk_salmon_dir, ".*/")
ont_sc.dge <- get_dge_from_salmon(params$ont_sc_salmon_dir, ".*/")
ont_sn.dge <- get_dge_from_salmon(params$ont_sn_salmon_dir, ".*/")
ont_sc_clean.dge <- get_dge_from_salmon(params$ont_sc_clean_salmon_dir, ".*/")
ont_sn_clean.dge <- get_dge_from_salmon(params$ont_sn_clean_salmon_dir, ".*/")



# Bulk ill
salmon.ill_bulk.dirs <- file.path(params$ill_bulk_salmon_dir,
                               list.dirs(params$ill_bulk_salmon_dir, full.names = FALSE, recursive = FALSE))
ill_bulk.salmon <- edgeR::catchSalmon(salmon.ill_bulk.dirs, verbose = TRUE)
ill_bulk.dge <- DGEList(counts=ill_bulk.salmon$counts, gene=ill_bulk.salmon$annotation)
rownames(ill_bulk.dge$samples) <- rownames(ill_bulk.dge$samples) %>% sub(".*_S", '', .)
rownames(ill_bulk.dge$samples) <- bulk.meta$sample[match(bulk.meta$index, rownames(ill_bulk.dge$samples))]
colnames(ill_bulk.dge$counts) <- colnames(ill_bulk.dge$counts) %>% sub(".*_S", '', .)
colnames(ill_bulk.dge$counts) <- bulk.meta$sample[match(bulk.meta$index, colnames(ill_bulk.dge$counts))]
ill_bulk.dge$samples <- merge(ill_bulk.dge$samples %>% select(-group), bulk.meta, by = "row.names", all.x = TRUE) %>% select(-Row.names)
ill_bulk.dge$samples <- ill_bulk.dge$samples[match(colnames(ill_bulk.dge$counts), ill_bulk.dge$samples$sample),]

```


# Split the origin
```{r}
is.sequins <- function(dge) {grepl("^R", rownames(dge))}
is.human <- function(dge) {grepl("^ENS", rownames(dge))}
is.sirv <- function(dge) {!is.human(dge) & !is.sequins(dge)}
is.sirv_E0 <- function(dge) {grepl("^SIRV\\d{3}$", rownames(dge))}
```


# Raw count distribution
```{r, fig.width=10, fig.height=5}

plot_df <- data.frame(
  ill_bulk = ill_bulk.dge[is.human(ill_bulk.dge), ]$counts %>% sum,
  ont_bulk = ont_bulk.dge[is.human(ont_bulk.dge), ]$counts %>% sum,
  pb_bulk = pb_bulk.dge[is.human(pb_bulk.dge), ]$counts %>% sum,
  ont_sc = ont_sc.dge[is.human(ont_sc.dge), ]$counts %>% sum,
  ont_sn = ont_sn.dge[is.human(ont_sn.dge), ]$counts %>% sum,
  ont_sn_clean = ont_sn_clean.dge[is.human(ont_sn_clean.dge), ]$counts %>% sum,
  ont_sc_clean = ont_sc_clean.dge[is.human(ont_sc_clean.dge), ]$counts %>% sum
) %>% pivot_longer(everything(), names_to = "Dataset", values_to = "Total Count")

p1 <- ggplot(plot_df, aes(x = Dataset, y = `Total Count`, fill = Dataset)) +
  geom_bar(stat = "identity", position = position_dodge()) +
  theme_minimal() +
  labs(title = "Total counts (HUMAN transcript)", x = "Sample", y = "Total Count") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

plot_df <- data.frame(
  ill_bulk = ill_bulk.dge[is.sequins(ill_bulk.dge), ]$counts %>% sum,
  ont_bulk = ont_bulk.dge[is.sequins(ont_bulk.dge), ]$counts %>% sum,
  pb_bulk = pb_bulk.dge[is.sequins(pb_bulk.dge), ]$counts %>% sum,
  ont_sc = ont_sc.dge[is.sequins(ont_sc.dge), ]$counts %>% sum,
  ont_sn = ont_sn.dge[is.sequins(ont_sn.dge), ]$counts %>% sum,
  ont_sn_clean = ont_sn_clean.dge[is.sequins(ont_sn_clean.dge), ]$counts %>% sum,
  ont_sc_clean = ont_sc_clean.dge[is.sequins(ont_sc_clean.dge), ]$counts %>% sum
) %>% pivot_longer(everything(), names_to = "Dataset", values_to = "Total Count")

p2 <- ggplot(plot_df, aes(x = Dataset, y = `Total Count`, fill = Dataset)) +
  geom_bar(stat = "identity", position = position_dodge()) +
  theme_minimal() +
  labs(title = "Total counts (Sequins transcript)", x = "Sample", y = "Total Count")+
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

# proportion of sequins counts
plot_df <- data.frame(
  ill_bulk = ill_bulk.dge[is.sequins(ill_bulk.dge), ]$counts %>% sum / ill_bulk.dge$counts %>% sum,
  ont_bulk = ont_bulk.dge[is.sequins(ont_bulk.dge), ]$counts %>% sum / ont_bulk.dge$counts %>% sum,
  pb_bulk = pb_bulk.dge[is.sequins(pb_bulk.dge), ]$counts %>% sum / pb_bulk.dge$counts %>% sum,
  ont_sc = ont_sc.dge[is.sequins(ont_sc.dge), ]$counts %>% sum / ont_sc.dge$counts %>% sum,
  ont_sn = ont_sn.dge[is.sequins(ont_sn.dge), ]$counts %>% sum / ont_sn.dge$counts %>% sum,
  ont_sn_clean = ont_sn_clean.dge[is.sequins(ont_sn_clean.dge), ]$counts %>% sum / ont_sn_clean.dge$counts %>% sum,
  ont_sc_clean = ont_sc_clean.dge[is.sequins(ont_sc_clean.dge), ]$counts %>% sum / ont_sc_clean.dge$counts %>% sum
) %>% pivot_longer(everything(), names_to = "Dataset", values_to = "Proportion of Sequins Count")

p3 <- ggplot(plot_df, aes(x = Dataset, y = `Proportion of Sequins Count`, fill = Dataset)) +
  geom_bar(stat = "identity", position = position_dodge()) +
  theme_minimal() +
  labs(title = "Proportion of Sequins counts", x = "Sample", y = "Proportion of Sequins Count") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))


plot_df <- data.frame(
  ill_bulk = ill_bulk.dge[is.sirv(ill_bulk.dge), ]$counts %>% sum,
  ont_bulk = ont_bulk.dge[is.sirv(ont_bulk.dge), ]$counts %>% sum,
  pb_bulk = pb_bulk.dge[is.sirv(pb_bulk.dge), ]$counts %>% sum,
  ont_sc = ont_sc.dge[is.sirv(ont_sc.dge), ]$counts %>% sum,
  ont_sn = ont_sn.dge[is.sirv(ont_sn.dge), ]$counts %>% sum,
  ont_sn_clean = ont_sn_clean.dge[is.sirv(ont_sn_clean.dge), ]$counts %>% sum,
  ont_sc_clean = ont_sc_clean.dge[is.sirv(ont_sc_clean.dge), ]$counts %>% sum
) %>% pivot_longer(everything(), names_to = "Dataset", values_to = "Total Count")

p4 <- ggplot(plot_df, aes(x = Dataset, y = `Total Count`, fill = Dataset)) +
  geom_bar(stat = "identity", position = position_dodge()) +
  theme_minimal() +
  labs(title = "Total counts (SIRV transcript)", x = "Sample", y = "Total Count")+
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

# proportion of sirv counts
plot_df <- data.frame(
  ill_bulk = ill_bulk.dge[is.sirv(ill_bulk.dge), ]$counts %>% sum / ill_bulk.dge$counts %>% sum,
  ont_bulk = ont_bulk.dge[is.sirv(ont_bulk.dge), ]$counts %>% sum / ont_bulk.dge$counts %>% sum,
  pb_bulk = pb_bulk.dge[is.sirv(pb_bulk.dge), ]$counts %>% sum / pb_bulk.dge$counts %>% sum,
  ont_sc = ont_sc.dge[is.sirv(ont_sc.dge), ]$counts %>% sum / ont_sc.dge$counts %>% sum,
  ont_sn = ont_sn.dge[is.sirv(ont_sn.dge), ]$counts %>% sum / ont_sn.dge$counts %>% sum,
  ont_sn_clean = ont_sn_clean.dge[is.sirv(ont_sn_clean.dge), ]$counts %>% sum / ont_sn_clean.dge$counts %>% sum,
  ont_sc_clean = ont_sc_clean.dge[is.sirv(ont_sc_clean.dge), ]$counts %>% sum / ont_sc_clean.dge$counts %>% sum
) %>% pivot_longer(everything(), names_to = "Dataset", values_to = "Proportion of SIRV Count")

p5 <- ggplot(plot_df, aes(x = Dataset, y = `Proportion of SIRV Count`, fill = Dataset)) +
  geom_bar(stat = "identity", position = position_dodge()) +
  theme_minimal() +
  labs(title = "Proportion of SIRV counts", x = "Sample", y = "Proportion of SIRV Count") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))


  

grid.arrange(p1, p2, p3, p4, p5, ncol = 3)
```

# Overdispesion
## Stratify by transcript number
```{r, functions}
calcTxNum <- function(G.Tx.map){
  geneid = rep(NA, nrow(G.Tx.map))
  txcount = table(G.Tx.map$gene_id)
  txnum = as.numeric(txcount[match(G.Tx.map$gene_id, names(txcount))])
}
# SIRV
SIRV_tx <- GenomicFeatures::makeTxDbFromGFF(params$sirv_ercc_gtf, format = "gtf") %>% GenomicFeatures::transcripts(columns = c("tx_id", "tx_name", "gene_id"))
SIRV.G.Tx.map <- tibble(
  gene_id = SIRV_tx$gene_id %>% unlist,
  tx_name = SIRV_tx$tx_name
)
rm(SIRV_tx)
SIRV.G.Tx.map$tx_count <- SIRV.G.Tx.map %>% calcTxNum

# Sequins
sequins_tx  <- GenomicFeatures::makeTxDbFromGFF(params$sequins_gtf, format = "gtf") %>% GenomicFeatures::transcripts(columns = c("tx_id", "tx_name", "gene_id"))
sequins.G.Tx.map <- tibble(
  gene_id = sequins_tx$gene_id %>% unlist,
  tx_name = sequins_tx$tx_name
)
rm(sequins_tx)
sequins.G.Tx.map$tx_count <- sequins.G.Tx.map %>% calcTxNum

# Human
human_tx  <- GenomicFeatures::makeTxDbFromGFF(params$human_gtf, format = "gtf") %>% GenomicFeatures::transcripts(columns = c("tx_id", "tx_name", "gene_id"))
human.G.Tx.map <- tibble(
  gene_id = human_tx$gene_id %>% unlist,
  tx_name = human_tx$tx_name
)
rm(human_tx)
human.G.Tx.map$tx_count <- human.G.Tx.map %>% calcTxNum
```


```{r, Overdispersion}
# Make boxplot of the overdispersion, stratified by transcript number, split figure by human, sequins and sirv
## human

ill_bulk.dge.human <-  ill_bulk.dge[is.human(ill_bulk.dge),]
ont_bulk.dge.human <- ont_bulk.dge[is.human(ont_bulk.dge),]
pb_bulk.dge.human <- pb_bulk.dge[is.human(pb_bulk.dge),]
ont_sc.dge.human <- ont_sc_clean.dge[is.human(ont_sc_clean.dge),]
ont_sn.dge.human <- ont_sn_clean.dge[is.human(ont_sn_clean.dge),]

overdisp.human <- data.frame(
  ont_bulk = ont_bulk.dge.human$genes$Overdispersion,
  ill_bulk = ill_bulk.dge.human$genes$Overdispersion[match(rownames(ont_bulk.dge.human), rownames(ill_bulk.dge.human))],
  pb_bulk = pb_bulk.dge.human$genes$Overdispersion[match(rownames(ont_bulk.dge.human), rownames(pb_bulk.dge.human))],
  ont_sc = ont_sc.dge.human$genes$Overdispersion[match(rownames(ont_bulk.dge.human), rownames(ont_sc.dge.human))],
  ont_sn = ont_sn.dge.human$genes$Overdispersion[match(rownames(ont_bulk.dge.human), rownames(ont_sn.dge.human))],
  tx_count = human.G.Tx.map$tx_count[match(rownames(ont_bulk.dge.human) %>% sub("\\|.*", "", .), human.G.Tx.map$tx_name)],
  tx_len =  ont_bulk.dge.human$genes$Length

)

overdisp.human$tx_count_bin <- cut(overdisp.human$tx_count, breaks = c(0,1,5,10,20,40, Inf))
overdisp.human$tx_count_bin <- factor(overdisp.human$tx_count_bin, ordered = TRUE)

overdisp.human$tx_len_bin <- cut(overdisp.human$tx_len, breaks = c(0,500,1500,2000,2500,3000, Inf))
overdisp.human$tx_len_bin <- factor(overdisp.human$tx_len_bin, ordered = TRUE)

overdisp.human_pivot <- pivot_longer(overdisp.human, 
                            cols = c(ont_bulk, pb_bulk, ill_bulk), 
                            names_to = "Method", values_to = "Overdispersion")

p1.1 <- ggplot(overdisp.human_pivot, aes(x = tx_count_bin, y = Overdispersion, fill = Method)) +
  geom_boxplot(position = position_dodge(width = 0.75), outliers=FALSE) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  labs(title = "Human transcript assignment ambiguity (Salmon overdispersion)", x = "Transcript number", y = "Overdispersion")


overdisp.human_pivot <- pivot_longer(overdisp.human, 
                            cols = c(ont_bulk, pb_bulk), 
                            names_to = "Method", values_to = "Overdispersion")

p1.2 <- ggplot(overdisp.human_pivot, aes(x = tx_count_bin, y = Overdispersion, fill = Method)) +
  geom_boxplot(position = position_dodge(width = 0.75), outliers=FALSE) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  labs(title = "Human transcript assignment ambiguity (Salmon overdispersion)", x = "Transcript number", y = "Overdispersion")
  


# ## sequins
ont_bulk.dge.sequins <- ont_bulk.dge[is.sequins(ont_bulk.dge),]
pb_bulk.dge.sequins <- pb_bulk.dge[is.sequins(pb_bulk.dge),]
ill_bulk.dge.sequins <- ill_bulk.dge[is.sequins(ill_bulk.dge),]

overdisp.sequins <- data.frame(
  ont_bulk = ont_bulk.dge.sequins$genes$Overdispersion,
  ill_bulk = ill_bulk.dge.sequins$genes$Overdispersion[match(rownames(ont_bulk.dge.sequins), rownames(ill_bulk.dge.sequins))],
  pb_bulk = pb_bulk.dge.sequins$genes$Overdispersion[match(rownames(ont_bulk.dge.sequins), rownames(pb_bulk.dge.sequins))],
  tx_count = sequins.G.Tx.map$tx_count[match(rownames(ont_bulk.dge.sequins), sequins.G.Tx.map$tx_name)]
)

overdisp.sequins$tx_count_bin <- cut(overdisp.sequins$tx_count, breaks = c(0,1,2,3,4,5, Inf))
overdisp.sequins$tx_count_bin <- factor(overdisp.sequins$tx_count_bin, ordered = TRUE)
overdisp.sequins_pivot <- pivot_longer(overdisp.sequins, 
                                        cols = c(ont_bulk, ill_bulk, pb_bulk), 
                                        names_to = "Method", values_to = "Overdispersion")

p2 <- ggplot(overdisp.sequins_pivot, aes(x = tx_count_bin, y = Overdispersion, fill = Method)) +
  geom_boxplot(position = position_dodge(width = 0.75), outliers=FALSE) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  labs(title = "Sequins assignment ambiguity (Salmon overdispersion)", x = "Transcript number", y = "Overdispersion")

## sirv
ont_bulk.dge.sirv <- ont_bulk.dge[is.sirv(ont_bulk.dge),]
ill_bulk.dge.sirv <- ill_bulk.dge[is.sirv(ill_bulk.dge),]
pb_bulk.dge.sirv <- pb_bulk.dge[is.sirv(pb_bulk.dge),]

overdisp.sirv <- data.frame(
  ont_bulk = ont_bulk.dge.sirv$genes$Overdispersion,
  ill_bulk = ill_bulk.dge.sirv$genes$Overdispersion[match(rownames(ill_bulk.dge.sirv), rownames(ill_bulk.dge.sirv))],
  pb_bulk = pb_bulk.dge.sirv$genes$Overdispersion[match(rownames(pb_bulk.dge.sirv), rownames(pb_bulk.dge.sirv))],
  tx_count = SIRV.G.Tx.map$tx_count[match(rownames(ont_bulk.dge.sirv), SIRV.G.Tx.map$tx_name)]
)

overdisp.sirv$tx_count_bin <- cut(overdisp.sirv$tx_count, breaks = c(0,1,2,3,4,5, 10,15,Inf))
overdisp.sirv$tx_count_bin <- factor(overdisp.sirv$tx_count_bin, ordered = TRUE)
overdisp.sirv_pivot <- pivot_longer(overdisp.sirv, 
                                    cols = c(ont_bulk, ill_bulk, pb_bulk), 
                                    names_to = "Method", values_to = "Overdispersion")

p3 <- ggplot(overdisp.sirv_pivot, aes(x = tx_count_bin, y = Overdispersion, fill = Method)) +
  geom_boxplot(position = position_dodge(width = 0.75), outliers=FALSE) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  labs(title = "SIRV assignment ambiguity (Salmon overdispersion)", x = "Transcript number", y = "Overdispersion") +   ylim(0,20)

grid.arrange(p1.1, p1.2,  p2, p3,  ncol=2)
```



## ## Stratify by transcript length
```{r}

overdisp.human <- data.frame(
  ont_bulk = ont_bulk.dge.human$genes$Overdispersion,
  ill_bulk = ill_bulk.dge.human$genes$Overdispersion[match(rownames(ont_bulk.dge.human), rownames(ill_bulk.dge.human))],
  pb_bulk = pb_bulk.dge.human$genes$Overdispersion[match(rownames(ont_bulk.dge.human), rownames(pb_bulk.dge.human))],
  ont_sc = ont_sc.dge.human$genes$Overdispersion[match(rownames(ont_bulk.dge.human), rownames(ont_sc.dge.human))],
  ont_sn = ont_sn.dge.human$genes$Overdispersion[match(rownames(ont_bulk.dge.human), rownames(ont_sn.dge.human))],
  tx_len =  ont_bulk.dge.human$genes$Length
)

overdisp.human$tx_len_bin <- cut(overdisp.human$tx_len, breaks = c(0,500,1500,2000,2500,3000, Inf))
overdisp.human$tx_len_bin <- factor(overdisp.human$tx_len_bin, ordered = TRUE)
overdisp.human_pivot <- pivot_longer(overdisp.human, 
                            cols = c(ont_bulk, pb_bulk,ill_bulk), 
                            names_to = "Method", values_to = "Overdispersion")

p1 <- ggplot(overdisp.human_pivot, aes(x = tx_len_bin, y = Overdispersion, fill = Method)) +
  geom_boxplot(position = position_dodge(width = 0.75), outliers=FALSE) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  labs(title = "Human transcript assignment ambiguity (Salmon overdispersion)", x = "Transcript length", y = "Overdispersion")


overdisp.human_pivot <- pivot_longer(overdisp.human, 
                            cols = c(ont_bulk, pb_bulk), 
                            names_to = "Method", values_to = "Overdispersion")
p2 <- ggplot(overdisp.human_pivot, aes(x = tx_len_bin, y = Overdispersion, fill = Method)) +
  geom_boxplot(position = position_dodge(width = 0.75), outliers=FALSE) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  labs(title = "Human transcript assignment ambiguity (Salmon overdispersion)", x = "Transcript length", y = "Overdispersion")
  
```


# Sequins (Scatter)
```{r, fig.width=20}
get_sequins_scatter_plot <- function(dge, figure_name, tx_len_norm = TRUE) {

    dge.sequins <- dge[is.sequins(dge),]

    # Normalisation functions
    get_ltpm <- function(counts,len) {
        x <- counts/len
        get_lcpm(x)
    }

    get_lcpm <- function(x) { 
        return(log2(x*1e6/sum(x) + 1))
    }
    
    # Prepare Sequins ground truth
    sequins.gt <- read_tsv(params$sequins_tsv)
    sequins.gt.norm <- sequins.gt %>% mutate(across(starts_with("MIX"), get_lcpm)) %>% select(NAME, MIX_A, MIX_B, LENGTH) %>% mutate(N_transcript_per_gene = gsub("^R.*_", "", NAME) %>% as.numeric)
    
    # Normalise observed data
    
    if (tx_len_norm) {
        dge.sequins.norm <- dge.sequins$counts %>% as_tibble(rownames="NAME")  %>% mutate(across(where(is.numeric), ~ get_ltpm(.x, dge.sequins$gene$Length)))
    } else {
        dge.sequins.norm <- dge.sequins$counts %>% as_tibble(rownames="NAME")  %>% mutate(across(where(is.numeric), get_lcpm))
    }
    dge.sequins.norm  <- dge.sequins.norm %>% left_join(dge.sequins$gene %>% as_tibble(rownames="NAME") %>% select(NAME, Overdispersion))



    # Merge the ground truth and observed data
    mixA_df <-dge.sequins.norm %>% select(-dge.sequins$samples$sample[dge.sequins$samples$sequins=='B']) %>% left_join(sequins.gt.norm)  %>% 
          mutate(GT=MIX_A, MIX='A')  %>% 
          select(-MIX_B, -MIX_A)
    mixB_df <- dge.sequins.norm %>% select(-dge.sequins$samples$sample[dge.sequins$samples$sequins=='A']) %>% left_join(sequins.gt.norm) %>% 
      mutate(GT=MIX_B, MIX='B')  %>% 
      select(-MIX_B, -MIX_A)
    
    mixA_df.long <- mixA_df %>% pivot_longer(cols = mixA_df %>% select(dge.sequins$samples$sample[dge.sequins$samples$sequins=='A']) %>% colnames, names_to = "Sample", values_to = "Observed log-CPM")
    mixB_df.long <- mixB_df %>% pivot_longer(cols = mixB_df %>% select(dge.sequins$samples$sample[dge.sequins$samples$sequins=='B']) %>% colnames, names_to = "Sample", values_to = "Observed log-CPM")
    combined_df <- rbind(mixA_df.long, mixB_df.long)
    
    per_sample_cor <- sapply(dge$samples$sample, function(x) {
      cor(combined_df[combined_df$Sample == x,]$GT, combined_df[combined_df$Sample == x,]$`Observed log-CPM`) %>% unlist
    })
    
    # Calculate the correlation
    combined_df$Sample <- factor(combined_df$Sample)
    combined_df$MIX <- factor(combined_df$MIX)
    correlation <- cor(combined_df$GT, combined_df$`Observed log-CPM`)
    
    p1 <- ggplot(combined_df[rownames(combined_df) %>% sample(nrow(combined_df), replace = F),], aes(y = GT, x = `Observed log-CPM`, color = Sample, shape = MIX)) +
      geom_point() +
      geom_abline(slope = 1, intercept = 0, linetype = "dotted") +  # Add a dotted line of x=y
      theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
      theme_classic() + 
      # scale_y_log10(labels = label_number()) + 
      labs(title = figure_name, y = "Expected abundance log2(TPM+1)", x = "Observed log2(TPM+1)") +
      annotate("text", x = min(combined_df$`Observed log-CPM`), y = max(combined_df$GT), 
               label = paste("Correlation: ", round(correlation, 2)), 
               hjust = 0, vjust = 1, 
               color = "black", size = 4, 
               fontface = "italic")
    
    combined_df <- combined_df %>% mutate(N_transcript_per_gene = N_transcript_per_gene %>% as.factor(), `Transcript length` = LENGTH)
    p2 <- ggplot(combined_df[rownames(combined_df) %>% sample(nrow(combined_df), replace = F),], aes(y = GT, x = `Observed log-CPM`, color = log10(`Transcript length`), shape = N_transcript_per_gene, size = Overdispersion)) +
      geom_point() +
      geom_abline(slope = 1, intercept = 0, linetype = "dotted") +  # Add a dotted line of x=y
      theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
      theme_classic() + 
      # scale_y_log10(labels = label_number()) + 
      labs(title = figure_name, y = "Expected abundance log2(TPM+1)", x = "Observed log2(TPM+1)") +
      annotate("text", x = min(combined_df$`Observed log-CPM`), y = max(combined_df$GT), 
               label = paste("Correlation: ", round(correlation, 2)), 
               hjust = 0, vjust = 1, 
               color = "black", size = 4, 
               fontface = "italic") +
      scale_color_viridis_c() +
      scale_size_continuous(limits = c(0, 5))
    
    t <- tibble(
      sample = dge$samples$sample,
      sample_correlation = per_sample_cor,
      merged_correlation = correlation
    )
    return(list(plot1 = p1,  plot2 = p2, table = t))
}

cor.ont_bulk <- get_sequins_scatter_plot(ont_bulk.dge, figure_name = "ONT BULK Sequins", tx_len_norm = FALSE)
cor.pb_bulk <- get_sequins_scatter_plot(pb_bulk.dge, figure_name = "Pacbio BULK Sequins", tx_len_norm = FALSE)
cor.ill_bulk <- get_sequins_scatter_plot(ill_bulk.dge, figure_name = "Illumina BULK Sequins", tx_len_norm = TRUE)


plot_df <- cor.ont_bulk$table %>% dplyr::rename(`ONT Bulk`=sample_correlation) %>% 
            left_join(cor.ill_bulk$table %>% dplyr::rename(`Illumina Bulk`=sample_correlation), by = 'sample') %>% 
            left_join(cor.pb_bulk$table %>% dplyr::rename(`PacBio Bulk`=sample_correlation), by = 'sample') %>% 
            select(`ONT Bulk`, `PacBio Bulk`, `Illumina Bulk`) %>% 
            pivot_longer(col=c(`ONT Bulk`, `PacBio Bulk`, `Illumina Bulk`), values_to = "correlation", names_to = "Sample")

bx_plot <- ggplot(plot_df, aes(x = Sample, y = correlation, fill = Sample)) +
  geom_boxplot(position = position_dodge(width = 0.75)) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  labs(title = "Sequins transcript: per sample correlation with true abundance")

grid.arrange(
            #cor.ont_bulk$plot1, 
            cor.ont_bulk$plot2, 
            #cor.pb_bulk$plot1,  
            cor.pb_bulk$plot2,
            #cor.ill_bulk$plot1, 
            cor.ill_bulk$plot2, 
            bx_plot, ncol=1)

```


# SIRV E0 (Scatter)
```{r, fig.width=20}
get_sirv_e0_scatter_plot <- function(dge, figure_name, tx_len_norm = TRUE) {

    dge.sirv_e0 <- dge[is.sirv_E0(dge),]

    # Normalisation functions
    get_ltpm <- function(counts,len) {
        x <- counts/len
        get_lcpm(x)
    }

    get_lcpm <- function(x) { 
        return(log2(x*1e6/sum(x) + 1))
    }

    # Prepare sirv_e0 ground truth
    sirv_e0.gt <- tibble(
        NAME = dge.sirv_e0 %>% row.names(),
        LENGTH = dge.sirv_e0$gene$Length,
        GT = rep(1, nrow(dge.sirv_e0$gene))
    )
    sirv_e0.gt.norm <- sirv_e0.gt %>% mutate(GT = get_lcpm(GT)) %>%
                                    mutate(N_transcript_per_gene = SIRV.G.Tx.map$tx_count[match(NAME, SIRV.G.Tx.map$tx_name)])
    
    # Normalise observed data
    
    if (tx_len_norm) {
        dge.sirv_e0.norm <- dge.sirv_e0$counts %>% as_tibble(rownames="NAME")  %>% mutate(across(where(is.numeric), ~ get_ltpm(.x, dge.sirv_e0$gene$Length)))
    } else {
        dge.sirv_e0.norm <- dge.sirv_e0$counts %>% as_tibble(rownames="NAME")  %>% mutate(across(where(is.numeric), get_lcpm))
    }
    dge.sirv_e0.norm  <- dge.sirv_e0.norm %>% left_join(dge.sirv_e0$gene %>% as_tibble(rownames="NAME") %>% select(NAME, Overdispersion))



    # Merge the ground truth and observed data
    plot_df <- dge.sirv_e0.norm %>% left_join(sirv_e0.gt.norm) %>% pivot_longer(cols=dge.sirv_e0 %>% colnames(), names_to = "Sample", values_to = "Observed log-CPM")

    
    per_sample_sd <- sapply(dge$samples$sample, function(x) {
      sd(plot_df[plot_df$Sample == x,]$`Observed log-CPM`) %>% unlist
    })
    
    # Calculate the standard deviation
    plot_df$Sample <- factor(plot_df$Sample)
    SD <- sd(plot_df$`Observed log-CPM`)


    # Scatter plot
    plot_df <- plot_df %>% mutate(N_transcript_per_gene = N_transcript_per_gene %>% as.factor(), `Transcript length` = LENGTH)
    p1 <- ggplot(plot_df[rownames(plot_df) %>% sample(nrow(plot_df), replace = F),], aes(y = `Transcript length`, x = `Observed log-CPM`, color = Sample)) +
      geom_point() +
      geom_vline(xintercept = plot_df$GT[1], linetype = "dashed", color = "red") +
      theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
      theme_classic() + 
      labs(title = figure_name, y = "Transcript length", x = "Observed log2(TPM+1)") +
      annotate("text", x = min(plot_df$`Observed log-CPM`), y = max(plot_df$`Transcript length`), 
               label = paste("sd: ", round(SD, 2)), 
               hjust = 0, vjust = 1, 
               color = "black", size = 4, 
               fontface = "italic")
    
    p2 <- ggplot(plot_df[rownames(plot_df) %>% sample(nrow(plot_df), replace = F),], aes(y =`Transcript length`, x = `Observed log-CPM`, color = log10(`Transcript length`), shape = N_transcript_per_gene, size = Overdispersion)) +
      geom_point() +
      geom_vline(xintercept = plot_df$GT[1], linetype = "dashed", color = "red") + 
      theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
      theme_classic() + 
      labs(title = figure_name, y = "Transcript length", x = "Observed log2(TPM+1)") +
      annotate("text", x = min(plot_df$`Observed log-CPM`), y = max(plot_df$`Transcript length`), 
               label = paste("sd: ", round(SD, 2)),
               hjust = 0, vjust = 1, 
               color = "black", size = 4, 
               fontface = "italic") +
      scale_color_viridis_c()
    p2 <- ggMarginal(p2, type = "histogram", margins = "x", fill = NA)
    
    t <- tibble(
      sample = dge$samples$sample,
      sample_sd = per_sample_sd,
      merged_sd = SD
    )
    return(list(plot1 = p1,  plot2 = p2, table = t))
}

cor.ont_bulk <- get_sirv_e0_scatter_plot(ont_bulk.dge, figure_name = "ONT BULK sirv_e0", tx_len_norm = FALSE)
cor.pb_bulk <- get_sirv_e0_scatter_plot(pb_bulk.dge, figure_name = "Pacbio BULK sirv_e0", tx_len_norm = FALSE)
cor.ill_bulk <- get_sirv_e0_scatter_plot(ill_bulk.dge, figure_name = "Illumina BULK sirv_e0", tx_len_norm = TRUE)


plot_df <- cor.ont_bulk$table %>% dplyr::rename(`ONT Bulk`=sample_sd) %>% 
            left_join(cor.ill_bulk$table %>% dplyr::rename(`Illumina Bulk`=sample_sd), by = 'sample') %>% 
            left_join(cor.pb_bulk$table %>% dplyr::rename(`PacBio Bulk`=sample_sd), by = 'sample') %>% 
            select(`ONT Bulk`, `PacBio Bulk`, `Illumina Bulk`) %>% 
            pivot_longer(col=c(`ONT Bulk`, `PacBio Bulk`, `Illumina Bulk`), values_to = "sd", names_to = "Sample")

bx_plot <- ggplot(plot_df, aes(x = Sample, y = sd, fill = Sample)) +
  geom_boxplot(position = position_dodge(width = 0.75)) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  labs(title = "sirv_e0 transcript: per sample sd")

grid.arrange(
            #cor.ont_bulk$plot1, 
            cor.ont_bulk$plot2, 
            #cor.pb_bulk$plot1,  
            cor.pb_bulk$plot2,
            #cor.ill_bulk$plot1, 
            cor.ill_bulk$plot2, 
            bx_plot, ncol=1)

```

## Transcript discovery rate (transcript with non-zero count)
```{r, fig.width=20}
get_tx_discovery_rate <- function(dge_list, dataset_list, figure_name="Transcript discovery rate"){
  plot_df <- tibble()
  for (i in 1:length(dge_list)) {
    dge <- dge_list[[i]]
    dataset <- dataset_list[[i]]
    non_zero_proportions <- dge$counts %>% apply(2, function(x) mean(x != 0)) %>% tibble()
    colnames(non_zero_proportions) <- "Transcript discovery rate"
    non_zero_proportions$sample <- dge$counts %>% colnames
    non_zero_proportions$Dataset <- dataset
    plot_df <- rbind(plot_df, non_zero_proportions)
  }
  ggplot(plot_df, aes(x = sample, y = `Transcript discovery rate`, fill = Dataset)) +
    geom_bar(stat = "identity", position = position_dodge()) +
    theme_minimal() +
    labs(title = figure_name, x = "Sample", y = "Transcript Discovery Rate") +
    geom_text(aes(label = scales::percent(`Transcript discovery rate`)), 
              position = position_dodge(width = 0.9), 
              vjust = -0.5, size = 2) 
}

p1 <- get_tx_discovery_rate(
  list(ont_bulk.dge[is.sequins(ont_bulk.dge),], ill_bulk.dge[is.sequins(ill_bulk.dge),], pb_bulk.dge[is.sequins(pb_bulk.dge),]),
  list("ONT BULK", "Illumina BULK", "PacBio BULK"),
  figure_name = "Sequins transcript discovery rate"
)

p2 <- get_tx_discovery_rate(
  list(ont_bulk.dge[is.sirv(ont_bulk.dge),], ill_bulk.dge[is.sirv(ill_bulk.dge),], pb_bulk.dge[is.sirv(pb_bulk.dge),]),
  list("ONT BULK", "Illumina BULK", "PacBio BULK"),
  figure_name = "SIRV transcript discovery rate"
)

p3 <- get_tx_discovery_rate(
  list(ont_bulk.dge[is.human(ont_bulk.dge),], 
       ill_bulk.dge[is.human(ill_bulk.dge),], 
       ont_sc_clean.dge[is.human(ont_sc_clean.dge),], 
       ont_sn_clean.dge[is.human(ont_sn_clean.dge),], 
       pb_bulk.dge[is.human(pb_bulk.dge),]),
  list("ONT BULK", "Illumina BULK", "ONT SC (cleaned)", "ONT SN (cleaned)", "PacBio BULK"),
  figure_name = "Human transcript discovery rate"
)

p4 <- get_tx_discovery_rate(
  list(ont_sc.dge[is.human(ont_sc.dge),], ont_sn.dge[is.human(ont_sn.dge),], ont_sc_clean.dge[is.human(ont_sc_clean.dge),], ont_sn_clean.dge[is.human(ont_sn_clean.dge),]),
  list("ONT SC", "ONT SN", "ONT SC (cleaned)", "ONT SN (cleaned)"),
  figure_name = "Human transcript discovery rate (SC/SN data)"
)

# Combine all plots
grid.arrange(p1, p2, p3, p4, ncol=2)


```


# Normalisation and MDS
```{r, fig.height=15, fig.width=10}
norm_and_mds <- function(dge, dataset='ont_bulk', col = NULL, text = NULL, norm.method = "TMM") {
  # filter 
  filter <- filterByExpr(dge, group=dge$samples$sample)
  dge <- dge[filter, , keep.lib.sizes = FALSE]
  
  
   # normalisation
  if (!is.null(norm.method)) {
     dge <- calcNormFactors(dge, method = norm.method)
  }
 
  cpm <- cpm(dge, log = TRUE)

  # MDS
  if (is.null(col)) {
    col <- dge$samples$group
  }
  
  if (is.null(text)) {
    text <- dge$samples$group
  }

  factor_vector <- as.factor(col)
  col <- brewer.pal(n = length(levels(factor_vector)), name = "Set1")
  color_mapping <- setNames(col, levels(factor_vector))
  boxplot(cpm, las=2, col=color_mapping[as.character(factor_vector)], main="")
  plotMDS(cpm, main = dataset,
          text = text,
          col = color_mapping[as.character(factor_vector)])
}


par(mfrow=c(3,2))

norm_and_mds(ont_bulk.dge[is.human(ont_bulk.dge),], 
            dataset='ont bulk human (colored by sample groups)',
            col = ont_bulk.dge$samples$group)

norm_and_mds(ont_bulk.dge[is.sequins(ont_bulk.dge),], 
            dataset='ont bulk sequins (colored by sequins mix A or B)',
            col = ont_bulk.dge$samples$sequins,
            norm.method = NULL)
norm_and_mds(ont_bulk.dge[is.sirv(ont_bulk.dge),], 
            dataset='ont bulk SIRV (colored by sample groups)',
            col = ont_bulk.dge$samples$group,
            norm.method = NULL)


norm_and_mds(pb_bulk.dge[is.human(pb_bulk.dge),], 
            dataset='PacBio bulk human (colored by sample groups)',
            col = pb_bulk.dge$samples$group)

norm_and_mds(pb_bulk.dge[is.sequins(pb_bulk.dge),], 
            dataset='PacBio bulk sequins (colored by sequins mix A or B)',
            col = pb_bulk.dge$samples$sequins,
            norm.method = NULL)
norm_and_mds(pb_bulk.dge[is.sirv(pb_bulk.dge),], 
            dataset='PacBio bulk SIRV (colored by sample groups)',
            col = pb_bulk.dge$samples$group,
            norm.method = NULL)


norm_and_mds(ill_bulk.dge[is.human(ill_bulk.dge),], 
            dataset='ILL bulk human (colored by sample groups)',
            col = ill_bulk.dge$samples$group)

norm_and_mds(ill_bulk.dge[is.sequins(ill_bulk.dge),], 
            dataset='ILL bulk sequins (colored by sequins mix A or B)',
            col = ill_bulk.dge$samples$sequins,
            norm.method = NULL)

norm_and_mds(ill_bulk.dge[is.sirv(ill_bulk.dge),], 
            dataset='ILL bulk SIRV (colored by sample groups)',
            col = ill_bulk.dge$samples$group,
            norm.method = NULL)

par(mfrow=c(3,2))
norm_and_mds(ont_sc_clean.dge[is.human(ont_sc_clean.dge),], 
            dataset='ONT SC pseudobulk human (colored by sample groups)',
            col = ont_sc_clean.dge$samples$group,
            norm.method = NULL)

norm_and_mds(ont_sn_clean.dge[is.human(ont_sn_clean.dge),], 
            dataset='ONT SN pseudobulk human (colored by sample groups)',
            col = ont_sn_clean.dge$samples$group,
            norm.method = NULL)

par(mfrow=c(3,2))
norm_and_mds(ont_sc.dge[is.human(ont_sc.dge),], 
            dataset='ONT SC pseudobulk human (colored by sample groups)',
            col = ont_sc_clean.dge$samples$group,
            norm.method = NULL)

norm_and_mds(ont_sn.dge[is.human(ont_sn.dge),], 
            dataset='ONT SN pseudobulk human (colored by sample groups)',
            col = ont_sn_clean.dge$samples$group,
            norm.method = NULL)


```

# Differential gene expression
## Aggregate counts (tximport)

```{r}
library(tximport)

# Extract tx_name and gene_id columns from human and sequins
human_tx2gene <- human.G.Tx.map[, c("tx_name", "gene_id")]
sequins_tx2gene <- sequins.G.Tx.map[, c("tx_name", "gene_id")]
combined_tx2gene <- rbind(human_tx2gene, sequins_tx2gene)

# get quant.sf file paths
get_quant_paths <- function(dirs, file_name = "quant.sf") {
  # Generate file paths for quantification files
  quant_paths <- file.path(dirs, file_name)
  return(quant_paths)
}

# Define directories
salmon_dirs <- list(
  ill_bulk = params$ill_bulk_salmon_dir,
  ont_bulk = params$ont_bulk_salmon_dir,
  pb_bulk = params$pb_bulk_salmon_dir,
  ont_sc = params$ont_sc_salmon_dir,
  ont_sn = params$ont_sn_salmon_dir,
  ont_sc_clean = params$ont_sc_clean_salmon_dir,
  ont_sn_clean = params$ont_sn_clean_salmon_dir
)

# Apply function to each directory
quant_paths <- lapply(salmon_dirs, function(dir) {
  salmon_dirs <- file.path(dir, list.dirs(dir, full.names = FALSE, recursive = FALSE))
  get_quant_paths(salmon_dirs)
})


get_dge_from_txi <- function(quant_dir,sample_prefix_regex) {
  # Import data using tximport
  txi <- tximport(quant_dir, 
                  type = "salmon", 
                  tx2gene = combined_tx2gene,
                  ignoreAfterBar = TRUE, 
                  countsFromAbundance = "no") # raw counts
  counts <- as.data.frame(txi$counts, stringAsFactors = FALSE)
  sample_names <- gsub(sample_prefix_regex, '', basename(dirname(quant_dir)))
  colnames(counts) <- sample_names
  rst.dge <- DGEList(counts = counts)
  rownames(rst.dge$samples) <- rownames(rst.dge$samples) %>% sub(sample_prefix_regex, '', .)
  colnames(rst.dge$counts) <- colnames(rst.dge$counts) %>% sub(sample_prefix_regex, '', .)
  rst.dge$samples <- merge(rst.dge$samples %>% select(-group), bulk.meta, by = "row.names", all.x = TRUE) %>% select(-Row.names)
  rst.dge$samples <- rst.dge$samples[match(colnames(rst.dge$counts), rst.dge$samples$sample),]
  return(rst.dge)
}

#txi_ill_bulk.dge <- get_dge_from_txi(quant_paths$ill_bulk,".*_S")
txi_ont_bulk.dge <- get_dge_from_txi(quant_paths$ont_bulk,".*/")
txi_pb_bulk.dge <- get_dge_from_txi(quant_paths$pb_bulk,".*/")
txi_ont_sc.dge <- get_dge_from_txi(quant_paths$ont_sc,".*/")
txi_ont_sn.dge <- get_dge_from_txi(quant_paths$ont_sn,".*/")
txi_ont_sc_clean.dge <- get_dge_from_txi(quant_paths$ont_sc_clean,".*/")
txi_ont_sn_clean.dge <- get_dge_from_txi(quant_paths$ont_sn_clean,".*/")


# Bulk ill
salmon_ill_bulk_quant <- quant_paths$ill_bulk
txi <- tximport(salmon_ill_bulk_quant, 
                  type = "salmon", 
                  tx2gene = combined_tx2gene,
                  ignoreAfterBar = TRUE, 
                  countsFromAbundance = "no") # raw counts
counts <- as.data.frame(txi$counts, stringAsFactors = FALSE)
sample_names <- (basename(dirname(salmon_ill_bulk_quant)))
colnames(counts) <- sample_names
txi_ill_bulk.dge <- DGEList(counts = counts)
rownames(txi_ill_bulk.dge$samples) <- rownames(txi_ill_bulk.dge$samples) %>% sub(".*_S", '', .)
rownames(txi_ill_bulk.dge$samples) <- bulk.meta$sample[match(bulk.meta$index, rownames(txi_ill_bulk.dge$samples))]
colnames(txi_ill_bulk.dge$counts) <- colnames(txi_ill_bulk.dge$counts) %>% sub(".*_S", '', .)
colnames(txi_ill_bulk.dge$counts) <- bulk.meta$sample[match(bulk.meta$index, colnames(txi_ill_bulk.dge$counts))]
txi_ill_bulk.dge$samples <- merge(txi_ill_bulk.dge$samples %>% select(-group), bulk.meta, by = "row.names", all.x = TRUE) %>% select(-Row.names)
correct_order <- match(colnames(txi_ill_bulk.dge$counts), txi_ill_bulk.dge$samples$sample)
txi_ill_bulk.dge$samples <- txi_ill_bulk.dge$samples[correct_order, ]


```



## DGE edgeR
```{r}

DGE.edgeR_human <- function(dge, norm.method = "TMM") {
  # QC and normalization
  filter <- filterByExpr(dge, group = dge$samples$sample)
  dge <- dge[filter, , keep.lib.sizes = FALSE]
  
  if (!is.null(norm.method)) {
    dge <- calcNormFactors(dge, method = norm.method)
  }
  
  design <- model.matrix(~0 + group, data = dge$samples)
  dge <- estimateDisp(dge, design)
  qlfit <- glmQLFit(dge, design)
  
  contrasts <- makeContrasts(
    AvsL = groupsclc_a - groupluad,
    PvsL = groupsclc_p - groupluad,
    AvsP = groupsclc_a - groupsclc_p,
    levels = design
  )
  
  AvsL.QLF.test <- glmQLFTest(qlfit, contrast = contrasts[, "AvsL"]) %>% decideTests
  PvsL.QLF.test <- glmQLFTest(qlfit, contrast = contrasts[, "PvsL"]) %>% decideTests
  AvsP.QLF.test <- glmQLFTest(qlfit, contrast = contrasts[, "AvsP"]) %>% decideTests
  
  list(
    AvsL.DE = rownames(AvsL.QLF.test)[AvsL.QLF.test != 0],
    PvsL.DE = rownames(PvsL.QLF.test)[PvsL.QLF.test != 0],
    AvsP.DE = rownames(AvsP.QLF.test)[AvsP.QLF.test != 0]
  )
}

DGE.edgeR_sequins <- function(dge, norm.method = "TMM") {
  # QC and normalization
  filter <- filterByExpr(dge, group = dge$samples$sample)
  dge <- dge[filter, , keep.lib.sizes = FALSE]
  
  if (!is.null(norm.method)) {
    dge <- calcNormFactors(dge, method = norm.method)
  }
  
  design <- model.matrix(~0 + sequins, data = dge$samples)
  dge <- estimateDisp(dge, design)
  qlfit <- glmQLFit(dge, design)
  
  contrasts <- makeContrasts(
    AvsB = sequinsA - sequinsB,
    levels = design
  )
  
  AvsB.QLF.test <- glmQLFTest(qlfit, contrast = contrasts[, "AvsB"]) %>% decideTests
  
  list(
    AvsB.DE = rownames(AvsB.QLF.test)[AvsB.QLF.test != 0]
  )
}


ill_bulk.dge_human.list <- DGE.edgeR_human(txi_ill_bulk.dge[is.human(txi_ill_bulk.dge),])
ont_bulk.dge_human.list <- DGE.edgeR_human(txi_ont_bulk.dge[is.human(txi_ont_bulk.dge),])
pb_bulk.dge_human.list <- DGE.edgeR_human(txi_pb_bulk.dge[is.human(txi_pb_bulk.dge),])
# ont_sc.dge_human.list <- DGE.edgeR_human(txi_ont_sc.dge[is.human(txi_ont_sc.dge),])
# ont_sn.dge_human.list <- DGE.edgeR_human(txi_ont_sn.dge[is.human(txi_ont_sn.dge),])
# ont_sn_clean.dge_human.list <- DGE.edgeR_human(txi_ont_sc_clean.dge[is.human(txi_ont_sc_clean.dge),])
# ont_sc_clean.dge_human.list <- DGE.edgeR_human(txi_ont_sn_clean.dge[is.human(txi_ont_sn_clean.dge),])

ill_bulk.dge_sequins.list <- DGE.edgeR_sequins(txi_ill_bulk.dge[is.sequins(txi_ill_bulk.dge),])
ont_bulk.dge_sequins.list <- DGE.edgeR_sequins(txi_ont_bulk.dge[is.sequins(txi_ont_bulk.dge),])
pb_bulk.dge_sequins.list <- DGE.edgeR_sequins(txi_pb_bulk.dge[is.sequins(txi_pb_bulk.dge),])



```


# Differential transcript expression
## DTE edgeR 
```{r get_dte, message=FALSE}


DTE.edgeR_human <- function(dge, norm.method="TMM") {
  # QC and normalisation:
  filter <- filterByExpr(dge, group=dge$samples$sample)
  dge <- dge[filter, , keep.lib.sizes = FALSE]

  if (!is.null(norm.method)) {
     dge <- calcNormFactors(dge, method = norm.method)
  }

  design <- model.matrix(~0+group, data = dge$samples)
  dge <- estimateDisp(dge, design)
  qlfit <- glmQLFit(dge, design)
  contrasts <- makeContrasts(AvsL=groupsclc_a-groupluad, PvsL=groupsclc_p-groupluad, AvsP = groupsclc_a-groupsclc_p, levels=design)
  # SCLC_A vs LUCA
  AvsL.QLF.test <- glmQLFTest(qlfit, contrast=contrasts[,"AvsL"]) %>% decideTests
  # SCLC_P vs LUCA
  PvsL.QLF.test <- glmQLFTest(qlfit, contrast=contrasts[,"PvsL"]) %>% decideTests
  # SCLC_A vs SCLC_P
  AvsP.QLF.test <- glmQLFTest(qlfit, contrast=contrasts[,"AvsP"]) %>% decideTests

 list(
    AvsL.DE = rownames(AvsL.QLF.test)[AvsL.QLF.test!=0],
    PvsL.DE = rownames(PvsL.QLF.test)[PvsL.QLF.test!=0],
    AvsP.DE = rownames(AvsP.QLF.test)[AvsP.QLF.test!=0]
  )
}


DTE.edgeR_sequins <- function(dge, norm.method = "TMM") {
  # QC and normalization
  filter <- filterByExpr(dge, group = dge$samples$sample)
  dge <- dge[filter, , keep.lib.sizes = FALSE]
  
  if (!is.null(norm.method)) {
    dge <- calcNormFactors(dge, method = norm.method)
  }
  
  design <- model.matrix(~0 + sequins, data = dge$samples)
  dge <- estimateDisp(dge, design)
  qlfit <- glmQLFit(dge, design)
  
  contrasts <- makeContrasts(
    AvsB = sequinsA - sequinsB,
    levels = design
  )
  
  AvsB.QLF.test <- glmQLFTest(qlfit, contrast = contrasts[, "AvsB"]) %>% decideTests
  
  list(
    AvsB.DE = rownames(AvsB.QLF.test)[AvsB.QLF.test != 0]
  )
}

ill_bulk.dte_human.list <- DTE.edgeR_human(ill_bulk.dge[is.human(ill_bulk.dge),])
ont_bulk.dte_human.list <- DTE.edgeR_human(ont_bulk.dge[is.human(ont_bulk.dge),])
pb_bulk.dte_human.list <- DTE.edgeR_human(pb_bulk.dge[is.human(pb_bulk.dge),])
# ont_sc.dte_human.list <- DTE.edgeR_human(ont_sc.dge[is.human(ont_sc.dge),])
# ont_sn.dte_human.list <- DTE.edgeR_human(ont_sn.dge[is.human(ont_sn.dge),])
# ont_sn_clean.dte_human.list <- DTE.edgeR_human(ont_sn_clean.dge[is.human(ont_sn_clean.dge),])
# ont_sc_clean.dte_human.list <- DTE.edgeR_human(ont_sc_clean.dge[is.human(ont_sc_clean.dge),])

ill_bulk.dte_sequins.list <- DTE.edgeR_sequins(ill_bulk.dge[is.sequins(ill_bulk.dge),])
ont_bulk.dte_sequins.list <- DTE.edgeR_sequins(ont_bulk.dge[is.sequins(ont_bulk.dge),])
pb_bulk.dte_sequins.list <- DTE.edgeR_sequins(pb_bulk.dge[is.sequins(pb_bulk.dge),])


```

## DTE edgeR+scaling

```{r}
bulk.meta <- read.csv(params$bulk_meta)
rownames(bulk.meta) <- bulk.meta$sample


scale_dge_from_salmon <- function(dir, sample_prefix_regex) {
  salmon.dirs <- file.path(dir, 
                               list.dirs(dir, full.names = FALSE, recursive = FALSE))
  catched.salmon <- edgeR::catchSalmon(salmon.dirs, verbose = TRUE)
  
  #scale the counts with the overdispersion values
  rst.dge <- DGEList(counts=catched.salmon$counts/catched.salmon$annotation$Overdispersion,
                     gene=catched.salmon$annotation)
  
  rownames(rst.dge$samples) <- rownames(rst.dge$samples) %>% sub(sample_prefix_regex, '', .)
  colnames(rst.dge$counts) <- colnames(rst.dge$counts) %>% sub(sample_prefix_regex, '', .)
  rst.dge$samples <- merge(rst.dge$samples %>% select(-group), bulk.meta, by = "row.names", all.x = TRUE) %>% select(-Row.names)
  rst.dge$samples <- rst.dge$samples[match(colnames(rst.dge$counts), rst.dge$samples$sample),]
  return(rst.dge)
}


ont_bulk.scale_dge <- scale_dge_from_salmon(params$ont_bulk_salmon_dir, ".*/")
pb_bulk.scale_dge <- scale_dge_from_salmon(params$pb_bulk_salmon_dir, ".*/")
# ont_sc.scale_dge <- scale_dge_from_salmon(params$ont_sc_salmon_dir, ".*/")
# ont_sn.scale_dge <- scale_dge_from_salmon(params$ont_sn_salmon_dir, ".*/")
# ont_sc_clean.scale_dge <- scale_dge_from_salmon(params$ont_sc_clean_salmon_dir, ".*/")
# ont_sn_clean.scale_dge <- scale_dge_from_salmon(params$ont_sn_clean_salmon_dir, ".*/")


# Bulk ill
salmon.ill_bulk.dirs <- file.path(params$ill_bulk_salmon_dir,
                               list.dirs(params$ill_bulk_salmon_dir, full.names = FALSE, recursive = FALSE))
ill_bulk.salmon <- edgeR::catchSalmon(salmon.ill_bulk.dirs, verbose = TRUE)
ill_bulk.scale_dge <- DGEList(counts=ill_bulk.salmon$counts/ill_bulk.salmon$annotation$Overdispersion,
                              gene=ill_bulk.salmon$annotation)
rownames(ill_bulk.scale_dge$samples) <- rownames(ill_bulk.scale_dge$samples) %>% sub(".*_S", '', .)
rownames(ill_bulk.scale_dge$samples) <- bulk.meta$sample[match(bulk.meta$index, rownames(ill_bulk.scale_dge$samples))]
colnames(ill_bulk.scale_dge$counts) <- colnames(ill_bulk.scale_dge$counts) %>% sub(".*_S", '', .)
colnames(ill_bulk.scale_dge$counts) <- bulk.meta$sample[match(bulk.meta$index, colnames(ill_bulk.scale_dge$counts))]
ill_bulk.scale_dge$samples <- merge(ill_bulk.scale_dge$samples %>% select(-group), bulk.meta, by = "row.names", all.x = TRUE) %>% select(-Row.names)
ill_bulk.scale_dge$samples <- ill_bulk.scale_dge$samples[match(colnames(ill_bulk.scale_dge$counts), ill_bulk.scale_dge$samples$sample),]

```


```{r}
ill_bulk.scl_dte_human.list <- DTE.edgeR_human(ill_bulk.scale_dge[is.human(ill_bulk.scale_dge),])
ont_bulk.scl_dte_human.list <- DTE.edgeR_human(ont_bulk.scale_dge[is.human(ont_bulk.scale_dge),])
pb_bulk.scl_dte_human.list <- DTE.edgeR_human(pb_bulk.scale_dge[is.human(pb_bulk.scale_dge),])
# ont_sc.scl_dte_human.list <- DTE.edgeR_human(ont_sc.scale_dge[is.human(ont_sc.scale_dge),])
# ont_sn.scl_dte_human.list <- DTE.edgeR_human(ont_sn.scale_dge[is.human(ont_sn.scale_dge),])
# ont_sn_clean.scl_dte_human.list <- DTE.edgeR_human(ont_sn_clean.scale_dge[is.human(ont_sn_clean.scale_dge),])
# ont_sc_clean.scl_dte_human.list <- DTE.edgeR_human(ont_sc_clean.scale_dge[is.human(ont_sc_clean.scale_dge),])

ill_bulk.scl_dte_sequins.list <- DTE.edgeR_sequins(ill_bulk.scale_dge[is.sequins(ill_bulk.scale_dge),])
ont_bulk.scl_dte_sequins.list <- DTE.edgeR_sequins(ont_bulk.scale_dge[is.sequins(ont_bulk.scale_dge),])
pb_bulk.scl_dte_sequins.list <- DTE.edgeR_sequins(pb_bulk.scale_dge[is.sequins(pb_bulk.scale_dge),])
```


# Summary DE
## DGE human
```{r upset plot dge human, fig.width=8, fig.height=6}

# Upset plot
## SCLC_A vs LUCA (AvsL)
all_elements <- unique(c(
  ILL_BULK = ill_bulk.dge_human.list$AvsL.DE,
  ONT_BULK = ont_bulk.dge_human.list$AvsL.DE,
  PB_BULK = pb_bulk.dge_human.list$AvsL.DE
  # ONT_SC = ont_sc.dge_human.list$AvsL.DE,
  # ONT_SN = ont_sn.dge_human.list$AvsL.DE,
  # ONT_SN_CLEAN = ont_sn_clean.dge_human.list$AvsL.DE,
  # ONT_SC_CLEAN = ont_sc_clean.dge_human.list$AvsL.DE
))

plot_d_AvsL.DE <- data.frame(
  Element = all_elements,
  ill_bulk= as.integer(all_elements %in% ill_bulk.dge_human.list$AvsL.DE),
  ont_bulk= as.integer(all_elements %in% ont_bulk.dge_human.list$AvsL.DE),
  pb_bulk = as.integer(all_elements %in% pb_bulk.dge_human.list$AvsL.DE)
  # ont_sc= as.integer(all_elements %in% ont_sc.dge_human.list$AvsL.DE),
  # ont_sn= as.integer(all_elements %in% ont_sn.dge_human.list$AvsL.DE),
  # ont_sn_clean= as.integer(all_elements %in% ont_sn_clean.dge_human.list$AvsL.DE),
  # ont_sc_clean= as.integer(all_elements %in% ont_sc_clean.dge_human.list$AvsL.DE)
)
plt1 <- upset(plot_d_AvsL.DE,  order.by = "freq", nsets = 8) 


## SCLC_P vs LUCA (PvsL)
all_elements <- unique(c(
 ILL_BULK = ill_bulk.dge_human.list$PvsL.DE,
 ONT_BULK = ont_bulk.dge_human.list$PvsL.DE,
 PB_BULK = pb_bulk.dge_human.list$PvsL.DE
#  ONT_SC = ont_sc.dge_human.list$PvsL.DE,
#  ONT_SN = ont_sn.dge_human.list$PvsL.DE,
#   ONT_SN_CLEAN = ont_sn_clean.dge_human.list$PvsL.DE,
#   ONT_SC_CLEAN = ont_sc_clean.dge_human.list$PvsL.DE
))

plot_d_PvsL.DE <- data.frame(
  Element = all_elements,
  ill_bulk= as.integer(all_elements %in% ill_bulk.dge_human.list$PvsL.DE),
  ont_bulk= as.integer(all_elements %in% ont_bulk.dge_human.list$PvsL.DE),
  pb_bulk = as.integer(all_elements %in% pb_bulk.dge_human.list$PvsL.DE)
  # ont_sc= as.integer(all_elements %in% ont_sc.dge_human.list$PvsL.DE),
  # ont_sn= as.integer(all_elements %in% ont_sn.dge_human.list$PvsL.DE),
  # ont_sn_clean= as.integer(all_elements %in% ont_sn_clean.dge_human.list$PvsL.DE),
  # ont_sc_clean= as.integer(all_elements %in% ont_sc_clean.dge_human.list$PvsL.DE)
)
plt2 <- upset(plot_d_PvsL.DE,  order.by = "freq", nsets = 8) 


## SCLC_A vs SCLC_P (AvsP)
all_elements <- unique(c(
 ILL_BULK = ill_bulk.dge_human.list$AvsP.DE,
 ONT_BULK = ont_bulk.dge_human.list$AvsP.DE,
 PB_BULK = pb_bulk.dge_human.list$AvsP.DE
#  ONT_SC = ont_sc.dge_human.list$AvsP.DE,
#  ONT_SN = ont_sn.dge_human.list$AvsP.DE,
#   ONT_SN_CLEAN = ont_sn_clean.dge_human.list$AvsP.DE,
#   ONT_SC_CLEAN = ont_sc_clean.dge_human.list$AvsP.DE
))

plot_d_AvsP.DE <- data.frame(
  Element = all_elements,
  ill_bulk= as.integer(all_elements %in% ill_bulk.dge_human.list$AvsP.DE),
  ont_bulk= as.integer(all_elements %in% ont_bulk.dge_human.list$AvsP.DE),
  pb_bulk = as.integer(all_elements %in% pb_bulk.dge_human.list$AvsP.DE)
  # ont_sc= as.integer(all_elements %in% ont_sc.dge_human.list$AvsP.DE),
  # ont_sn= as.integer(all_elements %in% ont_sn.dge_human.list$AvsP.DE),
  # ont_sn_clean= as.integer(all_elements %in% ont_sn_clean.dge_human.list$AvsP.DE),
  # ont_sc_clean= as.integer(all_elements %in% ont_sc_clean.dge_human.list$AvsP.DE)
)
plt3 <- upset(plot_d_AvsP.DE,  order.by = "freq", nsets = 8) 

plt1
grid::grid.text("DGE SCLC_A vs LUAD (EdgeR)", x = 0.5, y = 0.95)

plt2
grid::grid.text("DGE SCLC_P vs LUAD (EdgeR)", x = 0.5, y = 0.95)

plt3
grid::grid.text("DGE SCLC_P vs SCLC_A (EdgeR)", x = 0.5, y = 0.95)

```


## DGE sequins 

```{r upset plot dge sequins, fig.width=8, fig.height=6}

# Upset plot
## A vs B (AvsB)
all_elements <- unique(c(
  ILL_BULK = ill_bulk.dge_sequins.list$AvsB.DE,
  ONT_BULK = ont_bulk.dge_sequins.list$AvsB.DE,
  PB_BULK = pb_bulk.dge_sequins.list$AvsB.DE
))

plot_d_AvsB.DE <- data.frame(
  Element = all_elements,
  ill_bulk= as.integer(all_elements %in% ill_bulk.dge_sequins.list$AvsB.DE),
  ont_bulk= as.integer(all_elements %in% ont_bulk.dge_sequins.list$AvsB.DE),
  pb_bulk = as.integer(all_elements %in% pb_bulk.dge_sequins.list$AvsB.DE)
)
plt4 <- upset(plot_d_AvsB.DE,  order.by = "freq", nsets = 8) 


plt4
grid::grid.text("DGE SequinsA vs SequinsB (EdgeR)", x = 0.5, y = 0.95)


```



## DTE human(edgeR)
```{r upset plot dte human edgeR, fig.width=8, fig.height=6}

# Upset plot
## SCLC_A vs LUCA (AvsL)
all_elements <- unique(c(
  ILL_BULK = ill_bulk.dte_human.list$AvsL.DE,
  ONT_BULK = ont_bulk.dte_human.list$AvsL.DE,
  PB_BULK = pb_bulk.dte_human.list$AvsL.DE
  # ONT_SC = ont_sc.dte_human.list$AvsL.DE,
  # ONT_SN = ont_sn.dte_human.list$AvsL.DE,
  # ONT_SN_CLEAN = ont_sn_clean.dte_human.list$AvsL.DE,
  # ONT_SC_CLEAN = ont_sc_clean.dte_human.list$AvsL.DE
))

plot_d_AvsL.DE <- data.frame(
  Element = all_elements,
  ill_bulk= as.integer(all_elements %in% ill_bulk.dte_human.list$AvsL.DE),
  ont_bulk= as.integer(all_elements %in% ont_bulk.dte_human.list$AvsL.DE),
  pb_bulk = as.integer(all_elements %in% pb_bulk.dte_human.list$AvsL.DE)
  # ont_sc= as.integer(all_elements %in% ont_sc.dte_human.list$AvsL.DE),
  # ont_sn= as.integer(all_elements %in% ont_sn.dte_human.list$AvsL.DE),
  # ont_sn_clean= as.integer(all_elements %in% ont_sn_clean.dte_human.list$AvsL.DE),
  # ont_sc_clean= as.integer(all_elements %in% ont_sc_clean.dte_human.list$AvsL.DE)
)
p1 <- upset(plot_d_AvsL.DE,  order.by = "freq", nsets = 8) 


## SCLC_P vs LUCA (PvsL)
all_elements <- unique(c(
 ILL_BULK = ill_bulk.dte_human.list$PvsL.DE,
 ONT_BULK = ont_bulk.dte_human.list$PvsL.DE,
 PB_BULK = pb_bulk.dte_human.list$PvsL.DE
#  ONT_SC = ont_sc.dte_human.list$PvsL.DE,
#  ONT_SN = ont_sn.dte_human.list$PvsL.DE,
#   ONT_SN_CLEAN = ont_sn_clean.dte_human.list$PvsL.DE,
#   ONT_SC_CLEAN = ont_sc_clean.dte_human.list$PvsL.DE
))

plot_d_PvsL.DE <- data.frame(
  Element = all_elements,
  ill_bulk= as.integer(all_elements %in% ill_bulk.dte_human.list$PvsL.DE),
  ont_bulk= as.integer(all_elements %in% ont_bulk.dte_human.list$PvsL.DE),
  pb_bulk = as.integer(all_elements %in% pb_bulk.dte_human.list$PvsL.DE)
  # ont_sc= as.integer(all_elements %in% ont_sc.dte_human.list$PvsL.DE),
  # ont_sn= as.integer(all_elements %in% ont_sn.dte_human.list$PvsL.DE),
  # ont_sn_clean= as.integer(all_elements %in% ont_sn_clean.dte_human.list$PvsL.DE),
  # ont_sc_clean= as.integer(all_elements %in% ont_sc_clean.dte_human.list$PvsL.DE)
)
p2 <- upset(plot_d_PvsL.DE,  order.by = "freq", nsets = 8) 


## SCLC_A vs SCLC_P (AvsP)
all_elements <- unique(c(
 ILL_BULK = ill_bulk.dte_human.list$AvsP.DE,
 ONT_BULK = ont_bulk.dte_human.list$AvsP.DE,
 PB_BULK = pb_bulk.dte_human.list$AvsP.DE
#  ONT_SC = ont_sc.dte_human.list$AvsP.DE,
#  ONT_SN = ont_sn.dte_human.list$AvsP.DE,
#   ONT_SN_CLEAN = ont_sn_clean.dte_human.list$AvsP.DE,
#   ONT_SC_CLEAN = ont_sc_clean.dte_human.list$AvsP.DE
))

plot_d_AvsP.DE <- data.frame(
  Element = all_elements,
  ill_bulk= as.integer(all_elements %in% ill_bulk.dte_human.list$AvsP.DE),
  ont_bulk= as.integer(all_elements %in% ont_bulk.dte_human.list$AvsP.DE),
  pb_bulk = as.integer(all_elements %in% pb_bulk.dte_human.list$AvsP.DE)
  # ont_sc= as.integer(all_elements %in% ont_sc.dte_human.list$AvsP.DE),
  # ont_sn= as.integer(all_elements %in% ont_sn.dte_human.list$AvsP.DE),
  # ont_sn_clean= as.integer(all_elements %in% ont_sn_clean.dte_human.list$AvsP.DE),
  # ont_sc_clean= as.integer(all_elements %in% ont_sc_clean.dte_human.list$AvsP.DE)
)
p3 <- upset(plot_d_AvsP.DE,  order.by = "freq", nsets = 8) 

p1
grid::grid.text("DTE SCLC_A vs LUAD (EdgeR)", x = 0.5, y = 0.95)

p2
grid::grid.text("DTE SCLC_P vs LUAD (EdgeR)", x = 0.5, y = 0.95)

p3
grid::grid.text("DTE SCLC_P vs SCLC_A (EdgeR)", x = 0.5, y = 0.95)
```

## DTE sequins(edgeR) 
```{r upset plot dte sequins edgeR, fig.width=8, fig.height=6}

# Upset plot
## A vs B (AvsB)
all_elements <- unique(c(
  ILL_BULK = ill_bulk.dte_sequins.list$AvsB.DE,
  ONT_BULK = ont_bulk.dte_sequins.list$AvsB.DE,
  PB_BULK = pb_bulk.dte_sequins.list$AvsB.DE
))

plot_d_AvsB.DE <- data.frame(
  Element = all_elements,
  ill_bulk= as.integer(all_elements %in% ill_bulk.dte_sequins.list$AvsB.DE),
  ont_bulk= as.integer(all_elements %in% ont_bulk.dte_sequins.list$AvsB.DE),
  pb_bulk = as.integer(all_elements %in% pb_bulk.dte_sequins.list$AvsB.DE)
)
p4 <- upset(plot_d_AvsB.DE,  order.by = "freq", nsets = 8) 


p4
grid::grid.text("DTE SequinsA vs SequinsB", x = 0.5, y = 0.95)


```

## DTE human(edgeR + scaling)
```{r upset plot dte human edgeR + scaling, fig.width=8, fig.height=6}

# Upset plot
## SCLC_A vs LUCA (AvsL)
all_elements <- unique(c(
  ILL_BULK = ill_bulk.scl_dte_human.list$AvsL.DE,
  ONT_BULK = ont_bulk.scl_dte_human.list$AvsL.DE,
  PB_BULK = pb_bulk.scl_dte_human.list$AvsL.DE
  # ONT_SC = ont_sc.scl_dte_human.list$AvsL.DE,
  # ONT_SN = ont_sn.scl_dte_human.list$AvsL.DE,
  # ONT_SN_CLEAN = ont_sn_clean.scl_dte_human.list$AvsL.DE,
  # ONT_SC_CLEAN = ont_sc_clean.scl_dte_human.list$AvsL.DE
))

plot_d_AvsL.DE <- data.frame(
  Element = all_elements,
  ill_bulk= as.integer(all_elements %in% ill_bulk.scl_dte_human.list$AvsL.DE),
  ont_bulk= as.integer(all_elements %in% ont_bulk.scl_dte_human.list$AvsL.DE),
  pb_bulk = as.integer(all_elements %in% pb_bulk.scl_dte_human.list$AvsL.DE)
  # ont_sc= as.integer(all_elements %in% ont_sc.scl_dte_human.list$AvsL.DE),
  # ont_sn= as.integer(all_elements %in% ont_sn.scl_dte_human.list$AvsL.DE),
  # ont_sn_clean= as.integer(all_elements %in% ont_sn_clean.scl_dte_human.list$AvsL.DE),
  # ont_sc_clean= as.integer(all_elements %in% ont_sc_clean.scl_dte_human.list$AvsL.DE)
)
pp1 <- upset(plot_d_AvsL.DE,  order.by = "freq", nsets = 8) 


## SCLC_P vs LUCA (PvsL)
all_elements <- unique(c(
 ILL_BULK = ill_bulk.scl_dte_human.list$PvsL.DE,
 ONT_BULK = ont_bulk.scl_dte_human.list$PvsL.DE,
 PB_BULK = pb_bulk.scl_dte_human.list$PvsL.DE
#  ONT_SC = ont_sc.scl_dte_human.list$PvsL.DE,
#  ONT_SN = ont_sn.scl_dte_human.list$PvsL.DE,
#   ONT_SN_CLEAN = ont_sn_clean.scl_dte_human.list$PvsL.DE,
#   ONT_SC_CLEAN = ont_sc_clean.scl_dte_human.list$PvsL.DE
))

plot_d_PvsL.DE <- data.frame(
  Element = all_elements,
  ill_bulk= as.integer(all_elements %in% ill_bulk.scl_dte_human.list$PvsL.DE),
  ont_bulk= as.integer(all_elements %in% ont_bulk.scl_dte_human.list$PvsL.DE),
  pb_bulk = as.integer(all_elements %in% pb_bulk.scl_dte_human.list$PvsL.DE)
  # ont_sc= as.integer(all_elements %in% ont_sc.scl_dte_human.list$PvsL.DE),
  # ont_sn= as.integer(all_elements %in% ont_sn.scl_dte_human.list$PvsL.DE),
  # ont_sn_clean= as.integer(all_elements %in% ont_sn_clean.scl_dte_human.list$PvsL.DE),
  # ont_sc_clean= as.integer(all_elements %in% ont_sc_clean.scl_dte_human.list$PvsL.DE)
)
pp2 <- upset(plot_d_PvsL.DE,  order.by = "freq", nsets = 8) 


## SCLC_A vs SCLC_P (AvsP)
all_elements <- unique(c(
 ILL_BULK = ill_bulk.scl_dte_human.list$AvsP.DE,
 ONT_BULK = ont_bulk.scl_dte_human.list$AvsP.DE,
 PB_BULK = pb_bulk.scl_dte_human.list$AvsP.DE
#  ONT_SC = ont_sc.scl_dte_human.list$AvsP.DE,
#  ONT_SN = ont_sn.scl_dte_human.list$AvsP.DE,
#   ONT_SN_CLEAN = ont_sn_clean.scl_dte_human.list$AvsP.DE,
#   ONT_SC_CLEAN = ont_sc_clean.scl_dte_human.list$AvsP.DE
))

plot_d_AvsP.DE <- data.frame(
  Element = all_elements,
  ill_bulk= as.integer(all_elements %in% ill_bulk.scl_dte_human.list$AvsP.DE),
  ont_bulk= as.integer(all_elements %in% ont_bulk.scl_dte_human.list$AvsP.DE),
  pb_bulk = as.integer(all_elements %in% pb_bulk.scl_dte_human.list$AvsP.DE)
  # ont_sc= as.integer(all_elements %in% ont_sc.scl_dte_human.list$AvsP.DE),
  # ont_sn= as.integer(all_elements %in% ont_sn.scl_dte_human.list$AvsP.DE),
  # ont_sn_clean= as.integer(all_elements %in% ont_sn_clean.scl_dte_human.list$AvsP.DE),
  # ont_sc_clean= as.integer(all_elements %in% ont_sc_clean.scl_dte_human.list$AvsP.DE)
)
pp3 <- upset(plot_d_AvsP.DE,  order.by = "freq", nsets = 8) 

pp1
grid::grid.text("DTE SCLC_A vs LUAD (EdgeR + scale)", x = 0.5, y = 0.95)

pp2
grid::grid.text("DTE SCLC_P vs LUAD (EdgeR + scale)", x = 0.5, y = 0.95)

pp3
grid::grid.text("DTE SCLC_P vs SCLC_A (EdgeR + scale)", x = 0.5, y = 0.95)
```

## DTE sequins(edgeR + scaling)
```{r upset plot dte sequins edgeR+scaling, fig.width=8, fig.height=6}

# Upset plot
## A vs B (AvsB)
all_elements <- unique(c(
  ILL_BULK = ill_bulk.scl_dte_sequins.list$AvsB.DE,
  ONT_BULK = ont_bulk.scl_dte_sequins.list$AvsB.DE,
  PB_BULK = pb_bulk.scl_dte_sequins.list$AvsB.DE
))

plot_d_AvsB.DE <- data.frame(
  Element = all_elements,
  ill_bulk= as.integer(all_elements %in% ill_bulk.scl_dte_sequins.list$AvsB.DE),
  ont_bulk= as.integer(all_elements %in% ont_bulk.scl_dte_sequins.list$AvsB.DE),
  pb_bulk = as.integer(all_elements %in% pb_bulk.scl_dte_sequins.list$AvsB.DE)
)
pp4 <- upset(plot_d_AvsB.DE,  order.by = "freq", nsets = 8) 


pp4
grid::grid.text("DTE SequinsA vs SequinsB (EdgeR + scale)", x = 0.5, y = 0.95)


```
