---
title: "DE analysis"
author: "Yupei & Ash"
date: '`r Sys.Date()`'
version: 'v3.1_lfc0'
output:
  html_document:
    code_folding: hide
    toc: yes
    toc_float: yes
    number_sections: yes
    toc_depth: 4
  pdf_document:
    toc: yes
    toc_depth: '4'
params:
  random_seed: 2024
  cache_dir: NULL
  fig.path: NULL
  sirv_ercc_gtf: '/vast/projects/LongBench/reference_files/SIRV_Set4_Norm_Sequences_20210507/SIRV_ERCC_longSIRV_multi-fasta_20210507.gtf'
  sequins_gtf: '/vast/projects/LongBench/reference_files/Sequin_resources/v2.4/rnasequin_annotation_2.4.gtf'
  sequins_tsv: '/vast/projects/LongBench/reference_files/Sequin_resources/v2.4/rnasequin_isoforms_2.4.tsv'
  human_gtf: '/vast/projects/LongBench/reference_files/GRCh38/gencode.v44.annotation.gtf'
  sirv_csv: '/vast/projects/LongBench/reference_files/SIRV_Set4_Norm_Sequences_20210507/SIRV_set4_concentration.csv'
  lr_featureCcount: "/vast/projects/LongBench/analysis/lr_bulk/result/featureCounts"
  sr_featureCount: "/vast/projects/LongBench/analysis/sr_bulk/result/featureCounts"
  lr_pseudobulk_featureCount: "/vast/projects/LongBench/analysis/lr_sc_sn/result/PseudoBulkfeatureCounts"
  bulk_meta: "/vast/projects/LongBench/sequencing_data/illumina_bulk/metadata.txt"
  output_rds: "/vast/projects/LongBench/analysis/workflow/rmarkdown/RDS/bulk_DE.rds"  
  expected_de_table: '/vast/projects/LongBench/reference_files/Expected_DE.csv'
  log2FC_cutoff: 0

---

# Setup: Create DGE from Input

**Bulk RNA-seq** was performed to analyze gene and transcript expression across various cancer cell lines using both short- and long-read sequencing platforms.

1.  **Samples**:

    \- Total: 8 cancer cell lines (H146, H69, H526, H211, SHP77, H1975, H2228, HCC827) for each platform.

2.  **Groups by Cancer Cell Type**: - **Small Cell Lung Cancer (SCLC)**:

    \- **Subtype A (SCLC_A)**: H146, H69, SHP77.\

    -   **Subtype P (SCLC_P)**: H526, H211.\
    -   **Lung Adenocarcinoma (LUAD)**: - H1975, H2228, HCC827.

3.  **Sequencing Platforms**:

    \- **Short-Read Sequencing**: Illumina.

    \- **Long-Read Sequencing**: PacBio ,  Oxford Nanopore Technologies (ONT) (`ONT_1` (cDNA sequencing), `ONT_2` (dRNA sequencing))

```{r setup, include=FALSE}
knitr::opts_chunk$set(
	echo = TRUE,
	fig.height = 5,
	fig.width = 7,
	warning = FALSE,
	cache = FALSE
)
```

```{r Libraries, include=FALSE}
library(limma)
library(edgeR)
library(tidyverse)
library(RColorBrewer)
library(dplyr)
library(tidyr)
library(ggplot2)
library(ggrepel)
library(glue)
library(ggplot2)
library(gridExtra)
library(ggExtra)
library(knitr)
library(kableExtra)
library(rtracklayer)
library(scales)
library(UpSetR)
library(ggtree)
library(aplot)
source("/vast/projects/LongBench/analysis/workflow/scripts/Rfunctions.R")
bulk_rst <- list() # used to store result for downstream analysis

# setup some conflict functions
rename <- dplyr::rename
select <- dplyr::select

knitr::opts_chunk$set(
  cache = !is.null(params$cache_dir),  # Enable caching only if cache_dir is provided
  fig.width = 15,                     # Set default figure width
  fig.height = 10,                    # Set default figure height
  warning = FALSE,                    # Suppress warnings in the output
  fig.path = params$fig.path,         # Set figure path (NULL if not provided)
  dev = if (!is.null(params$fig.path)) c("png", "svg") else NULL  # Set output devices if fig.path is provided
)

# Set cache.path only if params$cache_dir is not NULL
if (!is.null(params$cache_dir)) {
  knitr::opts_chunk$set(cache.path = params$cache_dir)
}
```

```{r color code, include=FALSE}
color_palette <- c(
  PacBio = "#df1995",
  ONT = "#00789b",
  ONT_1 = "#04476c", # ONT cDNA
  ONT_2 = "#24cdcd", # ONT dRNA
  Illumina = "#e88b20"
)
```

```{r function for splitting the transcript origin, echo=TRUE, results='hide', message=FALSE, warning=FALSE}
is.sequins <- function(dge) {grepl("^R", rownames(dge))}
is.human <- function(dge) {grepl("^ENS", rownames(dge))}
is.sirv_or_ercc <- function(dge) {!is.human(dge) & !is.sequins(dge)} # this is the whole SIRV set4 (ERCC + SIRV E0 + Long SIRV)
is.sirv_E0 <- function(dge) {grepl("^SIRV\\d{3}$", rownames(dge))} 
is.sirv_all <- function(dge) {grepl("^SIRV", rownames(dge))} # including all SIRV E0 and Long SIRV
is.long_sirv <- function(dge) {grepl("^SIRV\\d{4}", rownames(dge))}
is.ercc <- function(dge) {!is.human(dge) & !is.sequins(dge) & !is.sirv_all(dge)}
```

```{r message=FALSE, warning=FALSE, include=FALSE}
calcTxNum <- function(G.Tx.map){
  geneid = rep(NA, nrow(G.Tx.map))
  txcount = table(G.Tx.map$gene_id)
  txnum = as.numeric(txcount[match(G.Tx.map$gene_id, names(txcount))])
}
# SIRV
SIRV_tx <- GenomicFeatures::makeTxDbFromGFF(params$sirv_ercc_gtf, format = "gtf") %>% GenomicFeatures::transcripts(columns = c("tx_id", "tx_name", "gene_id"))
SIRV.G.Tx.map <- tibble(
  gene_id = SIRV_tx$gene_id %>% unlist,
  tx_name = SIRV_tx$tx_name
)
rm(SIRV_tx)
SIRV.G.Tx.map$tx_count <- SIRV.G.Tx.map %>% calcTxNum

# Sequins
sequins_tx  <- GenomicFeatures::makeTxDbFromGFF(params$sequins_gtf, format = "gtf") %>% GenomicFeatures::transcripts(columns = c("tx_id", "tx_name", "gene_id"))
sequins.G.Tx.map <- tibble(
  gene_id = sequins_tx$gene_id %>% unlist,
  tx_name = sequins_tx$tx_name
)
rm(sequins_tx)
sequins.G.Tx.map$tx_count <- sequins.G.Tx.map %>% calcTxNum

# Human
human_tx  <- GenomicFeatures::makeTxDbFromGFF(params$human_gtf, format = "gtf") %>% GenomicFeatures::transcripts(columns = c("tx_id", "tx_name", "gene_id"))
human.G.Tx.map <- tibble(
  gene_id = human_tx$gene_id %>% unlist,
  tx_name = human_tx$tx_name
)
rm(human_tx)
human.G.Tx.map$tx_count <- human.G.Tx.map %>% calcTxNum
```

```{r Create DGE from input, echo=TRUE, message=FALSE, warning=FALSE, results='hide'}
bulk.meta <- read.csv(params$bulk_meta)
rownames(bulk.meta) <- bulk.meta$sample
```

# Count data Preprocessing

```{r, echo=TRUE, message=FALSE, warning=FALSE, results='hide'}
# SR
dir_path <- params$sr_featureCount
rds_files <- list.files(dir_path, pattern = "_featureCounts\\.rds$", full.names = TRUE)
dge_list <- lapply(rds_files, readRDS)
combined_counts <- do.call(cbind, lapply(dge_list, function(x) x$counts))
colnames(combined_counts) <- sub(".sorted.bam", "", colnames(combined_counts))
ill_bulk.gene.dge <- DGEList(counts = combined_counts)
ill_bulk.gene.dge$samples <- merge(ill_bulk.gene.dge$samples %>% select(-group), bulk.meta, by = "row.names", all.x = TRUE) %>% select(-Row.names)

# LR
dir_path <- params$lr_featureCcount

## dRNA
rds_files <- list.files(dir_path, pattern = "dRNA.*_featureCounts\\.rds$", full.names = TRUE)
dge_list <- lapply(rds_files, readRDS)
combined_counts <- do.call(cbind, lapply(dge_list, function(x) x$counts))
colnames(combined_counts) <- sub(".sorted.bam", "", colnames(combined_counts)) %>% sub("dRNA_bulk_", "", .)
drna_bulk.gene.dge <- DGEList(counts = combined_counts)
drna_bulk.gene.dge$samples <- merge(drna_bulk.gene.dge$samples %>% select(-group), bulk.meta, by = "row.names", all.x = TRUE) %>% select(-Row.names)

## cDNA
rds_files <- list.files(dir_path, pattern = "ont_bulk.*_featureCounts\\.rds$", full.names = TRUE)
dge_list <- lapply(rds_files, readRDS)
combined_counts <- do.call(cbind, lapply(dge_list, function(x) x$counts))
colnames(combined_counts) <- sub(".sorted.bam", "", colnames(combined_counts)) %>% sub("ont_bulk_", "", .)
ont_bulk.gene.dge <- DGEList(counts = combined_counts)
ont_bulk.gene.dge$samples <- merge(ont_bulk.gene.dge$samples %>% select(-group), bulk.meta, by = "row.names", all.x = TRUE) %>% select(-Row.names)

## PacBio
rds_files <- list.files(dir_path, pattern = "pb.*_featureCounts\\.rds$", full.names = TRUE)
dge_list <- lapply(rds_files, readRDS)
combined_counts <- do.call(cbind, lapply(dge_list, function(x) x$counts))
colnames(combined_counts) <- sub(".sorted.bam", "", colnames(combined_counts)) %>% sub("pb_bulk_", "", .)
pb_bulk.gene.dge <- DGEList(counts = combined_counts)
pb_bulk.gene.dge$samples <- merge(pb_bulk.gene.dge$samples %>% select(-group), bulk.meta, by = "row.names", all.x = TRUE) %>% select(-Row.names)

# SC SN
dir_path <- params$lr_pseudobulk_featureCount 
## ONT SC
rds_files <- list.files(dir_path, pattern = "ont_sc_.*_featureCounts\\.rds$", full.names = TRUE)
dge_list <- lapply(rds_files, readRDS)
combined_counts <- do.call(cbind, lapply(dge_list, function(x) x$counts))
colnames(combined_counts) <- sub(".sorted.bam", "", colnames(combined_counts)) %>% sub("ont_sc_", "", .)
ont_sc.gene.dge <- DGEList(counts = combined_counts)
ont_sc.gene.dge$samples <- merge(ont_sc.gene.dge$samples %>% select(-group), bulk.meta, by = "row.names", all.x = TRUE) %>% select(-Row.names)

## ONT SN
rds_files <- list.files(dir_path, pattern = "ont_sn_.*_featureCounts\\.rds$", full.names = TRUE)
dge_list <- lapply(rds_files, readRDS)
combined_counts <- do.call(cbind, lapply(dge_list, function(x) x$counts))
colnames(combined_counts) <- sub(".sorted.bam", "", colnames(combined_counts)) %>% sub("ont_sn_", "", .)
ont_sn.gene.dge <- DGEList(counts = combined_counts)
ont_sn.gene.dge$samples <- merge(ont_sn.gene.dge$samples %>% select(-group), bulk.meta, by = "row.names", all.x = TRUE) %>% select(-Row.names)

## PB SC
rds_files <- list.files(dir_path, pattern = "pb_sc_.*_featureCounts\\.rds$", full.names = TRUE)
dge_list <- lapply(rds_files, readRDS)
combined_counts <- do.call(cbind, lapply(dge_list, function(x) x$counts))
colnames(combined_counts) <- sub(".sorted.bam", "", colnames(combined_counts)) %>% sub("pb_sc_", "", .)
pb_sc.gene.dge <- DGEList(counts = combined_counts)
pb_sc.gene.dge$samples <- merge(pb_sc.gene.dge$samples %>% select(-group), bulk.meta, by = "row.names", all.x = TRUE) %>% select(-Row.names)

## PB SN
rds_files <- list.files(dir_path, pattern = "pb_sn_.*_featureCounts\\.rds$", full.names = TRUE)
dge_list <- lapply(rds_files, readRDS)
combined_counts <- do.call(cbind, lapply(dge_list, function(x) x$counts))
colnames(combined_counts) <- sub(".sorted.bam", "", colnames(combined_counts)) %>% sub("pb_sn_", "", .)
pb_sn.gene.dge <- DGEList(counts = combined_counts)
pb_sn.gene.dge$samples <- merge(pb_sn.gene.dge$samples %>% select(-group), bulk.meta, by = "row.names", all.x = TRUE) %>% select(-Row.names)



```

## Filter out lowly expressed genes and transcript using `filterByExpr` for each dataset, using default parameters

```{r gene filtering for DE, echo=TRUE, message=FALSE, warning=FALSE, results='hide'}
#GENE 
## 1. get common genes
common_genes <- Reduce(intersect, lapply(list(ill_bulk.gene.dge, ont_bulk.gene.dge, pb_bulk.gene.dge, drna_bulk.gene.dge), rownames))

ill_bulk.gene.dge <- ill_bulk.gene.dge[common_genes, ]
ont_bulk.gene.dge <- ont_bulk.gene.dge[common_genes, ]
pb_bulk.gene.dge <- pb_bulk.gene.dge[common_genes, ]
drna_bulk.gene.dge <- drna_bulk.gene.dge[common_genes, ]

## 2. filter by expression
keep <- filterByExpr(ill_bulk.gene.dge, group = ill_bulk.gene.dge$samples$group) &
          filterByExpr(ont_bulk.gene.dge, group = ont_bulk.gene.dge$samples$group) &
          filterByExpr(pb_bulk.gene.dge, group = pb_bulk.gene.dge$samples$group) &
          filterByExpr(drna_bulk.gene.dge, group = drna_bulk.gene.dge$samples$group)

#table(keep)
ill_bulk.gene.dge <- ill_bulk.gene.dge[keep, ]
ont_bulk.gene.dge <- ont_bulk.gene.dge[keep, ]
pb_bulk.gene.dge <- pb_bulk.gene.dge[keep, ]
drna_bulk.gene.dge <- drna_bulk.gene.dge[keep, ]
```

## Normalise the gene and transcript counts using `calcNormFactors` with `TMM` methods for each dataset, separately for human

```{r TMM normalisation, echo=TRUE, message=FALSE, warning=FALSE, results='hide'}
### 4. Normalise the gene counts using `calcNormFactors` with `TMM` methods for each dataset
ill_bulk.gene.dge <- calcNormFactors(ill_bulk.gene.dge, method = "TMM")
ont_bulk.gene.dge <- calcNormFactors(ont_bulk.gene.dge, method = "TMM")
pb_bulk.gene.dge <- calcNormFactors(pb_bulk.gene.dge, method = "TMM")
drna_bulk.gene.dge <- calcNormFactors(drna_bulk.gene.dge, method = "TMM")
```

# MDS (Merged)

## Merge DGElist

NOTE: The merged DGElist will only be used for the MDS plot. The DE analysis will be done separately for each dataset.

```{r Merge DGEs, include=FALSE}
# Gene
ill_bulk.gene.dge$samples$platform <- "Illumina"
ont_bulk.gene.dge$samples$platform <- "ONT cDNA"
pb_bulk.gene.dge$samples$platform <- "PacBio"
drna_bulk.gene.dge$samples$platform <- "ONT dRNA"
bulk.gene.dge.merge <- cbind(
  ill_bulk.gene.dge,
  ont_bulk.gene.dge,
  pb_bulk.gene.dge,
  drna_bulk.gene.dge
)

```

```{r include=FALSE}
norm_and_mds <- function(dge) {
  dge <- calcNormFactors(dge, method = "TMM")
  mds <- plotMDS(dge, plot = FALSE)
  return(mds)
}
myMDS <- function(dge, mds) {
  # normalisation
  mds_data <- data.frame(
    Dim1 = mds$x,
    Dim2 = mds$y,
    cell.line = dge$sample$sample,
    Cancer.type = dge$sample$group,
    Data.type = dge$sample$platform,
    Sequins.mix = dge$samples$sequins
  )

  p <- ggplot(mds_data, aes(x = Dim1, y = Dim2, color = Cancer.type, shape = Data.type, label=cell.line)) +
    geom_point(size = 3, alpha = 0.6) + # Plot points
    #geom_text_repel(size = 3) +
    labs(title = "MDS plot", 
        x = paste0("Dimension 1 (", round(mds$var.explained[1]*100, 2), "%)" ) , 
        y = paste0("Dimension 2 (", round(mds$var.explained[2]*100, 2), "%)" )) +
    scale_color_viridis_d() + # Set color scale
    theme_minimal() 
  return(p)
}
```

## Gene-level MDS

Here are key observations from the gene-level MDS plots:

1.  **Cancer type clustering:** samples cluster clearly by cancer type (**LUAD, SCLC type a, SCLC type p**), reflecting distinct gene expression profiles.
2.  **Platform type impact:** sequencing platforms (**Illumina, ONT cDNA, ONT dRNA, PacBio)** show slight variability but largely preserve cancer type clustering, indicating that the biological variation observed is largely due to cancer cell-type.
3.  **Variance explained:** The first two dimensions explain **19.25%** and **9.84%** of the variation, respectively, effectively separating sample groups.

```{r echo=FALSE, message=FALSE, warning=FALSE, fig.height=3, fig.width=8 }
mds <- norm_and_mds(bulk.gene.dge.merge[is.human(bulk.gene.dge.merge),])
p1 <- myMDS(bulk.gene.dge.merge[is.human(bulk.gene.dge.merge), ], mds) + ggtitle("Human gene counts")
p2 <- myMDS(bulk.gene.dge.merge[is.human(bulk.gene.dge.merge), ], mds) + ggtitle("Human gene counts")  + 
            aes(color = Data.type, shape = Cancer.type) + 
            scale_color_manual(values = color_palette[c("Illumina", "ONT_1", "ONT_2", "PacBio")] %>% unname())
p1 | p2 
```

# Differential gene expression

## DGE edgeR (FeatureCounts)

DGE analysis is performed on subsets for human genes, each with potentially different library size distributions compared to the entire dataset which included human genes and spikeins (sequins and SIRVs). To ensure consistency:

1.  Filter by Expression: Lowly expressed genes are filtered using `filterByExpr` for each subset of human-genes.
2.  Normalization: Gene counts are normalized using the `calcNormFactors` function in edgeR with the `TMM` method for each subset of human-genes.
3.  Differential Expression: Differential expression is assessed using a GLM-based model with the `glmQLFit` and `glmQLFTest` functions, comparing:

-   SCLC type a vs. LUAD
-   SCLC type p vs. LUAD
-   SCLC type a vs. SCLC type p
-   ***Model: model.matrix(\~0 + group, data = dge\$samples)***
4. Differential expression gene and transcript lists are defined using `DecideTests` with a log2 fold change cutoff of `r params$log2FC_cutoff` and adjusted p<0.05.
```{r echo=TRUE, message=FALSE, warning=FALSE, results='hide'}
DGE.edgeR_human <- function(dge, norm.method = "TMM") {
  # QC and normalization
  filter <- filterByExpr(dge, group = dge$samples$group)
  dge <- dge[filter, , keep.lib.sizes = FALSE]
  
  if (!is.null(norm.method)) {
     dge <- calcNormFactors(dge, method = norm.method)
  }
  
  design <- model.matrix(~0 + group, data = dge$samples)
  dge <- estimateDisp(dge, design)
  qlfit <- glmQLFit(dge, design)
  
  contrasts <- makeContrasts(
    AvsL = groupsclc_a - groupluad,
    PvsL = groupsclc_p - groupluad,
    AvsP = groupsclc_a - groupsclc_p,
    levels = design
  )
  
  AvsL.QLF.test <- glmQLFTest(qlfit, contrast = contrasts[, "AvsL"])
  PvsL.QLF.test <- glmQLFTest(qlfit, contrast = contrasts[, "PvsL"])
  AvsP.QLF.test <- glmQLFTest(qlfit, contrast = contrasts[, "AvsP"])
  
  list(
    AvsL.DE = AvsL.QLF.test,
    PvsL.DE = PvsL.QLF.test,
    AvsP.DE = AvsP.QLF.test 
  )
}

# DEG analsysis on human genes only 
ill_bulk.dge_human.list <- DGE.edgeR_human(ill_bulk.gene.dge[is.human(ill_bulk.gene.dge),])
ont_bulk.dge_human.list <- DGE.edgeR_human(ont_bulk.gene.dge[is.human(ont_bulk.gene.dge),])
drna_bulk.deg_human.list <- DGE.edgeR_human(drna_bulk.gene.dge[is.human(drna_bulk.gene.dge),])
pb_bulk.dge_human.list <- DGE.edgeR_human(pb_bulk.gene.dge[is.human(pb_bulk.gene.dge), ])
```

##Check Expected DE
```{r}
expected_DE <- read_csv(params$expected_de_table)

add_FC <- function(x){
  x %>% 
  left_join(ill_bulk.dge_human.list$AvsL.DE$table %>% 
              rownames_to_column('Gene')  %>% 
              select(Gene, logFC) %>%
              mutate(
                Gene = sub("\\..*", "", Gene)
                ) %>%
              rename(
                Illumina_logFC = logFC
              ), by = "Gene") %>% 
    left_join(ont_bulk.dge_human.list$AvsL.DE$table %>% 
              rownames_to_column('Gene')  %>% 
              select(Gene, logFC) %>%
              mutate(
                Gene = sub("\\..*", "", Gene)
                ) %>%
              rename(
                ONT_cDNA_logFC = logFC
              ), by = "Gene") %>%
    left_join(pb_bulk.dge_human.list$AvsL.DE$table %>% 
              rownames_to_column('Gene')  %>% 
              select(Gene, logFC) %>%
              mutate(
                Gene = sub("\\..*", "", Gene)
                ) %>%
              rename(
                Pacbio_logFC = logFC
              ), by = "Gene") %>% 
    left_join(drna_bulk.deg_human.list$AvsL.DE$table %>% 
              rownames_to_column('Gene')  %>% 
              select(Gene, logFC) %>%
              mutate(
                Gene = sub("\\..*", "", Gene)
                ) %>%
              rename(
                ONT_dRNA_logFC = logFC
              ), by = "Gene")
}

# expected_DE %>%  left_join(mat.AvsL.DE %>% mutate(Gene = sub("\\..*", "", Gene)), by="Gene") %>% add_FC
# expected_DE %>% left_join(mat.PvsL.DE %>% mutate(Gene = sub("\\..*", "", Gene)), by="Gene") %>% add_FC
# expected_DE %>% left_join(mat.AvsP.DE %>% mutate(Gene = sub("\\..*", "", Gene)), by="Gene") %>% add_FC



```

### MA plots
```{r MA plots, fig.height=3, fig.width=13}
library(ggplot2)
library(ggrepel)
plot.my.ma <- function(dge){
  # Get results and DE decisions
  res <- dge$table %>% 
    tibble::rownames_to_column("Gene") %>%
    mutate(Gene=sub("\\..*", "", Gene))
  
  # Join with expected_DE
  highlighted <- expected_DE %>%
    filter(Symbol %in% c("HMGA2", "POU2F3", "ASCL1")) %>%
    select(Gene, Symbol) %>%
    inner_join(res, by = "Gene")
  
  dges <- decideTests(dge)
  
  # Add DE status to results
  res$DE <- as.factor(dges[,1])  # Assuming single contrast
  
  # Subsets
  res_nonDE <- res %>% filter(DE == 0)
  res_DE <- res %>% filter(DE != 0)
  # MA plot
  ggplot() +
    geom_point(data = res_nonDE, aes(x = logCPM, y = logFC), color = "gray", alpha = 0.4) +
    geom_point(data = res_DE, aes(x = logCPM, y = logFC, color = DE), alpha = 0.6) +
    scale_color_manual(values = c("-1" = "blue", "1" = "red")) +
    geom_point(data = highlighted, aes(x = logCPM, y = logFC), color = "yellow", shape = 21, fill = NA, size = 2, stroke = 1) +
    geom_text_repel(data = highlighted, aes(x = logCPM, y = logFC, label = Symbol), size = 3, max.overlaps = 100) +
    geom_hline(yintercept = 0, linetype = "dashed", color = "gray") +
    labs(x = "Average log CPM", y = "logâ‚‚ Fold Change", color = "DE Status") +
    theme_minimal()
}
ill_bulk.dge_human.list %>% map(plot.my.ma) %>% patchwork::wrap_plots(ncol=3, guides='collect')
ont_bulk.dge_human.list %>% map(plot.my.ma) %>% patchwork::wrap_plots(ncol=3, guides='collect')
drna_bulk.deg_human.list %>% map(plot.my.ma) %>% patchwork::wrap_plots(ncol=3, guides='collect')
pb_bulk.dge_human.list %>% map(plot.my.ma) %>% patchwork::wrap_plots(ncol=3, guides='collect')


```


## Cross-platform comparison DGE human (upset)

Here the differential gene expression (DGE) results visualized in the upset plots show:

1.  ***SCLC_A vs LUAD:*** 580 shared differential expressed genes across all platforms. Highlighting platform-specific contributions to DGE detection with largest intersections from PacBio (790 genes).

2.  ***SCLC_P vs LUAD:*** The total number of differential expressed genes is smaller compared to SCLC_A vs LUAD. 140 shared differential expressed genes across all platforms. Highlighting platform-specific contributions to DGE detection with largest intersections from PacBio (173 genes).

3.  ***SCLC_A vs SCLC_P:*** A significantly smaller number of genes are differential expressed compared to LUAD comparisons. 16 shared differential expressed genes across all platforms. Highlighting platform-specific contributions to DGE detection with largest intersections from Illumina (48 genes).

***Overall, the results suggest:*** PacBio tends to identify highest number of platform-specific differential expressed genes. Differences between cancer subtypes (SCLC_A vs SCLC_P) are less pronounced compared to differences with LUAD.

```{r upset plot function SC, include=FALSE}

complex_upset_plot <- function(lst) {
  # apply decideTests to each list element
  DEGs <- lapply(lst, function(x) {
    x <- x %>% decideTests(lfc=params$log2FC_cutoff)
    x <- rownames(x)[x != 0]
    return(x)
  })

  long_df <- stack(DEGs)

  # Make the binary membership matrix
  binary_matrix <- long_df %>%
    mutate(present = 1) %>%
    pivot_wider(names_from = ind, values_from = present, values_fill = 0)

  p <- ComplexUpset::upset(
    binary_matrix %>% select(-values) %>% as.data.frame(),
    intersect = names(lst),
    base_annotations=list(
        'Intersection size'=ComplexUpset::intersection_size(
            counts=TRUE,
            text=element_text(size = 3, vjust=-0.2)
        )
    ),
    wrap = TRUE
  )

  return(list(
    plot = p,
    matrix = binary_matrix %>% rename(Gene = values)
  ))
}
```

```{r upset plot dge human, echo=FALSE, fig.height=9, fig.width=8}
# Upset plot
lst <- list(
  "Illumina" = ill_bulk.dge_human.list$AvsL.DE,
  "ONT cDNA" = ont_bulk.dge_human.list$AvsL.DE,
  "PacBio" = pb_bulk.dge_human.list$AvsL.DE,
  "ONT dRNA" = drna_bulk.deg_human.list$AvsL.DE
)
p1 <- complex_upset_plot(lst)$plot + ggtitle("SCLC_A vs LUAD") + theme(plot.title = element_text(size =10))
mat.AvsL.DE <- complex_upset_plot(lst)$matrix

lst <- list(
  "Illumina" = ill_bulk.dge_human.list$PvsL.DE,
  "ONT cDNA" = ont_bulk.dge_human.list$PvsL.DE,
  "PacBio" = pb_bulk.dge_human.list$PvsL.DE,
  "ONT dRNA" = drna_bulk.deg_human.list$PvsL.DE
)
p2 <- complex_upset_plot(lst)$plot + ggtitle("SCLC_P vs LUAD") + theme(plot.title = element_text(size = 10))
mat.PvsL.DE <- complex_upset_plot(lst)$matrix

lst <- list(
  "Illumina" = ill_bulk.dge_human.list$AvsP.DE,
  "ONT cDNA" = ont_bulk.dge_human.list$AvsP.DE,
  "PacBio" = pb_bulk.dge_human.list$AvsP.DE,
  "ONT dRNA" = drna_bulk.deg_human.list$AvsP.DE
)

p3 <- complex_upset_plot(lst)$plot + ggtitle("SCLC_A vs SCLC_P") + theme(plot.title = element_text(size = 10))
mat.AvsP.DE <- complex_upset_plot(lst)$matrix
(p1 / p2 / p3) + patchwork::plot_annotation(
  title = "DGE Human",
  theme = theme(plot.title = element_text(size =15))
)
```

```{r gene: platform specific detection summary, fig.height=2, fig.width=13}
# Pivot to long format
platform_specific_gene_summary <- function(df) {
  df_long <- df %>%
  pivot_longer(-Gene, names_to = "Platform", values_to = "Detected") %>%
  group_by(Gene) %>%
  mutate(TotalDetected = sum(Detected)) %>%
  ungroup() %>%
  mutate(
    Category = case_when(
      Detected == 1 & TotalDetected == 1 ~ "Platform-Specific Detection",  # Only detected by this platform
      Detected == 0 & TotalDetected == 3 ~ "Platform-Specific Absence",      # Not detected by this platform but present elsewhere
      Detected == 1 & TotalDetected > 1 ~ "Common Detection (also detected in >0 other platform)",
      Detected == 0 & TotalDetected < 3 ~ "Common Absence (also absent in >0 other platform)" # Detected by this platform + at least one other
    )
  ) %>%
  count(Platform, Category)

# Convert to factor for proper ordering in plot
df_long$Category <- factor(df_long$Category, 
                           levels = c(
                                      "Platform-Specific Absence", 
                                      "Common Absence (also absent in >0 other platform)",
                                      "Platform-Specific Detection", 
                                      "Common Detection (also detected in >0 other platform)"
                                      ))

# Stacked bar plot
df_long %>%
  mutate(n = case_when(
    Category %in% c("Platform-Specific Absence", "Common Absence (also absent in >0 other platform)") ~ -n,
    TRUE ~ n
  )) %>%
  ggplot(aes(x = Platform, y = n, fill = Category)) +
  geom_bar(stat = "identity") +
  labs(
    y = "Number of Genes",
    x = "Platform",
    fill = "Category"
  ) +
  theme_minimal() +
  scale_fill_manual(
    values = c(
      "Platform-Specific Detection" = "#0072B2",  # Blue (color-blind friendly)
      "Platform-Specific Absence" = "#E69F00",    # Orange (color-blind friendly)
      "Common Detection (also detected in >0 other platform)" = "#009E73",  # Green
      "Common Absence (also absent in >0 other platform)" = "#999999"  # Gray
    )
  ) +
  coord_flip() +  # Flip axes for horizontal bars
  theme(
    axis.text.x = element_text(angle = 0, hjust = 0.5, vjust = 0.5),  # Ensure x-axis labels are readable
    axis.title.x = element_text(size = 12),  # Adjust x-axis title size
    axis.text.y = element_text(size = 10)   # Adjust y-axis text size if needed
  )
}


map(list(mat.AvsL.DE, mat.PvsL.DE, mat.AvsP.DE), platform_specific_gene_summary) %>% do.call(patchwork::wrap_plots, .) +
  patchwork::plot_layout(guides = 'collect')
```


```{r}
# Save the de gene list to bulk_rst
bulk_rst[["AvsP.DEG.all"]] <- mat.AvsP.DE %>% 
  filter(rowSums(select(., where(is.numeric))) == 4) %>% pull(Gene)
bulk_rst[["AvsP.DEG.any"]] <- mat.AvsP.DE %>% 
  filter(rowSums(select(., where(is.numeric))) > 0) %>% pull(Gene)

bulk_rst[["AvsL.DEG.all"]] <- mat.AvsL.DE %>% 
  filter(rowSums(select(., where(is.numeric))) == 4) %>% pull(Gene)
bulk_rst[["AvsL.DEG.any"]] <- mat.AvsL.DE %>% 
  filter(rowSums(select(., where(is.numeric))) > 0) %>% pull(Gene)

bulk_rst[["PvsL.DEG.all"]] <- mat.PvsL.DE %>% 
  filter(rowSums(select(., where(is.numeric))) == 4) %>% pull(Gene)
bulk_rst[["PvsL.DEG.any"]] <- mat.PvsL.DE %>% 
  filter(rowSums(select(., where(is.numeric))) > 0) %>% pull(Gene)
```


# PseudoBulk
```{r}
color_palette <- c(
  PacBio = "#df1995",
  PacBio_1 = "#ff9b9b",
  PacBio_2 = "#a7129a",
  ONT = "#00789b",
  ONT_1 = "#24cdcd",
  ONT_2 = "#04476c",
  Illumina = "#e88b20"
)
plot_df <- tibble(
  "ONT SC" = ont_sc.gene.dge$counts %>% sum(),
  "ONT SN" = ont_sn.gene.dge$counts %>% sum(),
  "PacBio SC" = pb_sc.gene.dge$counts %>% sum(),
  "PacBio SN" = pb_sn.gene.dge$counts %>% sum()
) %>% pivot_longer(everything(), names_to = "Dataset", values_to = "Total Count")

plot_df$Dataset <- factor(plot_df$Dataset, levels = c("ONT SC", "ONT SN", "PacBio SC", "PacBio SN"))

ggplot(plot_df, aes(x = Dataset, y = `Total Count` / 1000000, fill = Dataset)) +
  geom_bar(stat = "identity", position = position_dodge()) +
  theme_minimal() +
  labs(x = "Protocols", y = "Total Count (Million)") +
  scale_fill_manual(values = color_palette[c( "ONT_1", "ONT_2", "PacBio", "PacBio_1")] %>% unname()) + ggtitle("Gene-level total counts")

```
```{r Merge DGEs SC}
# Gene
ont_sc.gene.dge$samples$platform <- "ONT"
ont_sn.gene.dge$samples$platform <- "ONT"
pb_sc.gene.dge$samples$platform <- "PacBio"
pb_sn.gene.dge$samples$platform <- "PacBio"

ont_sc.gene.dge$samples$library <- "SC"
ont_sn.gene.dge$samples$library <- "SN"
pb_sc.gene.dge$samples$library <- "SC"
pb_sn.gene.dge$samples$library <- "SN"

gene.dge.merge <- cbind(
  ont_sc.gene.dge,
  ont_sn.gene.dge,
  pb_sc.gene.dge,
  pb_sn.gene.dge
)
```

```{r}
norm_and_mds <- function(dge) {
  dge <- calcNormFactors(dge, method = "TMM")
  mds <- plotMDS(dge, plot = FALSE)
  return(mds)
}
myMDS <- function(dge, mds) {
  # normalisation
  mds_data <- tibble(
    Dim1 = mds$x,
    Dim2 = mds$y,
    "Cell lines" = dge$sample$sample,
    "Cancer types" = dge$sample$group,
    Platforms = dge$sample$platform,
    "Library types" = dge$sample$library
  )

  p <- ggplot(mds_data, aes(x = Dim1, y = Dim2, color =  `Cancer types`, shape = `Cell lines`)) +
    geom_point(size = 3, alpha = 0.6) + # Plot points +
    labs(title = "MDS plot", 
        x = paste0("Dimension 1 (", round(mds$var.explained[1]*100, 2), "%)" ) , 
        y = paste0("Dimension 2 (", round(mds$var.explained[2]*100, 2), "%)" )) +
    scale_color_viridis_d() + # Set color scale
    theme_minimal() 
  return(p)
}
```
## Gene count MDS
```{r Gene count MDS, fig.height=3, fig.width=8}
mds <- norm_and_mds(gene.dge.merge) 
p1 <- myMDS(gene.dge.merge, mds) + ggtitle("Human gene counts") +
  scale_shape_manual(values = c(7, 8, 9, 15, 16, 17, 18, 1))
p2 <- myMDS(gene.dge.merge, mds) + ggtitle("Human gene counts")  + 
            aes(color = Platforms, shape = `Library types`) + 
            scale_color_manual(values = color_palette[c("ONT",  "PacBio")] %>% unname())
p1 | p2 
```
## PVCA plots
```{r PVCA plots, fig.height=3, fig.width=4}
# Create a design matrix for covariates
library(pvca)
library(Biobase)


# Gene
log_cpm <- cpm(gene.dge.merge, log = TRUE, prior.count = 1) 
# Run PVCA
colnames(log_cpm) <- NULL
expr_set <- ExpressionSet(
  assayData = as.matrix(log_cpm),
  phenoData = AnnotatedDataFrame(gene.dge.merge$samples %>% rename(Platforms=platform, `Cancer_types`=group,  `Library_types`=library))
)
pvca_result <- pvcaBatchAssess(
  expr_set,
  batch.factors = c("Platforms", "Cancer_types", "Library_types"),  # Specify the metadata columns to test
  threshold = 0.1
)


# Create a data frame from the pvca_result list
plot_df <- as.data.frame(pvca_result$dat %>% t )
rownames(plot_df) <- pvca_result$label
plot_df <- plot_df %>% rownames_to_column("Variable") %>% rename("Value" = V1) %>% mutate(type="Gene-level")

# Create a bar plot using ggplot
ggplot(plot_df, aes(x=type, y = Value, fill = Variable)) +
  geom_bar(stat = "identity", position = "fill") +
  labs( x = "", y = "Weighted average proportion variance") +
  theme_minimal() +
  scale_fill_brewer(palette = "Set3") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

```





## Differential gene expression
### DGE edgeR (Salmon and Oarfish)
```{r SC MA Plots, fig.height=3, fig.width=13}
DGE.edgeR_human <- function(dge, norm.method = "TMM") {
  if (!is.null(norm.method)) {
    dge <- calcNormFactors(dge, method = norm.method)
  }
  
  design <- model.matrix(~0 + group, data = dge$samples)
  dge <- estimateDisp(dge, design)
  qlfit <- glmQLFit(dge, design)
  
  contrasts <- makeContrasts(
    AvsL = groupsclc_a - groupluad,
    PvsL = groupsclc_p - groupluad,
    AvsP = groupsclc_a - groupsclc_p,
    levels = design
  )
  
  AvsL.QLF.test <- glmQLFTest(qlfit, contrast = contrasts[, "AvsL"])
  PvsL.QLF.test <- glmQLFTest(qlfit, contrast = contrasts[, "PvsL"])
  AvsP.QLF.test <- glmQLFTest(qlfit, contrast = contrasts[, "AvsP"])
  
  list(
    AvsL.DE = AvsL.QLF.test,
    PvsL.DE = PvsL.QLF.test,
    AvsP.DE = AvsP.QLF.test 
  )
}

ont_sc.dge_human.list <- DGE.edgeR_human(ont_sc.gene.dge)
ont_sn.dge_human.list <- DGE.edgeR_human(ont_sn.gene.dge)
pb_sc.dge_human.list <- DGE.edgeR_human(pb_sc.gene.dge)
pb_sn.dge_human.list <- DGE.edgeR_human(pb_sn.gene.dge)

ont_sc.dge_human.list %>% map(plot.my.ma) %>% patchwork::wrap_plots(ncol=3, guides='collect')
ont_sn.dge_human.list %>% map(plot.my.ma) %>% patchwork::wrap_plots(ncol=3, guides='collect')
pb_sc.dge_human.list %>% map(plot.my.ma) %>% patchwork::wrap_plots(ncol=3, guides='collect')
pb_sn.dge_human.list %>% map(plot.my.ma) %>% patchwork::wrap_plots(ncol=3, guides='collect')
```

### Cross-platform comparison (upset)
```{r upset plot function}
library(ComplexUpset)
complex_upset_plot <- function(lst, bulk.any = NULL, bulk.all = NULL) {
  # apply decideTests to each list element
  DEGs <- lapply(lst, function(x) {
    x <- x %>% decideTests()
    x <- rownames(x)[x != 0]
    return(x)
  })
  DEGs[["Bulk DE any"]] <- bulk.any
  DEGs[["Bulk DE all"]] <- bulk.all
  long_df <- stack(DEGs)

  # Make the binary membership matrix
  binary_matrix <- long_df %>%
    mutate(present = 1) %>%
    pivot_wider(names_from = ind, values_from = present, values_fill = 0)
  
  binary_matrix$`Bulk DE any` <- binary_matrix$`Bulk DE any` %>% as.logical()
  binary_matrix$`Bulk DE all` <- binary_matrix$`Bulk DE all` %>% as.logical()
  binary_matrix$`Bulk DE` <- factor(
    ifelse(!binary_matrix$`Bulk DE any`, "Not DE in any bulk dataset",
      ifelse(
        binary_matrix$`Bulk DE all`, "DE in all bulk datasets", "DE in any bulk dataset"
      )
    ),
    levels = c("Not DE in any bulk dataset", "DE in any bulk dataset", "DE in all bulk datasets")
  )

  # remove the bulk Bulk DE levels


  p <- ComplexUpset::upset(
    binary_matrix %>% select(-values) %>% as.data.frame(),
    intersect = names(lst),
    wrap = TRUE,
    base_annotations=list(
      "Intersection size" = intersection_size(
        mapping = aes(fill = `Bulk DE`)
      ) + scale_fill_manual(
                values = c(
                  "Not DE in any bulk dataset" = "grey",
                  "DE in any bulk dataset" = "#84d28e", 
                  "DE in all bulk datasets" = "#27ca0b"
                )
            ) + theme(legend.position = "none")
    ),
    set_sizes=(
        upset_set_size(
            geom=geom_bar(
                aes(fill=`Bulk DE`, x=group),
                width=0.8
            ),
            position = 'right',
        )  + scale_fill_manual(
                values = c(
                  "Not DE in any bulk dataset" = "grey",
                  "DE in any bulk dataset" = "#84d28e", 
                  "DE in all bulk datasets" = "#27ca0b"
                )
            ) 
    ),
    guides='over'
  )

  return(list(
    plot = p,
    matrix = binary_matrix %>% rename(Gene = values)
  ))
}
```

```{r upset plot dge human SC,  fig.height=10, fig.width=8}
# Upset plot
lst <- list(
  "ONT SC" = ont_sc.dge_human.list$AvsL.DE,
  "ONT SN" = ont_sn.dge_human.list$AvsL.DE,
  "PacBio SC" = pb_sc.dge_human.list$AvsL.DE,
  "PacBio SN" = pb_sn.dge_human.list$AvsL.DE
)
bulk.any <- bulk_rst$AvsL.DEG.any
bulk.all <- bulk_rst$AvsL.DEG.all
p1 <- complex_upset_plot(lst, bulk.any, bulk.all)$plot + ggtitle("SCLC_A vs LUAD") + theme(plot.title = element_text(size =20))
mat.AvsL.DGE <- complex_upset_plot(lst, bulk.any, bulk.all)$matrix

lst <- list(
  "ONT SC" = ont_sc.dge_human.list$PvsL.DE,
  "ONT SN" = ont_sn.dge_human.list$PvsL.DE,
  "PacBio SC" = pb_sc.dge_human.list$PvsL.DE,
  "PacBio SN" = pb_sn.dge_human.list$PvsL.DE
)
bulk.any <- bulk_rst$PvsL.DEG.any
bulk.all <- bulk_rst$PvsL.DEG.all
p2 <- complex_upset_plot(lst, bulk.any, bulk.all)$plot + ggtitle("SCLC_P vs LUAD") + theme(plot.title = element_text(size = 20))
mat.PvsL.DGE <- complex_upset_plot(lst, bulk.any, bulk.all)$matrix

lst <- list(
  "ONT SC" = ont_sc.dge_human.list$AvsP.DE,
  "ONT SN" = ont_sn.dge_human.list$AvsP.DE,
  "PacBio SC" = pb_sc.dge_human.list$AvsP.DE,
  "PacBio SN" = pb_sn.dge_human.list$AvsP.DE
)
bulk.any <- bulk_rst$AvsP.DEG.any
bulk.all <- bulk_rst$AvsP.DEG.all
p3 <- complex_upset_plot(lst, bulk.any, bulk.all)$plot + ggtitle("SCLC_A vs SCLC_P") + theme(plot.title = element_text(size = 20))
mat.AvsP.DGE <- complex_upset_plot(lst, bulk.any, bulk.all)$matrix
(p1 / p2 / p3) + patchwork::plot_annotation(
  title = "DGE Human",
  theme = theme(plot.title = element_text(size =25))
)
```



