---
title: "LongBench Single-cell Sing-cell DE analysis"
author: "Yupei"
version: "v1.2"
output:
  html_document:
    code_folding: hide
    toc: yes
    toc_float: yes
    number_sections: yes
    toc_depth: 4
  pdf_document:
    toc: yes
    toc_depth: '4'
params:
  random_seed: 2024
  cache_dir: NULL
  ont_sc_dir: "/vast/projects/LongBench/analysis/lr_sc_sn/result/PseudoBulkOarfishCov/ont_sc"
  ont_sn_dir: "/vast/projects/LongBench/analysis/lr_sc_sn/result/PseudoBulkOarfishCov/ont_sn"
  pb_sc_dir: "/vast/projects/LongBench/analysis/lr_sc_sn/result/PseudoBulkOarfishCov/pb_sc"
  pb_sn_dir: "/vast/projects/LongBench/analysis/lr_sc_sn/result/PseudoBulkOarfishCov/pb_sn"
  human_gtf: '/vast/projects/LongBench/reference_files/GRCh38/gencode.v44.annotation.gtf'
  bulk_meta: "/vast/projects/LongBench/sequencing_data/illumina_bulk/metadata.txt"
  bulk_ident_rds: "/vast/projects/LongBench/analysis/workflow/rmarkdown/RDS/bulk_identification.rds"
  bulk_de_rds: "/vast/projects/LongBench/analysis/workflow/rmarkdown/RDS/bulk_DE.rds"
  expected_de_table: '/vast/projects/LongBench/reference_files/Expected_DE.csv'
  fig.path: NULL
---

# Setup: Create DGE from input
```{r setup, include=FALSE}
knitr::opts_chunk$set(
  cache = !is.null(params$cache_dir),  # Enable caching only if cache_dir is provided
  fig.width = 15,                     # Set default figure width
  fig.height = 10,                    # Set default figure height
  warning = FALSE,                    # Suppress warnings in the output
  fig.path = params$fig.path,         # Set figure path (NULL if not provided)
  dev = if (!is.null(params$fig.path)) c("png", "svg") else NULL  # Set output devices if fig.path is provided
)

# Set cache.path only if params$cache_dir is not NULL
if (!is.null(params$cache_dir)) {
  knitr::opts_chunk$set(cache.path = params$cache_dir)
}
```

```{r Libraries,include=FALSE}
library(limma)
library(edgeR)
library(tidyverse)
library(RColorBrewer)
library(dplyr)
library(tidyr)
library(ggplot2)
library(ggrepel)
library(glue)
library(ggplot2)
library(gridExtra)
library(ggExtra)
library(knitr)
library(kableExtra)
library(rtracklayer)
library(scales)
library(UpSetR)
library(ggtree)
library(aplot)
library(DTUrtle)
source("/home/users/allstaff/you.yu/LongBench/analysis/workflow/scripts/Rfunctions.R")

# setup some conflict functions
rename <- dplyr::rename
select <- dplyr::select
```



```{r color code, include=FALSE, eval=T}
color_palette <- c(
  PacBio = "#df1995",
  PacBio_1 = "#ff9b9b",
  PacBio_2 = "#a7129a",
  ONT = "#00789b",
  ONT_1 = "#24cdcd",
  ONT_2 = "#04476c",
  Illumina = "#e88b20"
)
```


```{r, get number of transcripts gene from gtf}
calcTxNum <- function(G.Tx.map){
  geneid = rep(NA, nrow(G.Tx.map))
  txcount = table(G.Tx.map$gene_id)
  txnum = as.numeric(txcount[match(G.Tx.map$gene_id, names(txcount))])
}

# Human
human_tx  <- GenomicFeatures::makeTxDbFromGFF(params$human_gtf, format = "gtf") %>% GenomicFeatures::transcripts(columns = c("tx_id", "tx_name", "gene_id"))
human.G.Tx.map <- tibble(
  gene_id = human_tx$gene_id %>% unlist,
  tx_name = human_tx$tx_name
)
rm(human_tx)
human.G.Tx.map$tx_count <- human.G.Tx.map %>% calcTxNum
```

```{r Create DGE from input, echo=F}
bulk.meta <- read.csv(params$bulk_meta)
bulk_ident_rds <- readRDS(params$bulk_ident_rds)
bulk_de_rds <- readRDS(params$bulk_de_rds)
rownames(bulk.meta) <- bulk.meta$sample

# SR salmon
# ill_bulk.dge <- get_dge_from_salmon(params$ill_bulk_salmon_dir, ".*/")

# LR oarfish
ont_sc.dge <- get_dge_from_oarfish(params$ont_sc_dir, ".*/")
ont_sn.dge <- get_dge_from_oarfish(params$ont_sn_dir, ".*/")
pb_sc.dge <- get_dge_from_oarfish(params$pb_sc_dir, ".*/")
pb_sn.dge <- get_dge_from_oarfish(params$pb_sn_dir, ".*/")
```

### Aggregate transcript counts to gene level using `tximport` with unfiltered transcript counts
```{r, echo=F}
library(tximport) # install latest version from github

# Extract tx_name and gene_id columns from human and sequins
human_tx2gene <- human.G.Tx.map[, c("tx_name", "gene_id")]

# Get quant.sf file paths
get_quant_paths <- function(dirs, file_name) {
  # Generate file paths for quantification files
  quant_paths <- file.path(dirs, file_name)
  return(quant_paths)
}

# Define directories with their corresponding expected file names
oarfish_dirs <- list(
  ont_sc = list(dir = params$ont_sc_dir, file_name = ".quant"),
  ont_sn = list(dir = params$ont_sn_dir, file_name = ".quant"),
  pb_sc = list(dir = params$pb_sc_dir, file_name = ".quant"),
  pb_sn = list(dir = params$pb_sn_dir, file_name = ".quant")
)

# Apply function to each directory to get all quant file paths
quant_paths <- lapply(oarfish_dirs, function(dir_info) {
  # Get the directory and file name from dir_info
  dir <- dir_info$dir
  file_name <- dir_info$file_name
  
  # List all subdirectories and get the quant file paths
  sub_dirs <- list.dirs(dir, full.names = TRUE, recursive = FALSE)
  get_quant_paths(sub_dirs, file_name)
})

# Flatten the list of quant_paths into a named list
quant_paths <- lapply(quant_paths, unlist)

# Function to import DGE from tximport
get_dge_from_txi <- function(quant_dir, sample_prefix_regex, type) {
  # Import data using tximport
  txi <- tximport(quant_dir,
                  type = type,
                  tx2gene = human_tx2gene,
                  ignoreAfterBar = TRUE,
                  countsFromAbundance = "no") # raw counts
  counts <- as.data.frame(txi$counts, stringAsFactors = FALSE)
  sample_names <- gsub(sample_prefix_regex, '', basename(dirname(quant_dir)))
  colnames(counts) <- sample_names
  rst.dge <- DGEList(counts = counts)
  rst.dge$genes <- data.frame(Length=txi$length %>% rowMeans())
  rownames(rst.dge$samples) <- sub(sample_prefix_regex, '', rownames(rst.dge$samples))
  colnames(rst.dge$counts) <- sub(sample_prefix_regex, "", colnames(rst.dge$counts))
  # colnames(rst.dge$gene$Length) <- sub(sample_prefix_regex, "", colnames(rst.dge$gene$Length))
  
  rst.dge$samples <- merge(rst.dge$samples %>% select(-group), bulk.meta, by = "row.names", all.x = TRUE) %>% select(-Row.names)
  rst.dge$samples <- rst.dge$samples[match(colnames(rst.dge$counts), rst.dge$samples$sample),]
  return(rst.dge)
}

# Process each quant path for different bulk types
ont_sc.gene.dge <- get_dge_from_txi(quant_paths$ont_sc, ".*/", "oarfish")
ont_sn.gene.dge <- get_dge_from_txi(quant_paths$ont_sn, ".*/", "oarfish")
pb_sc.gene.dge <- get_dge_from_txi(quant_paths$pb_sc, ".*/", "oarfish")
pb_sn.gene.dge <- get_dge_from_txi(quant_paths$pb_sn, ".*/", "oarfish")

# save all workpalce as Rdata
# save.image("/vast/scratch/users/you.yu/test/SC_DE_line197.Rdata")
```
### Plot: total Gene and transcript count
```{r sc total count plot, fig.height=3, fig.width=8}

plot_df <- tibble(
  "ONT SC" = ont_sc.dge$counts %>% sum(),
  "ONT SN" = ont_sn.dge$counts %>% sum(),
  "PacBio SC" = pb_sc.dge$counts %>% sum(),
  "PacBio SN" = pb_sn.dge$counts %>% sum()
) %>% pivot_longer(everything(), names_to = "Dataset", values_to = "Total Count")

plot_df$Dataset <- factor(plot_df$Dataset, levels = c("ONT SC", "ONT SN", "PacBio SC", "PacBio SN"))

p.tx <- ggplot(plot_df, aes(x = Dataset, y = `Total Count` / 1000000, fill = Dataset)) +
  geom_bar(stat = "identity", position = position_dodge()) +
  theme_minimal() +
  labs(x = "Protocols", y = "Total Count (Million)") +
  scale_fill_manual(values = color_palette[c( "ONT_1", "ONT_2", "PacBio", "PacBio_1")] %>% unname()) + ggtitle("Transcript-level total counts")

plot_df <- tibble(
  "ONT SC" = ont_sc.gene.dge$counts %>% sum(),
  "ONT SN" = ont_sn.gene.dge$counts %>% sum(),
  "PacBio SC" = pb_sc.gene.dge$counts %>% sum(),
  "PacBio SN" = pb_sn.gene.dge$counts %>% sum()
) %>% pivot_longer(everything(), names_to = "Dataset", values_to = "Total Count")

plot_df$Dataset <- factor(plot_df$Dataset, levels = c("ONT SC", "ONT SN", "PacBio SC", "PacBio SN"))

p.gene <- ggplot(plot_df, aes(x = Dataset, y = `Total Count` / 1000000, fill = Dataset)) +
  geom_bar(stat = "identity", position = position_dodge()) +
  theme_minimal() +
  labs(x = "Protocols", y = "Total Count (Million)") +
  scale_fill_manual(values = color_palette[c( "ONT_1", "ONT_2", "PacBio", "PacBio_1")] %>% unname()) + ggtitle("Gene-level total counts")

patchwork::wrap_plots(p.gene, p.tx) + 
  patchwork::plot_layout(guides = 'collect')
```

# Pseudobulk read count distribution:
```{r, fig.height=4, fig.width=10}
merged_data <- read_csv("/vast/projects/LongBench/analysis/lr_sc_sn/result/PseudoBulkQC/pseudo_bulk_read_count.csv")
```


```{r pseudobulked read count, fig.height=4, fig.width=10}
merged_data %>%
  ggplot(aes(x = sample, y = Count, fill = platform, colour = protocol)) +
  geom_bar(stat = "identity") +
  geom_text(aes(label = paste0(round(Count/1000000, 2), 'M')), hjust = 0, size = 4, angle = 90) +  # Add labels
  theme(axis.text.x = element_text(angle = 60, hjust = 1)) +
  labs(title = "Read count comparison", x = "Datasets", y = "Read counts")  +
  scale_fill_manual(values = color_palette[c("ONT", "PacBio")] %>% unname) +
  scale_color_manual(values = c('#a2b346', '#565750')) + ylim(0, 4.2e7)
```


```{r pseudobulked read count total, fig.height=4, fig.width=6}
merged_data %>%
  group_by(protocol, platform) %>%
  summarise(Count = sum(Count), .groups = "drop") %>%
  mutate(sample = paste(platform, protocol, sep = "_")) %>% 
  ggplot(aes(x = sample, y = Count, fill = platform, colour = protocol)) +
  geom_bar(stat = "identity") +
  geom_text(aes(label = paste0(round(Count/1000000, 2), 'M')), hjust = 0.5, vjust = -1, size = 4, angle = 0) +  # Add labels
  #theme(axis.text.x = element_text(angle = 0, hjust = 1)) +
  labs(title = "Read count comparison", x = "Datasets", y = "Read counts")  +
  scale_fill_manual(values = color_palette[c("ONT", "PacBio")] %>% unname) +
  scale_color_manual(values = c('#a2b346', '#565750')) + ylim(0, 1.5e8)
```



# Identification analysis:
## Gene level 
Gene identification: Filter out lowly expressed genes and transcript using `filterByExpr` for each dataset, using default parameters
```{r gene filtering for DE, fig.height=4, fig.width=8}
#GENE 
## 1. get common genes
common_genes <- Reduce(intersect, lapply(list(ont_sc.gene.dge, ont_sn.gene.dge, pb_sc.gene.dge, pb_sn.gene.dge), rownames))

ont_sc.gene.dge <- ont_sc.gene.dge[common_genes, ]
ont_sn.gene.dge <- ont_sn.gene.dge[common_genes, ]
pb_sc.gene.dge <- pb_sc.gene.dge[common_genes, ]
pb_sn.gene.dge <- pb_sn.gene.dge[common_genes, ]

gene_discovery <- tibble(
  gene_id = common_genes,
  "ONT SC" = filterByExpr(ont_sc.gene.dge, group = ont_sc.gene.dge$samples$group),
  "ONT SN" = filterByExpr(ont_sn.gene.dge, group = ont_sn.gene.dge$samples$group),
  "PacBio SC" = filterByExpr(pb_sc.gene.dge, group = pb_sc.gene.dge$samples$group),
  "PacBio SN" = filterByExpr(pb_sn.gene.dge, group = pb_sn.gene.dge$samples$group)
)


tx_ref_df <- ont_sc.dge$genes %>%
  rownames_to_column("tx_id") %>%
  select(tx_id, Length) %>%
  tibble() %>%
  mutate(
    biotype = strsplit(tx_id, "\\|") %>% sapply(function(x) x[8]),
    gene_id = strsplit(tx_id, "\\|") %>% sapply(function(x) x[2]),
    tx_id = strsplit(tx_id, "\\|") %>% sapply(function(x) x[1])
  )
gene_length_lookup <- tx_ref_df %>%
    group_by(gene_id) %>%
    summarise(mean_length = mean(Length, na.rm = TRUE), .groups = "drop")

# Connect to Ensembl
library(biomaRt)
ensembl <- useEnsembl(biomart = "genes", dataset = "hsapiens_gene_ensembl")

gene_discovery <- gene_discovery %>%
    left_join(gene_length_lookup, by = c("gene_id")) %>%
    rename(`Gene length` = mean_length)
gene_discovery$gene_id <- gene_discovery$gene_id %>% sub("\\..*", "", .)
biotypes <- getBM(
  attributes = c("ensembl_gene_id", "gene_biotype"),
  filters = "ensembl_gene_id",
  values = gene_discovery$gene_id,
  mart = ensembl
)


gene_discovery <- gene_discovery %>%
  left_join(biotypes, by = c("gene_id" = "ensembl_gene_id")) %>%
  rename(biotype = gene_biotype) %>%  
  mutate(biotype = ifelse(biotype %in% 
            names(sort(table(biotype), decreasing = TRUE)[1:3]), 
            biotype, "Others"))


gene_discovery$biotype <- factor(gene_discovery$biotype, levels = c(sort(unique(gene_discovery$biotype)) %>% setdiff("Others"), "Others"))
```

### Upset plots
```{r, fig.height=4, fig.width=8}
library(ComplexUpset)
ComplexUpset::upset(
  gene_discovery %>% mutate(
    biotype = case_when(
      biotype =="protein_coding" ~ "Protein-coding",
      TRUE ~ "Non-protein-coding"
    ) 
  ) %>% filter(
      `ONT SC` | `ONT SN` |  `PacBio SC` | `PacBio SN`
    ),
  intersect = c("ONT SC", "ONT SN", "PacBio SC", "PacBio SN"),
  annotations = list(
        # 1st method - passing list:
        "Transcript length" = list(
          aes = aes(x = intersection, y = `Gene length`),
          # provide a list if you wish to add several geoms
          geom = geom_boxplot(na.rm = TRUE, outliers = FALSE)
        )
        # ,
        # "BioType" = list(
        #   aes = aes(x = intersection, fill = biotype),
        #   geom = geom_bar(position = "fill")
        # )
  ),
  base_annotations=list(
        'Intersection size'=ComplexUpset::intersection_size(
            counts=TRUE,
            mapping=aes(fill=biotype),
            text=element_text(size = 3, vjust=-0.2)
        ) + scale_fill_manual(values = c("Non-protein-coding"="#BAC5CB",
                              "Protein-coding"='#B5C952'))
    ),
  set_sizes=(
        ComplexUpset::upset_set_size(
            geom=geom_bar(
                aes(fill=biotype),
                width=0.8
            ) ,
            position='right'
        )+ scale_fill_manual(values = c("Non-protein-coding"="#BAC5CB",
                              "Protein-coding"='#B5C952')) + 
          theme(legend.position = "none")
  ),
  guides='over',
  wrap = TRUE
)
```

### Upset plots: Overlap with Bulk
```{r Gene Identification Upset plot Overlap with Bulk, fig.height=3, fig.width=8}
library(ComplexUpset)

df <- gene_discovery %>%
  left_join(
    bulk_ident_rds$gene_discovery %>%
      mutate(bulk.n = Illumina + `ONT cDNA` + PacBio + `ONT dRNA`) %>%
      select(gene_id, bulk.n),
    by = c("gene_id" = "gene_id")
  )

# fill na as 0 in bulk.n
df$bulk.n[is.na(df$bulk.n)] <- 0
# df <- df%>% mutate(bulk.n = as.factor(bulk.n))


# remove the bulk Bulk DE levels
ComplexUpset::upset(
  df %>% 
    mutate("HC gene"= case_when(
                bulk.n == 4 ~ "High-confidence (shared by all 4 bulk protocols)",
                bulk.n == 0 ~ "Low-confidence (not detected by any bulk protocol)",
                TRUE ~ "Moderate-confidence (detected in some but not all protocols)"
                )
           ) %>%    
    mutate(`HC gene` = factor(`HC gene`, levels = c(
                "Low-confidence (not detected by any bulk protocol)",
                "Moderate-confidence (detected in some but not all protocols)",
                "High-confidence (shared by all 4 bulk protocols)"
              )))  %>%
    filter(`ONT SC` | `ONT SN` | `PacBio SC` | `PacBio SN` | bulk.n>0 ),
  intersect = c("ONT SC", "ONT SN", "PacBio SC", "PacBio SN"),
  wrap = TRUE,
  base_annotations=list(
    "Intersection size" = intersection_size(
      mapping = aes(fill = `HC gene`),
      text=element_text(size = 3, vjust=-0.2)
    )  +  scale_fill_manual(
              values = c(
                "red",
                "grey",
                "#1a4f07"
              )
          )
  ),
  set_sizes=(
      upset_set_size(
          geom=geom_bar(
              aes(fill=`HC gene`, x=group),
              width=0.8
          ),
          position = 'right',
      ) +  scale_fill_manual(
              values = c(
                "red",
                "grey",
                "#1a4f07"
              )
          ) +
        theme(legend.position = "none")
  ),
  guides='over'
)
```
Sup figure: Separating protein coding and non-protein coding
```{r SUP: Gene Identification Upset plot Overlap with Bulk, fig.height=3, fig.width=16}
p1 <- ComplexUpset::upset(
          df %>% 
            filter(biotype=="protein_coding") %>% 
            mutate("HC gene"= case_when(
                        bulk.n == 4 ~ "High-confidence (shared by all 4 bulk protocols)",
                        bulk.n == 0 ~ "Low-confidence (not detected by any bulk protocol)",
                        TRUE ~ "Moderate-confidence (detected in some but not all protocols)"
                        )
                   ) %>%    
            mutate(`HC gene` = factor(`HC gene`, levels = c(
                        "Low-confidence (not detected by any bulk protocol)",
                        "Moderate-confidence (detected in some but not all protocols)",
                        "High-confidence (shared by all 4 bulk protocols)"
                      )))  %>%
            filter(`ONT SC` | `ONT SN` | `PacBio SC` | `PacBio SN` | bulk.n>0 ),
          intersect = c("ONT SC", "ONT SN", "PacBio SC", "PacBio SN"),
          wrap = TRUE,
          base_annotations=list(
            "Intersection size" = intersection_size(
              mapping = aes(fill = `HC gene`),
              text=element_text(size = 3, vjust=-0.2)
            )  +  scale_fill_manual(
                      values = c(
                        "red",
                        "grey",
                        "#1a4f07"
                      )
                  )
          ),
          set_sizes=(
              upset_set_size(
                  geom=geom_bar(
                      aes(fill=`HC gene`, x=group),
                      width=0.8
                  ),
                  position = 'right',
              ) +  scale_fill_manual(
                      values = c(
                        "red",
                        "grey",
                        "#1a4f07"
                      )
                  ) +
                theme(legend.position = "none")
          ),
          guides='over'
        )
    ggtitle("Protein coding")

p2 <- ComplexUpset::upset(
        df %>% 
          filter(biotype!="protein_coding") %>% 
          mutate("HC gene"= case_when(
                      bulk.n == 4 ~ "High-confidence (shared by all 4 bulk protocols)",
                      bulk.n == 0 ~ "Low-confidence (not detected by any bulk protocol)",
                      TRUE ~ "Moderate-confidence (detected in some but not all protocols)"
                      )
                 ) %>%    
          mutate(`HC gene` = factor(`HC gene`, levels = c(
                      "Low-confidence (not detected by any bulk protocol)",
                      "Moderate-confidence (detected in some but not all protocols)",
                      "High-confidence (shared by all 4 bulk protocols)"
                    )))  %>%
          filter(`ONT SC` | `ONT SN` | `PacBio SC` | `PacBio SN` | bulk.n>0 ),
        intersect = c("ONT SC", "ONT SN", "PacBio SC", "PacBio SN"),
        wrap = TRUE,
        base_annotations=list(
          "Intersection size" = intersection_size(
            mapping = aes(fill = `HC gene`),
            text=element_text(size = 3, vjust=-0.2)
          )  +  scale_fill_manual(
                    values = c(
                      "#B67272",
                      "grey",
                      "#1a4f07"
                    )
                )
        ),
        set_sizes=(
            upset_set_size(
                geom=geom_bar(
                    aes(fill=`HC gene`, x=group),
                    width=0.8
                ),
                position = 'right',
            ) +  scale_fill_manual(
                    values = c(
                      "#B67272",
                      "grey",
                      "#1a4f07"
                    )
                ) +
              theme(legend.position = "none")
        ),
        guides='over'
      )
    ggtitle("Non protein coding")

p1 | p2
```

### Jaccard similarity
```{r gene Jaccard similarity, fig.height=3, fig.width=4.5}
library(ComplexHeatmap)
binary_matrix <- gene_discovery %>% 
  select("ONT SC", "ONT SN", "PacBio SC", "PacBio SN") %>% 
  mutate(across(everything(), as.integer)) %>% filter(
    `ONT SC` |  `ONT SN` |  `PacBio SC` | `PacBio SN`
  )

# Compute Jaccard similarity
jaccard_sim <- function(x, y) {
  sum(x & y) / sum(x | y)
}

platforms <- colnames(binary_matrix)
jaccard_matrix <- outer(platforms, platforms, Vectorize(function(i, j) {
  jaccard_sim(binary_matrix[[i]], binary_matrix[[j]])
}))

# Convert to matrix format
colnames(jaccard_matrix) <- rownames(jaccard_matrix) <- platforms

label_func <- function(j, i, x, y, width, height, fill) {
    grid.text(sprintf("%.2f", jaccard_matrix[i, j]), x, y, gp = gpar(fontsize = 10))
}
Heatmap(jaccard_matrix, 
        name = "Jaccard Similarity", 
        col = colorRampPalette(c("grey", "#06317F"))(50),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(round(jaccard_matrix[i, j], 2), x, y, 
                    gp = gpar(col = "white"))  # Set font color to white
        },
        na_col = "white")  # Keeps upper triangle blank
```
## Transcript level

FilterByExpr
```{r transcript filtering for DE, fig.height=15, fig.width=30}
# Transcript
## 1. get common transcripts
common_tx <- Reduce(intersect, lapply(list(ont_sc.dge, ont_sn.dge, pb_sc.dge, pb_sn.dge), rownames))

ont_sc.dge <- ont_sc.dge[common_tx, ]
ont_sn.dge <- ont_sn.dge[common_tx, ]
pb_sc.dge <- pb_sc.dge[common_tx, ]
pb_sn.dge <- pb_sn.dge[common_tx, ]


tx_discovery <- tibble(
  tx_id = common_tx,
  "ONT SC" = filterByExpr(ont_sc.dge, group = ont_sc.dge$samples$group, min.count=5),
  "ONT SN" = filterByExpr(ont_sn.dge, group = ont_sn.dge$samples$group, min.count=5),
  "PacBio SC" = filterByExpr(pb_sc.dge, group = pb_sc.dge$samples$group, min.count=5),
  "PacBio SN" = filterByExpr(pb_sn.dge, group = pb_sn.dge$samples$group, min.count=5)
) %>% mutate(
  tx_id = strsplit(tx_id, "\\|") %>% sapply(function(x) x[1])
) %>% left_join(tx_ref_df, by = c("tx_id")) %>% 
  mutate(top_biotype = ifelse(biotype %in% 
            names(sort(table(biotype), decreasing = TRUE)[1:3]), 
            biotype, "Others"))
tx_discovery$top_biotype <- factor(tx_discovery$top_biotype, levels = c(sort(unique(tx_discovery$top_biotype)) %>% setdiff("Others"), "Others"))
```

### Upset plot:
```{r transcript upset plot, fig.height=4, fig.width=8}
library(ComplexUpset)
ComplexUpset::upset(
  tx_discovery %>% mutate(
    biotype = case_when(
      biotype =="protein_coding" ~ "Protein-coding",
      TRUE ~ "Non-protein-coding"
    ) 
  ) %>% filter(
      `ONT SC` | `ONT SN` |  `PacBio SC` | `PacBio SN`
    ),
  intersect = c("ONT SC", "ONT SN", "PacBio SC", "PacBio SN"),
  annotations = list(
        # 1st method - passing list:
        "Transcript length" = list(
          aes = aes(x = intersection, y = Length),
          # provide a list if you wish to add several geoms
          geom = geom_boxplot(na.rm = TRUE, outliers = FALSE)
        )
        # ,
        # "BioType" = list(
        #   aes = aes(x = intersection, fill = biotype),
        #   geom = geom_bar(position = "fill")
        # )
  ),
  base_annotations=list(
        'Intersection size'=ComplexUpset::intersection_size(
            counts=FALSE,
            mapping=aes(fill=biotype)
        ) + scale_fill_manual(values = c("Non-protein-coding"="#BAC5CB",
                              "Protein-coding"='#B5C952')) 
    ),
  set_sizes=(
        ComplexUpset::upset_set_size(
            geom=geom_bar(
                aes(fill=biotype),
                width=0.8
            ) ,
            position='right'
        )+ scale_fill_manual(values = c("Non-protein-coding"="#BAC5CB",
                              "Protein-coding"='#B5C952')) + 
          theme(legend.position = "none")
  ),
  guides='over',
  wrap = TRUE
)

```

### Overlap with Bulk discovery

```{r transcript Upset plot Overlap with Bulk, fig.height=3, fig.width=8}
# Transcript 
df <- tx_discovery %>%
  left_join(
    bulk_ident_rds$tx_discovery %>%
      mutate(bulk.n = Illumina + `ONT cDNA` + PacBio + `ONT dRNA`) %>%
      select(tx_id, bulk.n),
    by = c("tx_id" = "tx_id")
  )

# fill na as 0 in bulk.n
df$bulk.n[is.na(df$bulk.n)] <- 0
#df <- df%>% mutate(bulk.n = as.factor(bulk.n))


# remove the bulk Bulk DE levels
ComplexUpset::upset(
  df %>% 
    mutate("HC transcript" = case_when(
                bulk.n == 4 ~ "High-confidence (shared by all 4 bulk protocols)",
                bulk.n == 0 ~ "Low-confidence (not detected by any bulk protocol)",
                TRUE ~ "Moderate-confidence (detected in some but not all protocols)"
                )) %>% 
    mutate(`HC transcript` = factor(`HC transcript`, levels = c(
                "Low-confidence (not detected by any bulk protocol)",
                "Moderate-confidence (detected in some but not all protocols)",
                "High-confidence (shared by all 4 bulk protocols)"
              )))  %>%
    filter(`ONT SC` | `ONT SN` | `PacBio SC` | `PacBio SN` | bulk.n > 0),
  intersect = c("ONT SC", "ONT SN", "PacBio SC", "PacBio SN"),
  wrap = TRUE,
  base_annotations=list(
    "Intersection size" = intersection_size(
      mapping = aes(fill = `HC transcript`),
      text=element_text(size = 3, vjust=-0.2)
    )  +  scale_fill_manual(
              values = c(
                "#B67272",
                "grey",
                "#1a4f07"
              )
          )
  ),
  set_sizes=(
      upset_set_size(
          geom=geom_bar(
              aes(fill=`HC transcript`, x=group),
              width=0.8
          ),
          position = 'right',
      ) +  scale_fill_manual(
              values = c(
                "#B67272",
                "grey",
                "#1a4f07"
              )
          ) +
        theme(legend.position = "none")
  ),
  guides='over'
)
```

Sup figure: Separating protein coding and non-protein coding
```{r SUP: Tx Identification Upset plot Overlap with Bulk, fig.height=3, fig.width=16}
p1 <- ComplexUpset::upset(
          df %>% 
            filter(biotype=="protein_coding")  %>%
            mutate("HC transcript" = case_when(
                bulk.n == 4 ~ "High-confidence (shared by all 4 bulk protocols)",
                bulk.n == 0 ~ "Low-confidence (not detected by any bulk protocol)",
                TRUE ~ "Moderate-confidence (detected in some but not all protocols)"
                )) %>% 
            mutate(`HC transcript` = factor(`HC transcript`, levels = c(
                        "Low-confidence (not detected by any bulk protocol)",
                        "Moderate-confidence (detected in some but not all protocols)",
                        "High-confidence (shared by all 4 bulk protocols)"
                      )))  %>%
            filter(`ONT SC` | `ONT SN` | `PacBio SC` | `PacBio SN` | bulk.n > 0 ),
          intersect = c("ONT SC", "ONT SN", "PacBio SC", "PacBio SN"),
          wrap = TRUE,
          base_annotations=list(
            "Intersection size" = intersection_size(
              mapping = aes(fill = `HC transcript`),
              text=element_text(size = 3, vjust=-0.2)
            )  +  scale_fill_manual(
                      values = c(
                        "#B67272",
                        "grey",
                        "#1a4f07"
                      )
                  )
          ),
          set_sizes=(
              upset_set_size(
                  geom=geom_bar(
                      aes(fill=`HC transcript`, x=group),
                      width=0.8
                  ),
                  position = 'right',
              ) +  scale_fill_manual(
                      values = c(
                        "#B67272",
                        "grey",
                        "#1a4f07"
                      )
                  ) +
                theme(legend.position = "none")
          ),
          guides='over'
        ) +
    ggtitle("Protain coding")

p2 <- ComplexUpset::upset(
          df %>% 
            filter(biotype!="protein_coding")  %>% 
            mutate("HC transcript" = case_when(
                bulk.n == 4 ~ "High-confidence (shared by all 4 bulk protocols)",
                bulk.n == 0 ~ "Low-confidence (not detected by any bulk protocol)",
                TRUE ~ "Moderate-confidence (detected in some but not all protocols)"
                )) %>% 
            mutate(`HC transcript` = factor(`HC transcript`, levels = c(
                        "Low-confidence (not detected by any bulk protocol)",
                        "Moderate-confidence (detected in some but not all protocols)",
                        "High-confidence (shared by all 4 bulk protocols)"
                      )))  %>%
            filter(`ONT SC` | `ONT SN` | `PacBio SC` | `PacBio SN` | bulk.n > 0 ),
          intersect = c("ONT SC", "ONT SN", "PacBio SC", "PacBio SN"),
          wrap = TRUE,
          base_annotations=list(
            "Intersection size" = intersection_size(
              mapping = aes(fill = `HC transcript`),
              text=element_text(size = 3, vjust=-0.2)
            )  +  scale_fill_manual(
                      values = c(
                        "#B67272",
                        "grey",
                        "#1a4f07"
                      )
                  )
          ),
          set_sizes=(
              upset_set_size(
                  geom=geom_bar(
                      aes(fill=`HC transcript`, x=group),
                      width=0.8
                  ),
                  position = 'right',
              ) +  scale_fill_manual(
                      values = c(
                        "#B67272",
                        "grey",
                        "#1a4f07"
                      )
                  ) +
                theme(legend.position = "none")
          ),
          guides='over'
        ) +
    ggtitle("Non protain coding")

p1 | p2
```
# Preprocess Gene and Transcript DGE for DE: Filtering, Normalisation
```{r Filtering}
## 2. filter by expression (Transcript)
df <- gene_discovery %>%
  left_join(
    bulk_ident_rds$gene_discovery %>%
      mutate(bulk.n = Illumina + `ONT cDNA` + PacBio + `ONT dRNA`) %>%
      select(gene_id, bulk.n),
    by = c("gene_id" = "gene_id")
  )
# fill na as 0 in bulk.n
df$bulk.n[is.na(df$bulk.n)] <- 0
df <- df%>% mutate(bulk.n = as.factor(bulk.n))

keep <- df %>% mutate(keep = `ONT SC` & `ONT SN` & `PacBio SC` & `PacBio SN` & bulk.n == 4) %>% pull(keep)


table(keep)
ont_sc.gene.dge <- ont_sc.gene.dge[keep, ]
ont_sn.gene.dge <- ont_sn.gene.dge[keep, ]
pb_sc.gene.dge <- pb_sc.gene.dge[keep, ]
pb_sn.gene.dge <- pb_sn.gene.dge[keep, ]


## 2. filter by expression (Transcript)
df <- tx_discovery %>%
  left_join(
    bulk_ident_rds$tx_discovery %>%
      mutate(bulk.n = Illumina + `ONT cDNA` + PacBio + `ONT dRNA`) %>%
      select(tx_id, bulk.n),
    by = c("tx_id" = "tx_id")
  )
# fill na as 0 in bulk.n
df$bulk.n[is.na(df$bulk.n)] <- 0
df <- df%>% mutate(bulk.n = as.factor(bulk.n))

keep <- df %>% mutate(keep = `ONT SC` & `ONT SN` & `PacBio SC` & `PacBio SN` & bulk.n == 4) %>% pull(keep)
table(keep)

ont_sc.dge <- ont_sc.dge[keep, ]
ont_sn.dge <- ont_sn.dge[keep, ]
pb_sc.dge <- pb_sc.dge[keep, ]
pb_sn.dge <- pb_sn.dge[keep, ]
```

Normalise the gene and transcript counts using `calcNormFactors` with `TMM` methods for each dataset, separately for human

```{r TMM normalisation}
### 4. Normalise the gene counts using `calcNormFactors` with `TMM` methods for each dataset
ont_sc.gene.dge <- calcNormFactors(ont_sc.gene.dge, method = "TMM")
ont_sn.gene.dge <- calcNormFactors(ont_sn.gene.dge, method = "TMM")
pb_sc.gene.dge <- calcNormFactors(pb_sc.gene.dge, method = "TMM")
pb_sn.gene.dge <- calcNormFactors(pb_sn.gene.dge, method = "TMM")


### 4. Normalise the transcript counts using `calcNormFactors` with `TMM` methods for each dataset
ont_sc.dge <- calcNormFactors(ont_sc.dge, method = "TMM")
ont_sn.dge <- calcNormFactors(ont_sn.dge, method = "TMM")
pb_sc.dge <- calcNormFactors(pb_sc.dge, method = "TMM")
pb_sn.dge <- calcNormFactors(pb_sn.dge, method = "TMM")
```

# MDS (Merged)
## Merge DGElist 
NOTE: The merged DGElist will only be used for the MDS plot. The DE analysis will be done separately for each dataset.
```{r Merge DGEs}
# Gene
ont_sc.gene.dge$samples$platform <- "ONT"
ont_sn.gene.dge$samples$platform <- "ONT"
pb_sc.gene.dge$samples$platform <- "PacBio"
pb_sn.gene.dge$samples$platform <- "PacBio"

ont_sc.gene.dge$samples$library <- "SC"
ont_sn.gene.dge$samples$library <- "SN"
pb_sc.gene.dge$samples$library <- "SC"
pb_sn.gene.dge$samples$library <- "SN"

gene.dge.merge <- cbind(
  ont_sc.gene.dge,
  ont_sn.gene.dge,
  pb_sc.gene.dge,
  pb_sn.gene.dge
)

# Transcript
ont_sc.dge$samples$platform <- "ONT"
ont_sn.dge$samples$platform <- "ONT"
pb_sc.dge$samples$platform <- "PacBio"
pb_sn.dge$samples$platform <- "PacBio"

ont_sc.dge$samples$library <- "SC"
ont_sn.dge$samples$library <- "SN"
pb_sc.dge$samples$library <- "SC"
pb_sn.dge$samples$library <- "SN"

tx.dge.merge <- cbind(
  ont_sc.dge,
  ont_sn.dge,
  pb_sc.dge,
  pb_sn.dge
)
```


```{r}
norm_and_mds <- function(dge) {
  dge <- calcNormFactors(dge, method = "TMM")
  mds <- plotMDS(dge, plot = FALSE)
  return(mds)
}
myMDS <- function(dge, mds, label=F) {
  # normalisation
  mds_data <- tibble(
    Dim1 = mds$x,
    Dim2 = mds$y,
    "Cell lines" = dge$sample$sample,
    "Cancer types" = dge$sample$group,
    Platforms = dge$sample$platform,
    "Library types" = dge$sample$library
  )

  p <- ggplot(mds_data, aes(x = Dim1, y = Dim2, color =  `Cancer types`, shape = `Cell lines`)) +
    geom_point(size = 3, alpha = 0.6) + # Plot points +
    labs(title = "MDS plot", 
        x = paste0("Dimension 1 (", round(mds$var.explained[1]*100, 2), "%)" ) , 
        y = paste0("Dimension 2 (", round(mds$var.explained[2]*100, 2), "%)" )) +
    scale_color_viridis_d() + # Set color scale
    theme_minimal() 
  if (label){
p <- ggplot(mds_data, aes(x = Dim1, y = Dim2, color = `Cancer types`, label = `Cell lines`)) +
  geom_text(vjust = 0.05, hjust = 0.05, size = 3) +
  labs(
    title = "MDS plot", 
    x = paste0("Dimension 1 (", round(mds$var.explained[1]*100, 2), "%)"), 
    y = paste0("Dimension 2 (", round(mds$var.explained[2]*100, 2), "%)")
  ) +
  scale_color_viridis_d() +
  theme_minimal()
  }
  return(p)
}
  
```
## Gene count MDS
```{r Gene count MDS, fig.height=3, fig.width=8}
mds <- norm_and_mds(gene.dge.merge) 
p1 <- myMDS(gene.dge.merge, mds) + ggtitle("Human gene counts") +
  scale_shape_manual(values = c(7, 8, 9, 15, 16, 17, 18, 1))
p2 <- myMDS(gene.dge.merge, mds) + ggtitle("Human gene counts")  + 
            aes(color = Platforms, shape = `Library types`) + 
            scale_color_manual(values = color_palette[c("ONT",  "PacBio")] %>% unname())



p3 <- myMDS(gene.dge.merge, mds, label=T) + ggtitle("Human gene counts") +
  scale_shape_manual(values = c(7, 8, 9, 15, 16, 17, 18, 1))
p4 <- myMDS(gene.dge.merge, mds, label=T) + ggtitle("Human gene counts")  + 
            aes(color = Platforms, shape = `Library types`) + 
            scale_color_manual(values = color_palette[c("ONT",  "PacBio")] %>% unname())
p1 | p2 
p3 | p4
```



## Transcript count MDS
```{r Transcript count MDS, fig.height=3, fig.width=8}
mds <- norm_and_mds(tx.dge.merge) 
p1 <- myMDS(tx.dge.merge, mds) + ggtitle("Human transcript counts") +
  scale_shape_manual(values = c(7, 8, 9, 15, 16, 17, 18, 1))
p2 <- myMDS(tx.dge.merge, mds) + ggtitle("Human transcript counts")  + 
            aes(color = Platforms, shape = `Library types`) + 
            scale_color_manual(values = color_palette[c("ONT",  "PacBio")] %>% unname())
p3 <- myMDS(tx.dge.merge, mds, label=T) + ggtitle("Human transcript counts") +
  scale_shape_manual(values = c(7, 8, 9, 15, 16, 17, 18, 1))
p4 <- myMDS(tx.dge.merge, mds, label=T) + ggtitle("Human transcript counts")  + 
            aes(color = Platforms, shape = `Library types`) + 
            scale_color_manual(values = color_palette[c("ONT",  "PacBio")] %>% unname())
p1 | p2 
p3 | p4
```

# PVCA plots
```{r PVCA plots, fig.height=3, fig.width=4}
# Create a design matrix for covariates
library(pvca)
library(Biobase)


# Gene
log_cpm <- cpm(gene.dge.merge, log = TRUE, prior.count = 1) 
# Run PVCA
colnames(log_cpm) <- NULL
expr_set <- ExpressionSet(
  assayData = as.matrix(log_cpm),
  phenoData = AnnotatedDataFrame(gene.dge.merge$samples %>% rename(Platforms=platform, `Cancer_types`=group,  `Library_types`=library))
)
pvca_result <- pvcaBatchAssess(
  expr_set,
  batch.factors = c("Platforms", "Cancer_types", "Library_types"),  # Specify the metadata columns to test
  threshold = 0.1
)

# Transcript
log_cpm <- cpm(tx.dge.merge, log = TRUE, prior.count = 1) 
# Run PVCA
colnames(log_cpm) <- NULL
expr_set <- ExpressionSet(
  assayData = as.matrix(log_cpm),
  phenoData = AnnotatedDataFrame(tx.dge.merge$samples %>% rename(Platforms=platform, `Cancer_types`=group,  `Library_types`=library))
)
pvca_result2 <- pvcaBatchAssess(
  expr_set,
  batch.factors = c("Platforms", "Cancer_types", "Library_types"),  # Specify the metadata columns to test
  threshold = 0.1
)

# Create a data frame from the pvca_result list
plot_df <- as.data.frame(pvca_result$dat %>% t )
rownames(plot_df) <- pvca_result$label
plot_df <- plot_df %>% rownames_to_column("Variable") %>% rename("Value" = V1) %>% mutate(type="Gene-level")

# Create a data frame from the pvca_result list
plot_df2 <- as.data.frame(pvca_result2$dat %>% t )
rownames(plot_df2) <- pvca_result2$label
plot_df2$type <- "Transcript quantification"
plot_df2 <- plot_df2 %>% rownames_to_column("Variable") %>% rename("Value" = V1) %>% mutate(type="Transcript-level")

plot_df <- plot_df %>% bind_rows(plot_df2)

# Create a bar plot using ggplot
ggplot(plot_df, aes(x=type, y = Value, fill = Variable)) +
  geom_bar(stat = "identity", position = "fill") +
  labs( x = "", y = "Weighted average proportion variance") +
  theme_minimal() +
  scale_fill_brewer(palette = "Set3") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

```

### Heatmap （placeholder for potential plot）

# Differential gene expression
## DGE edgeR (Salmon and Oarfish)
```{r}
DGE.edgeR_human <- function(dge, norm.method = "TMM") {
  if (!is.null(norm.method)) {
    dge <- calcNormFactors(dge, method = norm.method)
  }
  
  design <- model.matrix(~0 + group, data = dge$samples)
  dge <- estimateDisp(dge, design)
  qlfit <- glmQLFit(dge, design)
  
  contrasts <- makeContrasts(
    AvsL = groupsclc_a - groupluad,
    PvsL = groupsclc_p - groupluad,
    AvsP = groupsclc_a - groupsclc_p,
    levels = design
  )
  
  AvsL.QLF.test <- glmQLFTest(qlfit, contrast = contrasts[, "AvsL"])
  PvsL.QLF.test <- glmQLFTest(qlfit, contrast = contrasts[, "PvsL"])
  AvsP.QLF.test <- glmQLFTest(qlfit, contrast = contrasts[, "AvsP"])
  
  list(
    AvsL.DE = AvsL.QLF.test,
    PvsL.DE = PvsL.QLF.test,
    AvsP.DE = AvsP.QLF.test 
  )
}

ont_sc.dge_human.list <- DGE.edgeR_human(ont_sc.gene.dge)
ont_sn.dge_human.list <- DGE.edgeR_human(ont_sn.gene.dge)
pb_sc.dge_human.list <- DGE.edgeR_human(pb_sc.gene.dge)
pb_sn.dge_human.list <- DGE.edgeR_human(pb_sn.gene.dge)
```

## MA plot of expected DEG
```{r SC MA Plots, fig.height=3, fig.width=13}

expected_DE <- read_csv(params$expected_de_table)

plot.my.ma <- function(dge){
  # Get results and DE decisions
  res <- dge$table %>% 
    tibble::rownames_to_column("Gene") %>%
    mutate(Gene=sub("\\..*", "", Gene))
  
  # Join with expected_DE
  highlighted <- expected_DE %>%
    filter(Symbol %in% c("HMGA2", "POU2F3", "ASCL1")) %>%
    select(Gene, Symbol) %>%
    inner_join(res, by = "Gene")
  
  dges <- decideTests(dge)
  
  # Add DE status to results
  res$DE <- as.factor(dges[,1])  # Assuming single contrast
  
  # Subsets
  res_nonDE <- res %>% filter(DE == 0)
  res_DE <- res %>% filter(DE != 0)
  # MA plot
  ggplot() +
    geom_point(data = res_nonDE, aes(x = logCPM, y = logFC), color = "gray", alpha = 0.4) +
    geom_point(data = res_DE, aes(x = logCPM, y = logFC, color = DE), alpha = 0.6) +
    scale_color_manual(values = c("-1" = "blue", "1" = "red")) +
    geom_point(data = highlighted, aes(x = logCPM, y = logFC), color = "yellow", shape = 21, fill = NA, size = 2, stroke = 1) +
    geom_text_repel(data = highlighted, aes(x = logCPM, y = logFC, label = Symbol), size = 3, max.overlaps = 100) +
    geom_hline(yintercept = 0, linetype = "dashed", color = "gray") +
    labs(x = "Average log CPM", y = "log₂ Fold Change", color = "DE Status") +
    theme_minimal()
}
ont_sc.dge_human.list %>% map(plot.my.ma) %>% patchwork::wrap_plots(ncol=3, guides='collect')
ont_sn.dge_human.list %>% map(plot.my.ma) %>% patchwork::wrap_plots(ncol=3, guides='collect')
pb_sc.dge_human.list %>% map(plot.my.ma) %>% patchwork::wrap_plots(ncol=3, guides='collect')
pb_sn.dge_human.list %>% map(plot.my.ma) %>% patchwork::wrap_plots(ncol=3, guides='collect')
```

### Cross-platform comparison (upset)
```{r upset plot function}
library(ComplexUpset)
complex_upset_plot <- function(lst, bulk.any = NULL, bulk.all = NULL, retained_feature_set=NULL) {
  # apply decideTests to each list element
  DEGs <- lapply(lst, function(x) {
    x <- x %>% decideTests()
    x <- rownames(x)[x != 0]
    return(x)
  })
  DEGs[["Bulk DE any"]] <- bulk.any
  DEGs[["Bulk DE all"]] <- bulk.all
  long_df <- stack(DEGs)

  # Make the binary membership matrix
  binary_matrix <- long_df %>%
    mutate(present = 1) %>%
    pivot_wider(names_from = ind, values_from = present, values_fill = 0)
  
  if(!is.null(retained_feature_set)){
    binary_matrix$values <- sub("\\..*", "", binary_matrix$values)
    retained_feature_set <- sub("\\..*", "", retained_feature_set)
    binary_matrix <- binary_matrix %>% filter(values %in% retained_feature_set)
  }
  
  binary_matrix$`Bulk DE any` <- binary_matrix$`Bulk DE any` %>% as.logical()
  binary_matrix$`Bulk DE all` <- binary_matrix$`Bulk DE all` %>% as.logical()
  binary_matrix$`Bulk DE` <- factor(
    ifelse(!binary_matrix$`Bulk DE any`, "Not DE in any bulk dataset",
      ifelse(
        binary_matrix$`Bulk DE all`, "DE in all bulk datasets", "DE in any bulk dataset"
      )
    ),
    levels = c("Not DE in any bulk dataset", "DE in any bulk dataset", "DE in all bulk datasets")
  )

  # remove the bulk Bulk DE levels


  p <- ComplexUpset::upset(
    binary_matrix %>% select(-values) %>% as.data.frame(),
    intersect = names(lst),
    wrap = TRUE,
    base_annotations=list(
      "Intersection size" = intersection_size(
        mapping = aes(fill = `Bulk DE`)
      ) + scale_fill_manual(
                values = c(
                  "Not DE in any bulk dataset" = "#B67272",
                  "DE in any bulk dataset" = "grey", 
                  "DE in all bulk datasets" = "#1a4f07"
                )
            ) + theme(legend.position = "none")
    ),
    set_sizes=(
        upset_set_size(
            geom=geom_bar(
                aes(fill=`Bulk DE`, x=group),
                width=0.8
            ),
            position = 'right',
        )  + scale_fill_manual(
                values = c(
                  "Not DE in any bulk dataset" = "#B67272",
                  "DE in any bulk dataset" = "grey", 
                  "DE in all bulk datasets" = "#1a4f07"
                )
            ) 
    ),
    guides='over'
  )

  return(list(
    plot = p,
    matrix = binary_matrix %>% rename(Gene = values)
  ))
}
```

```{r upset plot dge human,  fig.height=10, fig.width=8}
# Upset plot
lst <- list(
  "ONT SC" = ont_sc.dge_human.list$AvsL.DE,
  "ONT SN" = ont_sn.dge_human.list$AvsL.DE,
  "PacBio SC" = pb_sc.dge_human.list$AvsL.DE,
  "PacBio SN" = pb_sn.dge_human.list$AvsL.DE
)
bulk.any <- bulk_de_rds$AvsL.DEG.any
bulk.all <- bulk_de_rds$AvsL.DEG.all
retained_feature_set <- gene_discovery %>%
      left_join(
        bulk_ident_rds$gene_discovery %>%
          mutate(bulk.n = Illumina + `ONT cDNA` + PacBio + `ONT dRNA`) %>%
          select(gene_id, bulk.n),
        by = c("gene_id" = "gene_id")
      ) %>% 
  filter(`ONT SC` & `ONT SN` & `PacBio SC` & `PacBio SN` & bulk.n==4) %>% 
  pull(gene_id)

# fill na as 0 in bulk.n
df$bulk.n[is.na(df$bulk.n)] <- 0
# df <- df%>% mutate(bulk.n = as.factor(bulk.n))
p1 <- complex_upset_plot(lst, bulk.any, bulk.all, retained_feature_set)$plot + ggtitle("SCLC_A vs LUAD") + theme(plot.title = element_text(size =20))
mat.AvsL.DGE <- complex_upset_plot(lst, bulk.any, bulk.all)$matrix

lst <- list(
  "ONT SC" = ont_sc.dge_human.list$PvsL.DE,
  "ONT SN" = ont_sn.dge_human.list$PvsL.DE,
  "PacBio SC" = pb_sc.dge_human.list$PvsL.DE,
  "PacBio SN" = pb_sn.dge_human.list$PvsL.DE
)
bulk.any <- bulk_de_rds$PvsL.DEG.any
bulk.all <- bulk_de_rds$PvsL.DEG.all
p2 <- complex_upset_plot(lst, bulk.any, bulk.all, retained_feature_set)$plot + ggtitle("SCLC_P vs LUAD") + theme(plot.title = element_text(size = 20))
mat.PvsL.DGE <- complex_upset_plot(lst, bulk.any, bulk.all)$matrix

lst <- list(
  "ONT SC" = ont_sc.dge_human.list$AvsP.DE,
  "ONT SN" = ont_sn.dge_human.list$AvsP.DE,
  "PacBio SC" = pb_sc.dge_human.list$AvsP.DE,
  "PacBio SN" = pb_sn.dge_human.list$AvsP.DE
)
bulk.any <- bulk_de_rds$AvsP.DEG.any
bulk.all <- bulk_de_rds$AvsP.DEG.all
p3 <- complex_upset_plot(lst, bulk.any, bulk.all, retained_feature_set)$plot + ggtitle("SCLC_A vs SCLC_P") + theme(plot.title = element_text(size = 20))
mat.AvsP.DGE <- complex_upset_plot(lst, bulk.any, bulk.all)$matrix
(p1 / p2 / p3) + patchwork::plot_annotation(
  title = "DGE Human",
  theme = theme(plot.title = element_text(size =25))
)
```
```{r explore the SC specific DE,  fig.height=5, fig.width=5}
plot_d <- tibble(
  gene = pb_sc.dge_human.list$AvsP.DE$genes %>% rownames,
  AveLogCPM =  pb_sc.dge_human.list$AvsP.DE$AveLogCPM,
  Categories = case_when(gene %in% bulk.all~"common",
                         gene %in% bulk.any~"any",
                         TRUE~"Not in"),
  DE=pb_sc.dge_human.list$AvsP.DE %>% decideTests() %>% unname
) %>% filter(DE != 0 )

# plot 
p1 <- ggplot(plot_d %>% filter(DE==1), aes(x = Categories, y = AveLogCPM)) +
  geom_boxplot() +
  scale_fill_manual(values = c("-1" = "blue", "0" = "grey", "1" = "red")) +
  theme_minimal() +
  labs(fill = "DE") + ggtitle("Up regulated")

p2 <- ggplot(plot_d %>% filter(DE==-1), aes(x = Categories, y = AveLogCPM)) +
  geom_boxplot() +
  scale_fill_manual(values = c("-1" = "blue", "0" = "grey", "1" = "red")) +
  theme_minimal() +
  labs(fill = "DE")+ ggtitle("Down regulated")
p1 | p2

```
```{r}
library(Glimma)
glimmaMDS(pb_sc.gene.dge)
```

### Sup figure: Protein coding gene only
```{r upset plot dge human (Protein-coding),  fig.height=10, fig.width=8}
# Upset plot
lst <- list(
  "ONT SC" = ont_sc.dge_human.list$AvsL.DE,
  "ONT SN" = ont_sn.dge_human.list$AvsL.DE,
  "PacBio SC" = pb_sc.dge_human.list$AvsL.DE,
  "PacBio SN" = pb_sn.dge_human.list$AvsL.DE
)
bulk.any <- bulk_de_rds$AvsL.DEG.any
bulk.all <- bulk_de_rds$AvsL.DEG.all
retained_feature_set <- gene_discovery %>%
      left_join(
        bulk_ident_rds$gene_discovery %>%
          mutate(bulk.n = Illumina + `ONT cDNA` + PacBio + `ONT dRNA`) %>%
          select(gene_id, bulk.n),
        by = c("gene_id" = "gene_id")
      ) %>% 
  filter(`ONT SC` & `ONT SN` & `PacBio SC` & `PacBio SN` & bulk.n==4 & biotype=="protein_coding") %>% 
  pull(gene_id)

# fill na as 0 in bulk.n
df$bulk.n[is.na(df$bulk.n)] <- 0
# df <- df%>% mutate(bulk.n = as.factor(bulk.n))
p1 <- complex_upset_plot(lst, bulk.any, bulk.all, retained_feature_set)$plot + ggtitle("SCLC_A vs LUAD") + theme(plot.title = element_text(size =20))
mat.AvsL.DGE <- complex_upset_plot(lst, bulk.any, bulk.all)$matrix

lst <- list(
  "ONT SC" = ont_sc.dge_human.list$PvsL.DE,
  "ONT SN" = ont_sn.dge_human.list$PvsL.DE,
  "PacBio SC" = pb_sc.dge_human.list$PvsL.DE,
  "PacBio SN" = pb_sn.dge_human.list$PvsL.DE
)
bulk.any <- bulk_de_rds$PvsL.DEG.any
bulk.all <- bulk_de_rds$PvsL.DEG.all
p2 <- complex_upset_plot(lst, bulk.any, bulk.all, retained_feature_set)$plot + ggtitle("SCLC_P vs LUAD") + theme(plot.title = element_text(size = 20))
mat.PvsL.DGE <- complex_upset_plot(lst, bulk.any, bulk.all)$matrix

lst <- list(
  "ONT SC" = ont_sc.dge_human.list$AvsP.DE,
  "ONT SN" = ont_sn.dge_human.list$AvsP.DE,
  "PacBio SC" = pb_sc.dge_human.list$AvsP.DE,
  "PacBio SN" = pb_sn.dge_human.list$AvsP.DE
)
bulk.any <- bulk_de_rds$AvsP.DEG.any
bulk.all <- bulk_de_rds$AvsP.DEG.all
p3 <- complex_upset_plot(lst, bulk.any, bulk.all, retained_feature_set)$plot + ggtitle("SCLC_A vs SCLC_P") + theme(plot.title = element_text(size = 20))
mat.AvsP.DGE <- complex_upset_plot(lst, bulk.any, bulk.all)$matrix
(p1 / p2 / p3) + patchwork::plot_annotation(
  title = "DGE Human (Protein coding only)",
  theme = theme(plot.title = element_text(size =25))
)
```
### Sup figure: Non Protein coding gene only
```{r upset plot dge human (Non-Protein-coding),  fig.height=10, fig.width=8}
# Upset plot
lst <- list(
  "ONT SC" = ont_sc.dge_human.list$AvsL.DE,
  "ONT SN" = ont_sn.dge_human.list$AvsL.DE,
  "PacBio SC" = pb_sc.dge_human.list$AvsL.DE,
  "PacBio SN" = pb_sn.dge_human.list$AvsL.DE
)
bulk.any <- bulk_de_rds$AvsL.DEG.any
bulk.all <- bulk_de_rds$AvsL.DEG.all
retained_feature_set <- gene_discovery %>%
      left_join(
        bulk_ident_rds$gene_discovery %>%
          mutate(bulk.n = Illumina + `ONT cDNA` + PacBio + `ONT dRNA`) %>%
          select(gene_id, bulk.n),
        by = c("gene_id" = "gene_id")
      ) %>% 
  filter(`ONT SC` & `ONT SN` & `PacBio SC` & `PacBio SN` & bulk.n==4 & biotype!="protein_coding") %>% 
  pull(gene_id)

# fill na as 0 in bulk.n
df$bulk.n[is.na(df$bulk.n)] <- 0
# df <- df%>% mutate(bulk.n = as.factor(bulk.n))
p1 <- complex_upset_plot(lst, bulk.any, bulk.all, retained_feature_set)$plot + ggtitle("SCLC_A vs LUAD") + theme(plot.title = element_text(size =20))
mat.AvsL.DGE <- complex_upset_plot(lst, bulk.any, bulk.all)$matrix

lst <- list(
  "ONT SC" = ont_sc.dge_human.list$PvsL.DE,
  "ONT SN" = ont_sn.dge_human.list$PvsL.DE,
  "PacBio SC" = pb_sc.dge_human.list$PvsL.DE,
  "PacBio SN" = pb_sn.dge_human.list$PvsL.DE
)
bulk.any <- bulk_de_rds$PvsL.DEG.any
bulk.all <- bulk_de_rds$PvsL.DEG.all
p2 <- complex_upset_plot(lst, bulk.any, bulk.all, retained_feature_set)$plot + ggtitle("SCLC_P vs LUAD") + theme(plot.title = element_text(size = 20))
mat.PvsL.DGE <- complex_upset_plot(lst, bulk.any, bulk.all)$matrix

lst <- list(
  "ONT SC" = ont_sc.dge_human.list$AvsP.DE,
  "ONT SN" = ont_sn.dge_human.list$AvsP.DE,
  "PacBio SC" = pb_sc.dge_human.list$AvsP.DE,
  "PacBio SN" = pb_sn.dge_human.list$AvsP.DE
)
bulk.any <- bulk_de_rds$AvsP.DEG.any
bulk.all <- bulk_de_rds$AvsP.DEG.all
p3 <- complex_upset_plot(lst, bulk.any, bulk.all, retained_feature_set)$plot + ggtitle("SCLC_A vs SCLC_P") + theme(plot.title = element_text(size = 20))
mat.AvsP.DGE <- complex_upset_plot(lst, bulk.any, bulk.all)$matrix
(p1 / p2 / p3) + patchwork::plot_annotation(
  title = "DGE Human (Non-Protein-coding only)",
  theme = theme(plot.title = element_text(size =25))
)
```

# Differential transcript expression
```{r}
DTE.edgeR_human <- function(dge, norm.method="TMM") {
  if (!is.null(norm.method)) {
     dge <- calcNormFactors(dge, method = norm.method)
  }

  design <- model.matrix(~0+group, data = dge$samples)
  dge <- estimateDisp(dge, design)
  qlfit <- glmQLFit(dge, design)
  contrasts <- makeContrasts(AvsL=groupsclc_a-groupluad, PvsL=groupsclc_p-groupluad, AvsP = groupsclc_a-groupsclc_p, levels=design)
  # SCLC_A vs LUAD
  AvsL.QLF.test <- glmQLFTest(qlfit, contrast=contrasts[,"AvsL"])
  # SCLC_P vs LUAD
  PvsL.QLF.test <- glmQLFTest(qlfit, contrast=contrasts[,"PvsL"])
  # SCLC_A vs SCLC_P
  AvsP.QLF.test <- glmQLFTest(qlfit, contrast=contrasts[,"AvsP"])

 list(
    AvsL.DE = AvsL.QLF.test,
    PvsL.DE = PvsL.QLF.test,
    AvsP.DE = AvsP.QLF.test
  )
}
```


```{r}
design <- readRDS(
  system.file("RNAseq123/design.rds", package = "Glimma"))
contr.matrix <- readRDS(
  system.file("RNAseq123/contr.matrix.rds", package = "Glimma"))
dge <- readRDS(system.file("RNAseq123/dge.rds", package = "Glimma"))
dge <- estimateDisp(dge, design)
gfit <- glmQLFit(dge, design)
glrt <- glmQLFTest(gfit, design, contrast = contr.matrix)
glimmaMA(glrt) # glimmaMA(glrt, dge = dge) to use edgeR results
```


## DTE edgeR+scaling
```{r get dte scaling}
# scale
ont_sc.tx.scale_dge <- ont_sc.dge
ont_sn.tx.scale_dge <- ont_sn.dge
pb_sc.tx.scale_dge <- pb_sc.dge
pb_sn.tx.scale_dge <- pb_sn.dge

ont_sc.tx.scale_dge$counts  <- ont_sc.tx.scale_dge$counts / ont_sc.tx.scale_dge$genes$Overdispersion
ont_sn.tx.scale_dge$counts  <- ont_sn.tx.scale_dge$counts / ont_sn.tx.scale_dge$genes$Overdispersion
pb_sc.tx.scale_dge$counts  <- pb_sc.tx.scale_dge$counts / pb_sc.tx.scale_dge$genes$Overdispersion
pb_sn.tx.scale_dge$counts  <- pb_sn.tx.scale_dge$counts / pb_sn.tx.scale_dge$genes$Overdispersion

# get human list
ont_sc.scl_dte_human.list <- DTE.edgeR_human(ont_sc.tx.scale_dge)
ont_sn.scl_dte_human.list <- DTE.edgeR_human(ont_sn.tx.scale_dge)
pb_sc.scl_dte_human.list <- DTE.edgeR_human(pb_sc.tx.scale_dge)
pb_sn.scl_dte_human.list <- DTE.edgeR_human(pb_sn.tx.scale_dge)
```

### Upset plot
### DTE Upset plot (human transcript)
```{r upset plot dte human edgeR with scaling, fig.height=10, fig.width=8}

# Upset plot
## SCLC_A vs LUAD (AvsL)
lst <- list(
  "ONT SC" = ont_sc.scl_dte_human.list$AvsL.DE,
  "ONT SN" = ont_sn.scl_dte_human.list$AvsL.DE,
  "PacBio SC" = pb_sc.scl_dte_human.list$AvsL.DE,
  "PacBio SN" = pb_sn.scl_dte_human.list$AvsL.DE
)
bulk.any <- bulk_de_rds$AvsL.DTE.any
bulk.all <- bulk_de_rds$AvsL.DTE.all
p1 <- complex_upset_plot(lst, bulk.any, bulk.all)$plot + ggtitle("SCLC_A vs LUAD") + theme(plot.title = element_text(size = 20))
mat.AvsL.DTE <- complex_upset_plot(lst, bulk.any, bulk.all)$matrix

lst <- list(
  "ONT SC" = ont_sc.scl_dte_human.list$PvsL.DE,
  "ONT SN" = ont_sn.scl_dte_human.list$PvsL.DE,
  "PacBio SC" = pb_sc.scl_dte_human.list$PvsL.DE,
  "PacBio SN" = pb_sn.scl_dte_human.list$PvsL.DE
)
bulk.any <- bulk_de_rds$PvsL.DTE.any
bulk.all <- bulk_de_rds$PvsL.DTE.all
p2 <- complex_upset_plot(lst, bulk.any, bulk.all)$plot + ggtitle("SCLC_P vs LUAD") + theme(plot.title = element_text(size = 20))
mat.PvsL.DTE <- complex_upset_plot(lst, bulk.any, bulk.all)$matrix

lst <- list(
  "ONT SC" = ont_sc.scl_dte_human.list$AvsP.DE,
  "ONT SN" = ont_sn.scl_dte_human.list$AvsP.DE,
  "PacBio SC" = pb_sc.scl_dte_human.list$AvsP.DE,
  "PacBio SN" = pb_sn.scl_dte_human.list$AvsP.DE
)
bulk.any <- bulk_de_rds$AvsP.DTE.any
bulk.all <- bulk_de_rds$AvsP.DTE.all
p3 <- complex_upset_plot(lst, bulk.any, bulk.all)$plot + ggtitle("SCLC_A vs SCLC_P") + theme(plot.title = element_text(size = 20))
mat.AvsP.DTE <- complex_upset_plot(lst, bulk.any, bulk.all)$matrix

p1 / p2 / p3 + patchwork::plot_annotation(
  title = "DTE Human (with scaling)",
  theme = theme(plot.title = element_text(size =25))
)
```




# Unincluded analysis
## Understand the difference (SCLC_A vs LUAD)
```{r}
volcano_plot <- function(rst,  main="Volcano plot", xlab="M", ylab="-10*log(P-val)", highligh_gene = NULL) {
  if (is.null(highligh_gene)) {
    dot_col <- ifelse(rst$table$PValue < 0.05 & rst$table$logFC > 0, "red",
      ifelse(rst$table$PValue < 0.05 & rst$table$logFC < 0, "blue", "black")
    )
  } else {
    dot_col <- ifelse(rst$table %>% rownames() %in% highligh_gene, "red", "grey")
  }
  plot(rst$table$logFC, -10 * log10(rst$table$PValue),
    main = main, xlab = xlab, ylab = ylab,
    pch = 20, col = dot_col,
                      )
}

# SCLC_A vs LUAD
par(mfrow=c(2,2))
plotMD(ont_sc.dge_human.list$AvsL.DE, main = "ONT SC")
plotMD(ont_sn.dge_human.list$AvsL.DE, main = "ONT SN")
plotMD(pb_sc.dge_human.list$AvsL.DE, main = "PacBio SC")
plotMD(pb_sn.dge_human.list$AvsL.DE, main = "PacBio SN")




# filter1
filtered_gene <- mat.AvsL.DGE %>%
  filter(`Bulk DE any` == F & `ONT SC` == 1 & `PacBio SC` == 1) %>%
  pull(Gene)
plotMD(ont_sc.dge_human.list$AvsL.DE[filtered_gene, ], main = "ONT SC")
plotMD(ont_sn.dge_human.list$AvsL.DE[filtered_gene, ], main = "ONT SN")
plotMD(pb_sc.dge_human.list$AvsL.DE[filtered_gene, ], main = "PacBio SC")
plotMD(pb_sn.dge_human.list$AvsL.DE[filtered_gene, ], main = "PacBio SN")

par(mfrow=c(2,2))
volcano_plot(ont_sc.dge_human.list$AvsL.DE, main = "ONT SC", highligh_gene = filtered_gene)
volcano_plot(ont_sn.dge_human.list$AvsL.DE, main = "ONT SN", highligh_gene = filtered_gene)
volcano_plot(pb_sc.dge_human.list$AvsL.DE, main = "PacBio SC", highligh_gene = filtered_gene)
volcano_plot(pb_sn.dge_human.list$AvsL.DE, main = "PacBio SN", highligh_gene = filtered_gene)


filtered_tx <- mat.AvsL.DTE %>% 
  filter(`Bulk DE any` == F & `ONT SC` == 1 & `PacBio SC` == 1) %>%
  pull(Gene)
```
